---
title: "00.7_Extract-T2DLoci-Results.Rmd"
author: "Jason Torres"
date: "August 22, 2016"
output:
  pdf_document:
    toc: yes
  html_document:
    fig_caption: yes
    highlight: textmate
    number_sections: yes
    theme: journal
    toc: yes
---

Need to generate file "t2d.loci.df" with these fields:
names(t2d.loci.df)

[1] "chrom"            "win.start"        "win.end"          "win.length"       "most.sig.snp"    
[6] "sig.snp.pos"      "min.gwas.pval"    "reported.gene"    "reported.txstart" "reported.txend"  
[11] "nearby.gene"      "nearby.txstart"   "nearby.txend"

```{r setup}

"%&%" <- function(a,b) paste0(a,b) 

library("data.table")
library("dplyr")

root.dir <- "/Users/jtorres/Google Drive/Projects/MetaXcan/MetaXcanT2D/"
rds.dir <- root.dir %&% "RDS.gtexV6p/"

ggd.df <- readRDS(rds.dir%&%"ggd.ref.df.RDS")

server.dir <- "/Volumes/im-lab/nas40t2/jason/projects/ReferenceData/"
diag37.path  <- server.dir %&% "diagramT2D/build37/diagram.mega-meta.build37.txt.gz"
diag37.df <- fread("cat " %&% diag37.path %&% " | zmore")

```


Process reference df of nearest genes to significant GWAS snps; get SNP with the lowest p-value 

```{r process_df}


process_neargene_df <- function(){
  df  <- readRDS(rds.dir%&%"neargene.10min6.df.RDS")
  pval <- c()
  for (i in 1:dim(df)[1]){
    snp <- df[i,]$rsid
    p <- filter(diag37.df,SNP==snp)$P
    pval <- append(pval,p)
  }
  out.df <- cbind(df,pval)
  return(out.df)
}

g.df <- process_neargene_df()

prune_down_df <- function(g.df){
  genes <- unique(g.df$gene)
  out.df <- c() 
  for (g in genes){
    sub.df <- (filter(g.df,gene==g) %>% arrange(pval))[1,]
    out.df <- rbind(out.df,sub.df)
  }
  return(out.df)
}

ref.df <- prune_down_df(g.df)

add_window_info <- function(ref.df){
  win.start <- c()
  win.end <- c()
  win.length <- c()
  for (i in 1:dim(ref.df)[1]){
    tx.start <- as.integer(ref.df[i,]$tx.start)
    tx.end <- as.integer(ref.df[i,]$tx.end)
    lower.bound <- min(tx.start,tx.end)
    upper.bound <- max(tx.start,tx.end)
    ws <- lower.bound - (10^6)
    we <- upper.bound + (10^6)
    wl <- we - ws
    win.start <- append(win.start,ws)
    win.end <- append(win.end,we)
    win.length <- append(win.length,wl)
  }
  out.df <- cbind(ref.df, win.start,win.end,win.length)
  out.df$tx.start <- as.integer(out.df$tx.start)
  out.df$tx.end <- as.integer(out.df$tx.end)
  out.df$win.start <- as.integer(out.df$win.start)
  out.df$win.end <- as.integer(out.df$win.end)
  out.df$win.length <- as.integer(out.df$win.length)
  return(out.df)
}

ref.df <- add_window_info(ref.df)

```


# Build T2D loci df 


```{r build_t2dlocidf}

build_t2d_loci_df <- function(){
  ggd.temp.df <- mutate(ggd.df,lower.bound = pmin(start,end)) %>% mutate(upper.bound = (pmax(start,end)))
  chrom <- c() 
  win.start <- c()
  win.end <- c()
  win.length <- c()
  most.sig.snp <- c()    
  sig.snp.pos <- c()
  min.gwas.pval <- c()
  reported.gene <- c()
  reported.txstart <- c() 
  reported.txend <- c()  
  nearby.gene <- c()
  nearby.txstart <- c()
  nearby.txend <- c()
  for (i in 1:dim(ref.df)[1]){
    c <- ref.df[i,]$chrom
    ws <- ref.df[i,]$win.start
    we <- ref.df[i,]$win.end
    wl <- ref.df[i,]$win.length
    mss <- ref.df[i,]$rsid
    ssp <- ref.df[i,]$pos
    mgp <- ref.df[i,]$pval
    rg <- ref.df[i,]$gene
    print(rg)
    rtss <- ref.df[i,]$tx.start
    rtes <- ref.df[i,]$tx.end
    # determine which genes are within a window 
    sub.df <- filter(ggd.temp.df,lower.bound >= ws & upper.bound <= we &
                       biotype=="protein_coding" & chr == c)
    for (e in 1:dim(sub.df)[1]){
      ng <- sub.df[e,]$symbol
      ntss <- sub.df[e,]$start
      ntes <- sub.df[e,]$end
      chrom <- append(chrom,c) 
      win.start <- append(win.start,ws)
      win.end <- append(win.end,we)
      win.length <- append(win.length,wl)
      most.sig.snp <- append(most.sig.snp,mss)    
      sig.snp.pos <- append(sig.snp.pos,ssp)
      min.gwas.pval <- append(min.gwas.pval,mgp)
      reported.gene <- append(reported.gene,rg)
      reported.txstart <- append(reported.txstart,rtss) 
      reported.txend <- append(reported.txend,rtes)  
      nearby.gene <- append(nearby.gene,ng)
      nearby.txstart <- append(nearby.txstart,ntss)
      nearby.txend <- append(nearby.txend,ntes)      
    }
  }
  out.df <- cbind(chrom,win.start,win.end,win.length,most.sig.snp,sig.snp.pos,
                  min.gwas.pval,reported.gene,reported.txstart,reported.txend,
                  nearby.gene,nearby.txstart,nearby.txend)
  out.df <- as.data.frame(out.df,stringsAsFactors=FALSE)
  out.df$chrom <- as.integer(out.df$chrom)
  out.df$win.start <- as.integer(out.df$win.start)
  out.df$win.end <- as.integer(out.df$win.end)
  out.df$win.length <- as.integer(out.df$win.length)
  out.df$sig.snp.pos <- as.integer(out.df$sig.snp.pos)
  out.df$min.gwas.pval <- as.numeric(out.df$min.gwas.pval)
  out.df$reported.txstart <- as.integer(out.df$reported.txstart)
  out.df$reported.txend <- as.integer(out.df$reported.txend)
  out.df$nearby.txstart <- as.integer(out.df$nearby.txstart)
  out.df$nearby.txend <- as.integer(out.df$nearby.txend)
  return(out.df) 
}

t2d.loci.df <- build_t2d_loci_df()
saveRDS(t2d.loci.df,file=rds.dir%&%"t2d.loci.df.RDS")

```

