---
title: "Implement analytic workflow"
author: "Jason Torres"
date: "August 6, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup
```{r}

"%&%" <- function(a,b) paste0(a,b)
library("dplyr")
library("data.table")

serv.dir <- "/Users/jtorres/FUSE4/"
root.dir <- serv.dir %&% "projects/MetaXcan/"
res.dir <- root.dir %&% "results/tables/GTExV6p/" # DIAGRAM results with COLOC output appended 
meta.dir <- root.dir %&% "results/meta-analyses/DIAGRAM-GERA_ImpG_0.80_gtexV6p/tables/"
save.dir <- root.dir %&% "MetaXcanT2D/analyzed_result_files/"

pfile <- res.dir %&% "DIAGRAM_ImpG_0.8_gtexV6p.0.5.pvalue.table.csv.gz"
p.df <- fread("cat " %&% pfile %&% " | zmore")
gene.vec <- p.df$Gene

ens.df <- fread("cat " %&% res.dir %&% 
                "DIAGRAM_ImpG_0.8_gtexV6p.0.5.ensid.table.csv.gz" %&% " | zmore")
z.df <- fread("cat " %&% res.dir %&% 
                "DIAGRAM_ImpG_0.8_gtexV6p.0.5.zscore.table.csv.gz" %&% " | zmore")
r2.df <- fread("cat " %&% res.dir %&% 
                 "DIAGRAM_ImpG_0.8_gtexV6p.0.5.r2.table.csv.gz" %&% " | zmore")
ppp.df <- fread("cat " %&% res.dir %&% 
                 "DIAGRAM_ImpG_0.8_gtexV6p.0.5.ppp.table.csv.gz" %&% " | zmore")
nsm.df <- fread("cat " %&% res.dir %&% 
                  "DIAGRAM_ImpG_0.8_gtexV6p.0.5.nsnpsmod.table.csv.gz" %&% " | zmore")
ns.df <- fread("cat " %&% res.dir %&% 
                 "DIAGRAM_ImpG_0.8_gtexV6p.0.5.nsnps.table.csv.gz" %&% " | zmore")
ph3.df <- fread("cat " %&% res.dir %&% 
                  "DIAGRAM_ImpG_0.8_gtexV6p.0.5.ph3.table.csv.gz" %&% " | zmore")
ph4.df <- fread("cat " %&% res.dir %&% 
                  "DIAGRAM_ImpG_0.8_gtexV6p.0.5.ph4.table.csv.gz" %&% " | zmore")
gera.z.df <- fread("cat " %&% res.dir %&% 
                     "GERA_ImpG_0.8_gtexV6p.0.5.zscore.table.csv.gz" %&% " | zmore")
gera.p.df <- fread("cat " %&% res.dir %&% 
                     "GERA_ImpG_0.8_gtexV6p.0.5.pvalue.table.csv.gz" %&% " | zmore")
gera.ns.df <- fread("cat " %&% res.dir %&% 
                      "GERA_ImpG_0.8_gtexV6p.0.5.nsnps.table.csv.gz" %&% " | zmore")
meta.z.df <- fread("cat " %&% meta.dir %&% 
                      "DIAGRAM-GERA_ImpG_0.80_gtexV6p.0.5.zscore.table.csv.gz" %&% " | zmore")
meta.p.df <- fread("cat " %&% meta.dir %&% 
                      "DIAGRAM-GERA_ImpG_0.80_gtexV6p.0.5.pvalue.table.csv.gz" %&% " | zmore")
meta.moresig.df <- fread("cat " %&% meta.dir %&% 
                      "DIAGRAM-GERA_ImpG_0.80_gtexV6p.0.5.moresig.table.csv.gz" %&% " | zmore")

```


# Step 1 

Perform PrediXcan or S-PrediXcan using all tissues. Use Bonferroni correction for all gene-tissue pairs: keep p< 0.05/ number of gene-tissue pairs tested.

```{r}

build_p.mat.gtex <- function(){
  gene.index <- (1:length(colnames(p.df)))[grepl("Gene",colnames(p.df))]
  dgn.index <- (1:length(colnames(p.df)))[grepl("DGN",colnames(p.df))]
  keep.vec <- (1:length(colnames(p.df)))[!((1:length(colnames(p.df))) %in% c(gene.index,dgn.index))]
  p.mat <- as.data.frame(as.data.frame(p.df)[,keep.vec]) 
  return(p.mat)
}

build_p.mat.dgn <- function(){
  gene.index <- (1:length(colnames(p.df)))[grepl("Gene",colnames(p.df))]
  dgn.index <- (1:length(colnames(p.df)))[grepl("DGN",colnames(p.df))]
  gtex.index <- (1:length(colnames(p.df)))[!grepl("DGN",colnames(p.df))]
  keep.vec <- (1:length(colnames(p.df)))[!((1:length(colnames(p.df))) %in% c(gtex.index))]
  p.mat <- as.data.frame(as.data.frame(p.df)[,keep.vec])
  names(p.mat) <- names(p.df)[dgn.index]
  return(p.mat)
}

build_p.mat.full <- function(){
  gene.index <- (1:length(colnames(p.df)))[grepl("Gene",colnames(p.df))]
  keep.vec <- (1:length(colnames(p.df)))[!((1:length(colnames(p.df))) %in% c(gene.index))]
  p.mat <- as.data.frame(as.data.frame(p.df)[,keep.vec]) 
  return(p.mat)
}


build.sig.df <- function(p.mat){

  not.missing <- sapply(1:dim(p.mat)[2],function(i){
    count <- length(na.omit(p.mat[,i]))
    return(count)
  })
  names(not.missing) <- colnames(p.mat)
  num.pairs <- sum(not.missing)

  bonfer <- 0.05 / num.pairs # 2.561633e-07  
  print("Bonferroni threshold: p <= " %&% bonfer)
  sig.df <- c() 
  for (i in 1:dim(p.mat)[2]){
    vec <- p.mat[,i]; vec[is.na(vec)] <- 1
    sig <- vec <= bonfer
    #tiss.vec <- rep(gsub("TW_","",colnames(p.mat)[i]),sum(sig))
    tiss.vec <- rep(colnames(p.mat)[i],sum(sig))
    p.vec <- vec[sig]
    g.vec <- gene.vec[sig]
    df <- data.frame(tissue=tiss.vec,gene=g.vec,pvalue=p.vec,
                     stringsAsFactors = FALSE)
    sig.df <- rbind(sig.df,df)
  }
  return(sig.df)
}


```


**Functions for retrieving additional information for significant associations** 


```{r}

get_vals <- function(sig.df, ref.df,is.string=FALSE){
  pb <- txtProgressBar(min=0,max=dim(sig.df)[1],initial=0,style=3)
  vec <- sapply(1:dim(sig.df)[1],function(i){
    setTxtProgressBar(pb,i)
    tiss <- sig.df[i,1]; gene <- sig.df[i,2]
    
    sub.df <- dplyr::select(ref.df,one_of("Gene",tiss)) %>% filter(Gene==gene) %>% 
      dplyr::select(one_of(tiss))
    if (is.string==TRUE){
      val <- as.character(sub.df)
    } else{
      val <- as.numeric(sub.df)
    }
    return(val)
  })
  return(vec)
}

append_to_sigdf <- function(sig.df){
  ensid <- get_vals(sig.df, ens.df, is.string=TRUE)
  zscore <- get_vals(sig.df, z.df)
  pred.perf.r2 <- get_vals(sig.df, r2.df) 
  pred.perf.pvalue <- get_vals(sig.df, ppp.df) 
  n.snps.model <- get_vals(sig.df, nsm.df) 
  n.snps <- get_vals(sig.df, ns.df)
  temp.df <- sig.df
  temp.df$tissue <- gsub("TW_Whole_Blood_DGN","TW_Whole_Blood",temp.df$tissue)
  ph3 <- get_vals(temp.df, ph3.df)
  ph4 <- get_vals(temp.df, ph4.df)
  #if (is.list(ph3)==TRUE & is.vector(ph3)==TRUE){
  #  ph3 <- rep(NA,length(ensid))
  #  ph4 <- rep(NA,length(ensid))    
  #}
  gera.zscore <- get_vals(sig.df, gera.z.df)
  gera.pvalue <- get_vals(sig.df, gera.p.df)
  gera.nsnps <- get_vals(sig.df, gera.ns.df)
  meta.zscore <- get_vals(sig.df, meta.z.df)
  meta.pvalue <- get_vals(sig.df, meta.p.df)
  meta.more.sig <- get_vals(sig.df, meta.moresig.df)
  out.df <- cbind(ensid,sig.df,zscore,pred.perf.r2,pred.perf.pvalue,n.snps.model,n.snps,
                  ph3,ph4,gera.zscore,gera.pvalue,gera.nsnps,
                  meta.zscore,meta.pvalue,meta.more.sig)
  out.df$tissue <- gsub("TW_","",out.df$tissue)
  out.df$ensid <- as.character(out.df$ensid)
  return(out.df)
}


```


**Append nearest gene information** 


```{r}

library("GenomicRanges")
library("GenomicFeatures")
library("org.Hs.eg.db")
library("annotate")
library("Homo.sapiens")

get_tx_gr <- function(ensid){
  tx.chrom <- tryCatch({(select(Homo.sapiens,key=ensid,keytype="ENSEMBL",
                                columns="TXCHROM"))$TXCHROM}, 
                       warning=function(war){return(NA)},
                       error=function(err){return(NA)})   
  tx.start <- tryCatch({(select(Homo.sapiens,key=ensid,keytype="ENSEMBL",
                                columns="TXSTART"))$TXSTART}, 
                       warning=function(war){return(NA)},
                       error=function(err){return(NA)})                          
  tx.end <- tryCatch({(select(Homo.sapiens,key=ensid,keytype="ENSEMBL",
                                columns="TXEND"))$TXEND}, 
                       warning=function(war){return(NA)},
                       error=function(err){return(NA)}) 
  tx.strand <- tryCatch({(select(Homo.sapiens,key=ensid,keytype="ENSEMBL",
                                columns="TXSTRAND"))$TXSTRAND}, 
                       warning=function(war){return(NA)},
                       error=function(err){return(NA)})   
  #print(tx.chrom %&% " : " %&% tx.start %&% " : " %&% tx.end %&% " : " %&% tx.strand)
  tx.gr <- tryCatch({(GRanges(tx.chrom,IRanges(tx.start, tx.end),tx.strand))}, 
                       warning=function(war){return(NA)},
                       error=function(err){return(NA)})  
  return(tx.gr)
}

append_tx_info <- function(sig.df){ # output from append_to_sigdf
  chrom <- c()
  tx.start <- c()
  tx.end <- c()
  tx.strand <- c()
  pb <- txtProgressBar(min=0,max=length(sig.df$ensid),initial=0,style=3)
  for (i in 1:length(sig.df$ensid)){
    setTxtProgressBar(pb,i)
    ensid <- sig.df$ensid[i]
    #print(ensid)
    tx.gr <- suppressMessages(get_tx_gr(ensid))
    if (is.na(tx.gr)){
      chrom <- append(chrom,NA)
      tx.start <- append(tx.start,NA)
      tx.end <- append(tx.end,NA)
      tx.strand <- append(tx.strand,NA)      
    } else{
      c <- as.character(seqnames(tx.gr))[1]; chrom <- append(chrom,c)
      ts <- paste0(unique(start(tx.gr)),collapse=","); tx.start <- append(tx.start,ts)
      te <- paste0(unique(end(tx.gr)),collapse=","); tx.end <- append(tx.end,te)
      tst <- paste0(unique(strand(tx.gr)),collapse=","); tx.strand <- append(tx.strand,tst)      
    }
  }
  out.df <- cbind(chrom,tx.start,tx.end,tx.strand,sig.df)
  out.df$chrom <- as.character(out.df$chrom)
  out.df$tx.start <- as.character(out.df$tx.start)
  out.df$tx.end <- as.character(out.df$tx.end)
  out.df$tx.strand <- as.character(out.df$tx.strand)
  return(out.df)
}

```


# Steps 2-4 

2. Keep associations with significant prediction performance adjusting for number of PrediXcan significant gene-tissue pairs: keep prediction performance p-values < 0.05/(number of significant associations from previous step).

3. Filter out LD-contaminated associations, i.e. gene-tissue pairs in the ‘independent signal’ (=‘non-colocalized’) region of the ternary plot (See Figure 3A): keep COLOC P3> 0.5 (Blue and gray regions in Figure 3A).

4. If further reduction of number of genes to be taken to replication or validation is desired, filter out hits without explicit evidence of colocalization: keep only P4> 0.5 (Blue region in Figure 3A).

**Functions for steps 2,3, and step 4** 

```{r}

step2 <- function(df){ # df is sig df with appended features  
  print("Performing Step 2...")
  num.pairs <- dim(df)[1]
  bonfer <- 0.05 / num.pairs   
  print("Bonferroni threshold: p <= " %&% bonfer)  
  sub.df <- filter(df,pred.perf.pvalue<=bonfer)
  print("Number of associations before filtering: " %&% dim(df)[1])
  print("Number of associations after filtering: " %&% dim(sub.df)[1])
  return(sub.df)
}

step3 <- function(df,ph3.thresh=0.5){ # df is output from step2 
  print("Performing Step 3...")
  sub.df <- filter(df,ph3>=ph3.thresh)
  print("Number of associations before filtering: " %&% dim(df)[1])
  print("Number of associations after filtering: " %&% dim(sub.df)[1])
  return(sub.df)
}

step4 <- function(df,ph4.thresh=0.5){ # df is output from step 3 
  print("Performing Step 4...")
  sub.df <- filter(df,ph4>=ph4.thresh)
  print("Number of associations before filtering: " %&% dim(df)[1])
  print("Number of associations after filtering: " %&% dim(sub.df)[1])
  return(sub.df)
}


```


# Peform analysis

**Build results data frames** 

```{r}

# GTEx only 
p.mat.gtex <- build_p.mat.gtex()
gtex.sig.df <- build.sig.df(p.mat.gtex) # "Bonferroni threshold: p <= 2.56163288726766e-07"
gtex.sig.df <- append_to_sigdf(gtex.sig.df)
gtex.sig.df <- append_tx_info(gtex.sig.df)

# DGN only 
p.mat.dgn <- build_p.mat.dgn()
dgn.sig.df <- build.sig.df(p.mat.dgn) # "Bonferroni threshold: p <= 4.33839479392625e-06"
dgn.sig.df <- append_to_sigdf(dgn.sig.df)
dgn.sig.df <- append_tx_info(dgn.sig.df)

# GTEx and DGN models 
p.mat.full <- build_p.mat.full()
full.sig.df <- build.sig.df(p.mat.full) # "Bonferroni threshold: p <= 2.41881255653974e-07"
full.sig.df <- append_to_sigdf(full.sig.df)
full.sig.df <- append_tx_info(full.sig.df)

write.table(x=gtex.sig.df,file=save.dir%&%"gtex-only.results.txt",
            sep="\t",quote=FALSE,row.names=FALSE)
write.table(x=dgn.sig.df,file=save.dir%&%"dgn-only.results.txt",
            sep="\t",quote=FALSE,row.names=FALSE)
write.table(x=full.sig.df,file=save.dir%&%"full.results.txt",
            sep="\t",quote=FALSE,row.names=FALSE)

```


## GTEx only 

```{r}

gtex.step2 <- step2(gtex.sig.df)
gtex.step3 <- step3(gtex.step2)
gtex.step4 <- step4(gtex.step3,ph4.thresh=0.5)

```


## DGN only 

```{r}

dgn.step2 <- step2(dgn.sig.df)
dgn.step3 <- step3(dgn.step2)
dgn.step4 <- step4(dgn.step3,ph4.thresh=0.5)

```

## GTEx and DGN 


```{r}

full.step2 <- step2(full.sig.df)
full.step3 <- step3(full.step2)
dgn.step4 <- step4(full.step3,ph4.thresh=0.5)

```
