---
title: "Locus based analysis"
author: "Jason Torres"
date: "August 13, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup 

```{r}

"%&%" <- function(a,b) paste0(a,b)
library("dplyr")
library("data.table")

#serv.dir <- "/group/im-lab/nas40t2/jason/" # server
serv.dir <- "/Users/jtorres/FUSE4/"
root.dir <- serv.dir %&% "projects/MetaXcan/"
rds.dir <- root.dir %&% "/MetaXcanT2D/RDS/"
save.dir <- root.dir %&% "MetaXcanT2D/analyzed_result_files/"
gwas.df <- readRDS(rds.dir %&% "sig.gwas.annot.df.RDS") # dim(gwas.df) is 566 15 

```

Read in results data frames 

```{r}
res.dir <- root.dir %&% "results/tables/GTExV6p/" # DIAGRAM results with COLOC output appended 
meta.dir <- root.dir %&% "results/meta-analyses/DIAGRAM-GERA_ImpG_0.80_gtexV6p/tables/"

pfile <- res.dir %&% "DIAGRAM_ImpG_0.8_gtexV6p.0.5.pvalue.table.csv.gz"
p.df <- fread("cat " %&% pfile %&% " | zmore")
gene.vec <- p.df$Gene

ens.df <- fread("cat " %&% res.dir %&% 
                "DIAGRAM_ImpG_0.8_gtexV6p.0.5.ensid.table.csv.gz" %&% " | zmore")
z.df <- fread("cat " %&% res.dir %&% 
                "DIAGRAM_ImpG_0.8_gtexV6p.0.5.zscore.table.csv.gz" %&% " | zmore")
r2.df <- fread("cat " %&% res.dir %&% 
                 "DIAGRAM_ImpG_0.8_gtexV6p.0.5.r2.table.csv.gz" %&% " | zmore")
ppp.df <- fread("cat " %&% res.dir %&% 
                 "DIAGRAM_ImpG_0.8_gtexV6p.0.5.ppp.table.csv.gz" %&% " | zmore")
nsm.df <- fread("cat " %&% res.dir %&% 
                  "DIAGRAM_ImpG_0.8_gtexV6p.0.5.nsnpsmod.table.csv.gz" %&% " | zmore")
ns.df <- fread("cat " %&% res.dir %&% 
                 "DIAGRAM_ImpG_0.8_gtexV6p.0.5.nsnps.table.csv.gz" %&% " | zmore")
ph3.df <- fread("cat " %&% res.dir %&% 
                  "DIAGRAM_ImpG_0.8_gtexV6p.0.5.ph3.table.csv.gz" %&% " | zmore")
ph4.df <- fread("cat " %&% res.dir %&% 
                  "DIAGRAM_ImpG_0.8_gtexV6p.0.5.ph4.table.csv.gz" %&% " | zmore")
gera.z.df <- fread("cat " %&% res.dir %&% 
                     "GERA_ImpG_0.8_gtexV6p.0.5.zscore.table.csv.gz" %&% " | zmore")
gera.p.df <- fread("cat " %&% res.dir %&% 
                     "GERA_ImpG_0.8_gtexV6p.0.5.pvalue.table.csv.gz" %&% " | zmore")
gera.ns.df <- fread("cat " %&% res.dir %&% 
                      "GERA_ImpG_0.8_gtexV6p.0.5.nsnps.table.csv.gz" %&% " | zmore")
meta.z.df <- fread("cat " %&% meta.dir %&% 
                      "DIAGRAM-GERA_ImpG_0.80_gtexV6p.0.5.zscore.table.csv.gz" %&% " | zmore")
meta.p.df <- fread("cat " %&% meta.dir %&% 
                      "DIAGRAM-GERA_ImpG_0.80_gtexV6p.0.5.pvalue.table.csv.gz" %&% " | zmore")
meta.moresig.df <- fread("cat " %&% meta.dir %&% 
                      "DIAGRAM-GERA_ImpG_0.80_gtexV6p.0.5.moresig.table.csv.gz" %&% " | zmore")

```


# Partition Loci (~ 1 Mb windows) 

```{r}

library("GenomicRanges")

build_locus_df <- function(){
  gwas.df$win.start <- gwas.df$POS - 1000000
  gwas.df$win.end <- gwas.df$POS + 1000000
  gwas.df$chrom <- as.integer(gsub("chr","",gwas.df$CHR)) # 2 SNPs are missing 
  gwas.df <- arrange(gwas.df,chrom,POS)
  gwas.gr <- GRanges(seqnames=na.omit(gwas.df)$CHR,IRanges(start=na.omit(gwas.df)$win.start,
                     end=na.omit(gwas.df)$win.end))  
  ind.gr <- reduce(gwas.gr)
  chromo.vec <- as.character(seqnames(ind.gr))
  locus <- c()
  id.refGene <- c()
  id.ensembl <- c()
  id.ucsc <- c() 
  pb <- txtProgressBar(min=0,max=length(ind.gr),style=3)
  for (i in 1:length(ind.gr)){
    setTxtProgressBar(pb,i)
    rang <- ind.gr[i]
    #print(seqnames(rang)[])
    chromo <- chromo.vec[i]
    start <- start(ranges(ind.gr[i]))
    end <- end(ranges(ind.gr[i]))
    sub.df <- filter(gwas.df,CHR==chromo,POS>=start,POS<=end)
    locus <- append(locus,i)
    id.refGene <- append(id.refGene,paste0(unique(sub.df$refGene),collapse=","))
    id.ensembl <- append(id.ensembl,paste0(unique(sub.df$Ensembl),collapse=","))
    id.ucsc <- append(id.ucsc,paste0(unique(sub.df$hgnc),collapse=","))    
  }
  out.df <- data.frame("Locus"=locus,"CHR"=chromo.vec,
                       "Start"=start(ranges(ind.gr)),"End"=end(ranges(ind.gr)),
                       "refGene"=id.refGene,"Ensembl"=id.ensembl,"UCSC"=id.ucsc,
                       stringsAsFactors = FALSE)
  return(out.df)
}

locus.df <- build_locus_df()
write.table(x=locus.df,file=save.dir%&%"locus-table.txt",
            sep="\t",quote=FALSE,row.names=FALSE)
saveRDS(object=locus.df,file=rds.dir%&%"locus.df.RDS")

```


# Build Ensemble Ref Data frame 

```{r}

library("GenomicRanges")
library("GenomicFeatures")
library("org.Hs.eg.db")
library("annotate")
library("Homo.sapiens")
#install.packages("devtools")
#devtools::install_github("stephenturner/annotables")
library("annotables")


get_tx_gr <- function(ensid){
  tx.chrom <- tryCatch({(select(Homo.sapiens,key=ensid,keytype="ENSEMBL",
                                columns="TXCHROM"))$TXCHROM}, 
                       warning=function(war){return(NA)},
                       error=function(err){return(NA)})   
  tx.start <- tryCatch({(select(Homo.sapiens,key=ensid,keytype="ENSEMBL",
                                columns="TXSTART"))$TXSTART}, 
                       warning=function(war){return(NA)},
                       error=function(err){return(NA)})                          
  tx.end <- tryCatch({(select(Homo.sapiens,key=ensid,keytype="ENSEMBL",
                                columns="TXEND"))$TXEND}, 
                       warning=function(war){return(NA)},
                       error=function(err){return(NA)}) 
  tx.strand <- tryCatch({(select(Homo.sapiens,key=ensid,keytype="ENSEMBL",
                                columns="TXSTRAND"))$TXSTRAND}, 
                       warning=function(war){return(NA)},
                       error=function(err){return(NA)})   
  #print(tx.chrom %&% " : " %&% tx.start %&% " : " %&% tx.end %&% " : " %&% tx.strand)
  tx.gr <- tryCatch({(GRanges(tx.chrom,IRanges(tx.start, tx.end),tx.strand))}, 
                       warning=function(war){return(NA)},
                       error=function(err){return(NA)})  
  return(tx.gr)
}

build_ensid_ref <- function(){
  Gene <- ens.df$Gene
  Ensemble <- c()
  CHR <- c()
  Start <- c()
  End <- c() 
  pb <- txtProgressBar(min=0,max=length(Gene),style=3)
  for (i in 1:length(Gene)){
    setTxtProgressBar(pb,i)
    sub.df <- filter(grch37,symbol==Gene[i])
    if (dim(sub.df)[1]>0){
      #if(dim(sub.df)[1]>1){
      #  print(sub.df)
      #}
      ensid <- sub.df[1,]$ensgene 
      chromo <- sub.df[1,]$chr
      tx.start <- sub.df[1,]$start
      tx.end <- sub.df[1,]$end      
    } else{
      ensid <- NA 
      chromo <- NA
      tx.start <- NA
      tx.end <- NA    
    }
    Ensemble <- append(Ensemble,ensid)
    CHR <- append(CHR,chromo)
    Start <- append(Start,tx.start)
    End <- append(End,tx.end)
  }
  out.df <- data.frame("Gene"=Gene,"Ensemble"=Ensemble,
                       "CHR"=CHR,"TX.Start"=Start,"TX.End"=End,
                       stringsAsFactors=FALSE)
  out.df$CHR <- as.integer(out.df$CHR)
  return(out.df)
}

ens.ref.df <- build_ensid_ref()

write.table(x=ens.ref.df,file=save.dir%&%"Ensemble-reference.txt",
            sep="\t",quote=FALSE,row.names=FALSE)
saveRDS(object=ens.ref.df,file=rds.dir%&%"ens.ref.df.RDS")

missing.info.genes <- ens.ref.df[is.na(ens.ref.df$Ensemble),]$Gene # possibly outdated gene symbols (i.e. KLHDC5)

#vec <- c()
#for (i in 1:dim(ens.df)[1]){
#  v <- unique(na.omit(as.character(as.data.frame(ens.df)[i,2:46])))
#  vec <- append(vec,v)
#  
#}


```



# Step 1 functions 


```{r}

  #k <- filter(ens.ref.df,CHR==chrom,TX.Start>=win.start,TX.End<=win.end)
  #print(c(locus.df$refGene[i],dim(k)[1]))

build_p.mat.gtex <- function(chrom,win.start,win.end){
  ensgenes <- strsplit(filter(locus.df,CHR==chrom,Start>=win.start,End<=win.end)$Ensembl,",")[[1]]
  keep.genes <- filter(grch37,ensgene%in%ensgenes)$symbol
  keep.genes <- append(keep.genes,filter(ens.ref.df,CHR==(as.integer(gsub("chr","",chrom)))
                       ,TX.Start>=win.start,TX.End<=win.end)$Gene)
  psub.df <- filter(p.df,Gene %in% keep.genes)
  gene.index <- (1:length(colnames(psub.df)))[grepl("Gene",colnames(psub.df))]
  dgn.index <- (1:length(colnames(psub.df)))[grepl("DGN",colnames(psub.df))]
  keep.vec <- (1:length(colnames(psub.df)))[!((1:length(colnames(psub.df))) %in% c(gene.index,dgn.index))]
  p.mat <- as.data.frame(as.data.frame(psub.df)[,keep.vec])
  row.names(p.mat) <- psub.df$Gene
  return(p.mat)
}

build.sig.df <- function(p.mat){

  not.missing <- sapply(1:dim(p.mat)[2],function(i){
    count <- length(na.omit(p.mat[,i]))
    return(count)
  })
  names(not.missing) <- colnames(p.mat)
  num.pairs <- sum(not.missing)

  bonfer <- 0.05 / num.pairs 
  #print("Bonferroni threshold: p <= " %&% bonfer)
  sig.df <- c() 
  for (i in 1:dim(p.mat)[2]){
    vec <- p.mat[,i]; vec[is.na(vec)] <- 1
    sig <- vec <= bonfer
    #tiss.vec <- rep(gsub("TW_","",colnames(p.mat)[i]),sum(sig))
    tiss.vec <- rep(colnames(p.mat)[i],sum(sig))
    p.vec <- vec[sig]
    g.vec <- row.names(p.mat)[sig]#gene.vec[sig]
    df <- data.frame(tissue=tiss.vec,gene=g.vec,pvalue=p.vec,
                     stringsAsFactors = FALSE)
    sig.df <- rbind(sig.df,df)
  }
  return(sig.df)
}

get_vals <- function(sig.df, ref.df,is.string=FALSE){
  pb <- txtProgressBar(min=0,max=dim(sig.df)[1],initial=0,style=3)
  vec <- sapply(1:dim(sig.df)[1],function(i){
    setTxtProgressBar(pb,i)
    tiss <- sig.df[i,1]; gene <- sig.df[i,2]
    
    sub.df <- dplyr::select(ref.df,one_of("Gene",tiss)) %>% filter(Gene==gene) %>% 
      dplyr::select(one_of(tiss))
    if (is.string==TRUE){
      val <- as.character(sub.df)
    } else{
      val <- as.numeric(sub.df)
    }
    return(val)
  })
  return(vec)
}

append_to_sigdf <- function(sig.df){
  ensid <- get_vals(sig.df, ens.df, is.string=TRUE)
  zscore <- get_vals(sig.df, z.df)
  pred.perf.r2 <- get_vals(sig.df, r2.df) 
  pred.perf.pvalue <- get_vals(sig.df, ppp.df) 
  n.snps.model <- get_vals(sig.df, nsm.df) 
  n.snps <- get_vals(sig.df, ns.df)
  temp.df <- sig.df
  temp.df$tissue <- gsub("TW_Whole_Blood_DGN","TW_Whole_Blood",temp.df$tissue)
  ph3 <- get_vals(temp.df, ph3.df)
  ph4 <- get_vals(temp.df, ph4.df)
  #if (is.list(ph3)==TRUE & is.vector(ph3)==TRUE){
  #  ph3 <- rep(NA,length(ensid))
  #  ph4 <- rep(NA,length(ensid))    
  #}
  gera.zscore <- get_vals(sig.df, gera.z.df)
  gera.pvalue <- get_vals(sig.df, gera.p.df)
  gera.nsnps <- get_vals(sig.df, gera.ns.df)
  meta.zscore <- get_vals(sig.df, meta.z.df)
  meta.pvalue <- get_vals(sig.df, meta.p.df)
  meta.more.sig <- get_vals(sig.df, meta.moresig.df)
  out.df <- cbind(ensid,sig.df,zscore,pred.perf.r2,pred.perf.pvalue,n.snps.model,n.snps,
                  ph3,ph4,gera.zscore,gera.pvalue,gera.nsnps,
                  meta.zscore,meta.pvalue,meta.more.sig)
  out.df$tissue <- gsub("TW_","",out.df$tissue)
  out.df$ensid <- as.character(out.df$ensid)
  return(out.df)
}

append_tx_info <- function(sig.df){ # output from append_to_sigdf
  chrom <- c()
  tx.start <- c()
  tx.end <- c()
  tx.strand <- c()
  pb <- txtProgressBar(min=0,max=length(sig.df$ensid),initial=0,style=3)
  for (i in 1:length(sig.df$ensid)){
    setTxtProgressBar(pb,i)
    ensid <- sig.df$ensid[i]
    #print(ensid)
    tx.gr <- suppressMessages(get_tx_gr(ensid))
    if (is.na(tx.gr)){
      chrom <- append(chrom,NA)
      tx.start <- append(tx.start,NA)
      tx.end <- append(tx.end,NA)
      tx.strand <- append(tx.strand,NA)      
    } else{
      c <- as.character(seqnames(tx.gr))[1]; chrom <- append(chrom,c)
      ts <- paste0(unique(start(tx.gr)),collapse=","); tx.start <- append(tx.start,ts)
      te <- paste0(unique(end(tx.gr)),collapse=","); tx.end <- append(tx.end,te)
      tst <- paste0(unique(strand(tx.gr)),collapse=","); tx.strand <- append(tx.strand,tst)      
    }
  }
  out.df <- cbind(chrom,tx.start,tx.end,tx.strand,sig.df)
  out.df$chrom <- as.character(out.df$chrom)
  out.df$tx.start <- as.character(out.df$tx.start)
  out.df$tx.end <- as.character(out.df$tx.end)
  out.df$tx.strand <- as.character(out.df$tx.strand)
  return(out.df)
}


build_siggene_results <- function(){
  out.df <- c()
  #pb <- txtProgressBar(min=0,max=dim(locus.df)[1],style = 3)
  for (i in 1:dim(locus.df)[1]){
    #setTxtProgressBar(pb,i)
    print(i)
    chrom <- locus.df$CHR[i] #as.integer(gsub("chr","",locus.df$CHR[i]))
    win.start <- locus.df$Start[i]
    win.end <- locus.df$End[i]
    p.mat <- build_p.mat.gtex(chrom,win.start,win.end)
    sig.df <- build.sig.df(p.mat)
    if (dim(sig.df)[1] > 0){
      #print(locus.df$UCSC[i])
      sig.df <- append_to_sigdf(sig.df)
      sig.df <- append_tx_info(sig.df) 
      sig.df <- cbind(Locus=rep(i,dim(sig.df)[1]),sig.df)
      out.df <- rbind(out.df,sig.df)
    }
  }  
  # Fill in missing information with ensgene lookup in annotable 
  pb <- txtProgressBar(min=0,max=dim(out.df)[1],style=3)
  for (i in 1:dim(out.df)[1]){
    setTxtProgressBar(pb,i)
    eid <- out.df$ensid[i]; chromo <- out.df$chrom[i]
    if (is.na(chromo)==TRUE){
      sub.df <- filter(grch37,ensgene==eid)
      if (dim(sub.df)[1]>0){
        out.df$chrom[i] <- "chr"%&%sub.df$chr[1]
        out.df$tx.start[i] <- sub.df$start[1] 
        out.df$tx.end[i] <- sub.df$end[1]
      }
    }
  }
  return(out.df)

}



```


# Step 2-4 functions 

```{r}

step2 <- function(df){ # df is sig df with appended features  
  print("Performing Step 2...")
  num.pairs <- dim(df)[1]
  bonfer <- 0.05 / num.pairs   
  print("Bonferroni threshold: p <= " %&% bonfer)  
  sub.df <- filter(df,pred.perf.pvalue<=bonfer)
  print("Number of associations before filtering: " %&% dim(df)[1])
  print("Number of associations after filtering: " %&% dim(sub.df)[1])
  return(sub.df)
}

step3 <- function(df,ph3.thresh=0.5){ # df is output from step2 
  print("Performing Step 3...")
  sub.df <- filter(df,ph3>=ph3.thresh)
  print("Number of associations before filtering: " %&% dim(df)[1])
  print("Number of associations after filtering: " %&% dim(sub.df)[1])
  return(sub.df)
}

step4 <- function(df,ph4.thresh=0.5){ # df is output from step 3 
  print("Performing Step 4...")
  sub.df <- filter(df,ph4>=ph4.thresh)
  print("Number of associations before filtering: " %&% dim(df)[1])
  print("Number of associations after filtering: " %&% dim(sub.df)[1])
  return(sub.df)
}


```


# Perform analysis 

```{r}

gtex.locsig.df <- build_siggene_results()

write.table(x=gtex.locsig.df,file=save.dir%&%"gtex-only.locus-results.txt",
            sep="\t",quote=FALSE,row.names=FALSE)

saveRDS(object=gtex.locsig.df,file=rds.dir%&%"gtex.locsig.df.RDS")

step2.df <- step2(gtex.locsig.df)

#[1] "Performing Step 2..."
#[1] "Bonferroni threshold: p <= 0.000271739130434783"
#[1] "Number of associations before filtering: 184"
#[1] "Number of associations after filtering: 109"

step3.df <- step3(step2.df)

#[1] "Performing Step 3..."
#[1] "Number of associations before filtering: 109"
#[1] "Number of associations after filtering: 22"

step4.df <- step4(step3.df,ph4.thresh=0.01)

#[1] "Performing Step 4..."
#[1] "Number of associations before filtering: 22"
#[1] "Number of associations after filtering: 0"

```


