---
title: "SignficantGenesProfile"
author: "Jason Torres"
date: "August 4, 2016"
output:
  pdf_document:
    toc: yes
  html_document:
    fig_caption: yes
    highlight: textmate
    number_sections: yes
    theme: journal
    toc: yes
---


# Assemble result and reference data frames

Create reference data frames of signficant MetaXcan genes from DIAGRAM3 analysis 

```{r setup, echo=FALSE}
library("data.table")
library("dplyr")
library("ggplot2")
library("stargazer")
"%&%" <- function(a,b) paste0(a,b) 

fig.dir <- "figures/"
dir.create(fig.dir,showWarnings=FALSE)
rds.dir <- "intermediate_files/diag/" 
dir.create(rds.dir,showWarnings=FALSE)
pdf.dir <- "figures/" 
dir.create(pdf.dir,showWarnings=FALSE)
tab.dir <- "intermediate_files/diag/"
dir.create(tab.dir,showWarnings=FALSE)
table.dir<-"input_files/tables/"

meta.name <- "DIAGRAM_ImpG_0.8_gtexV6p" 

```

```{r load_hidden, echo=FALSE, results="hide", warning=FALSE}
suppressPackageStartupMessages({
  library("data.table")
  library("dplyr")
  library("ggplot2")
})
```


```{r table_function, echo=FALSE}

make.latex.table <- function(df,tabname,fname){
  outpath <- tab.dir %&% fname %&% ".tex"
  code <- stargazer(df,type="latex",title=tabname,summary=FALSE,
                    out=outpath,rownames=FALSE)
}
```

Read in MetaXcan all z-scores for DIAGRAM3 analysis - use table file 
Create z-score df and pval df for each alpha 

```{r z_pval_dfs, echo=FALSE}

makeZdf <- function(alpha){
  dir <- table.dir %&% meta.name %&% "."
  file <- dir %&% alpha %&% ".zscore" %&% ".table.csv.gz"
  arg <- "cat " %&% file %&% " | zmore"
  df <- as.data.frame(fread(arg))
  return(df)
}
makePdf <- function(alpha){
  dir <- table.dir %&% meta.name %&% "."
  file <- dir %&% alpha %&% ".pvalue" %&% ".table.csv.gz"
  arg <- "cat " %&% file %&% " | zmore"
  df <- as.data.frame(fread(arg))
  return(df)
}

z.EN.df <- makeZdf(0.5); p.EN.df <- makePdf(0.5)
```

Create q-value dfs 

```{r qvalues,echo=FALSE}
library("qvalue")
makeQdf <- function(p.df){
  Gene <- p.df[,1]
  q.df <- c()
  for (col in 2:dim(p.df)[2]){
    pvec <- p.df[,col]
    qvec <- qvalue(pvec)$qvalues
    q.df <- cbind(q.df,qvec)
  }
  q.df <- as.data.frame(q.df)
  q.df <- cbind(Gene,q.df)
  names(q.df) <- names(p.df)
  q.df$Gene <- as.character(q.df$Gene)
  return(q.df)
}
q.EN.df <- makeQdf(p.EN.df)
```



```{r checkpoint1,echo=FALSE}
save(z.EN.df,p.EN.df,q.EN.df,file=paste0(rds.dir,"metaXcan.results.Rdata"))
load(paste0(rds.dir,"metaXcan.results.Rdata"))
```

Make data frames with the number of predictor SNPs used to predictor gene expression for each tested in MetaXcan association analysis 

```{r snp_num,echo=FALSE}
makeNSNPSMODdf <- function(alpha){
  dir <- table.dir %&% meta.name %&% "."
  file <- dir %&% alpha %&% ".nsnpsmod" %&% ".table.csv.gz"
  arg <- "cat " %&% file %&% " | zmore"
  df <- as.data.frame(fread(arg))
  return(df)
}
s.EN.df <- makeNSNPSMODdf(0.5)
saveRDS(s.EN.df,paste0(rds.dir,"s.EN.df.RDS"))
s.EN.df <- readRDS(paste0(rds.dir,"s.EN.df.RDS"))

makeNSNPSdf <- function(alpha){
  dir <- table.dir %&% meta.name %&% "."
  file <- dir %&% alpha %&% ".nsnps" %&% ".table.csv.gz"
  arg <- "cat " %&% file %&% " | zmore"
  df <- as.data.frame(fread(arg))
  return(df)
}
n.EN.df <- makeNSNPSdf(0.5)
saveRDS(n.EN.df,paste0(rds.dir,"n.EN.df.RDS"))
n.EN.df <- readRDS(paste0(rds.dir,"n.EN.df.RDS"))
```

Create Reference with ENSEMBLE Ids 

```{r ensid_ref, echo=FALSE}

server.dir <- "input_files/"
gencode.path <- server.dir %&%  "GENCODE.Genes.V19.Basic.annotated.txt.gz"
gencode.ref.df <- fread("cat " %&% gencode.path %&% " | zmore")

gene2ensid <- function(mygene){
  genename <- filter(gencode.ref.df,hg19.wgEncodeGencodeAttrsV19.geneName==mygene)$hg19.wgEncodeGencodeAttrsV19.geneId[1]
  return(genename)
}
```

Create Gene Annotation reference Build37 Genecode V18

```{r build37_ref, echo=FALSE}

make_ggd_ref <- function(){
  print("Gene Annotatino Table from Steve Turner")
  print("Getting Genetics Done")
  #http://www.gettinggeneticsdone.com/2015/11/annotables-convert-gene-ids.html
  library("devtools")
  #devtools::install_github("stephenturner/annotables")
  library(annotables)
  df <- grch37
  saveRDS(df,paste0(rds.dir,"ggd.ref.df.RDS"))
}
make_ggd_ref()
ggd.ref.df <- readRDS(paste0(rds.dir,"ggd.ref.df.RDS"))
```

Build Cross Validation $r^2$ data frame 

```{r cross_val_r2, echo=FALSE}
makeR2df <- function(alpha){
  dir <- table.dir %&% meta.name %&% "."
  file <- dir %&% alpha %&% ".r2" %&% ".table.csv.gz"
  arg <- "cat " %&% file %&% " | zmore"
  df <- as.data.frame(fread(arg))
  return(df)
}

r.EN.df <- makeR2df(0.5)
saveRDS(r.EN.df,paste0(rds.dir,"r.EN.df.RDS"))
r.EN.df <- readRDS(paste0(rds.dir,"r.EN.df.RDS"))
```

# Build MetaXcan data frame for all genes 

Building core of Gene, Model (i.e. Tissue), and Alpha 

```{r build_core, echo=FALSE}

build_core <- function(alpha){
  gvec.mod <- z.EN.df$Gene
  tvec.mod <- names(z.EN.df)[2:dim(z.EN.df)[2]]
  core.df <- c()
  for (t in tvec.mod){
    tvec <- rep(t,length(gvec.mod))
    avec <- rep(alpha,length(gvec.mod))
    stack <- cbind(gvec.mod,tvec,avec)
    core.df <- rbind(core.df,stack)
  }
  core.df <- as.data.frame(core.df)
  names(core.df) <- c("gene","model","alpha")
  core.df$gene <- as.character(core.df$gene)
  core.df$model <- as.character(core.df$model)
  core.df$alpha <- as.numeric(as.character(core.df$alpha))
  saveRDS(core.df,paste0(rds.dir,"core.df.RDS"))
  print("Saving RDS file")
}
build_core(0.5)
core.df <- readRDS(paste0(rds.dir,"core.df.RDS"))
```

# Build MetaXcan data frame for Genome-Significant Genes (Bonferroni and FDR05 only)

Identify significant gene-model pairs 

```{r significant_genes, echo=FALSE}

# Note KLHDC5 is an outdated name for KLHL42
# (http://www.genecards.org/cgi-bin/carddisp.pl?gene=KLHL42)

make_sig_gene.df <- function(myalpha){
  sig.gene.df <- c()
  d1 <- readRDS(paste0(rds.dir,"elasticNet_FDR05.RDS")) %>% select(Gene,Type)
  d1$Gene <- as.character(d1$Gene); d1$Type <- as.character(d1$Type)
  Alpha <- rep(myalpha,dim(d1)[1])
  d1 <- cbind(d1,Alpha)
  df <- d1
  # Iterate over significant genes associations 
  for (i in 1:dim(df)[1]){
    gene <- df[i,1]; type <- df[i,2]; alpha <- df[i,3]
    print(c(gene,i))
    if (gene == "KLHDC5"){
      gene = "KLHL42"
    }
    if (alpha==0.5) {
      z.df <- filter(z.EN.df,Gene==gene)
      p.df <- filter(p.EN.df,Gene==gene)
      q.df <- filter(q.EN.df,Gene==gene)
      s.df <- filter(s.EN.df,Gene==gene)
      r.df <- filter(r.EN.df,Gene==gene)
      n.df <- filter(n.EN.df,Gene==gene)
    } 
      gene.df <- rbind(z.df,p.df,q.df,s.df,n.df,r.df)
      gene.df <- gene.df[ , colSums(is.na(gene.df)) == 0]
      keep.vec <- c(1)
      for (c in 2:dim(gene.df)[2]){if (gene.df[3,c] < 0.05) keep.vec <- append(keep.vec,c)}
      gene.df <- gene.df[,keep.vec]
      tvec <- names(gene.df)
      for (c in 2:dim(gene.df)[2]){
        model <- tvec[c]; z.score <- gene.df[1,c]
        p.value <- gene.df[2,c]; q.value <- gene.df[3,c]
        num.model.snps <- as.integer(gene.df[4,c])
        num.used.snps <- as.integer(gene.df[5,c])
        cv.r2 <- gene.df[6,c]
        # Updated approach to evaluate bonferonni significance
        if (alpha==1) {
          mytemp.df <- p.L.df
        } else if (alpha==0.5){
          mytemp.df <- p.EN.df
        } else{
          stop("Need a valid alpha value")
        }
        mytemp.df <- select(mytemp.df,contains(model))
        refnum <- length(na.omit(as.numeric(as.matrix(mytemp.df))))
        bonfer <- 0.05 / refnum
        bonferonni <- (p.value < bonfer)
        ensid.vers <- gene2ensid(gene)
        ensid <- gsub("\\..+","",ensid.vers)
        if (length(c(ensid,model,z.score,p.value,q.value,num.model.snps,bonferonni,ensid.vers)) == 8) {
          ggd.df <- filter(ggd.ref.df,ensgene==ensid) %>% select(chr,start,end,strand,biotype,description)
          chrom <- as.integer(ggd.df$chr)
          start <- ggd.df$start; end <- ggd.df$end
          strand <- ggd.df$strand; biotype <- ggd.df$biotype
          description <- ggd.df$description
        } else if (dim(filter(ggd.ref.df,symbol==gene))[1]==0){
            print(gene)
            ensid <- gene; ensid.vers <- gene
            chrom <- NA
            start <- NA; end <- NA
            strand <- NA; biotype <- NA
            description <- NA
        } else {
          print("There is an unknown problem, please inspect")
          break
        }
          stack.df <- cbind(ensid,ensid.vers,gene,type,biotype,chrom,start,end,model,
                            alpha,z.score,p.value,q.value,num.model.snps,num.used.snps,cv.r2,
                            bonferonni,strand,description)
          if (dim(stack.df)[2] != 19){
            ggd.df <- filter(ggd.ref.df,symbol==gene) %>% select(chr,start,end,strand,biotype,description)
            chrom <- as.integer(ggd.df$chr)
            start <- ggd.df$start; end <- ggd.df$end
            strand <- ggd.df$strand; biotype <- ggd.df$biotype
            description <- ggd.df$description
            stack.df <- cbind(ensid,ensid.vers,gene,type,biotype,chrom,start,end,model,
                              alpha,z.score,p.value,q.value,num.model.snps,num.used.snps,cv.r2,
                              bonferonni,strand,description)
          }
          sig.gene.df <- as.data.frame(rbind(sig.gene.df,stack.df))

      }
  }
  sig.gene.df$ensid <- as.character(sig.gene.df$ensid)
  sig.gene.df$ensid.vers <- as.character(sig.gene.df$ensid.vers)
  sig.gene.df$gene <- as.character(sig.gene.df$gene)
  sig.gene.df$type <- as.character(sig.gene.df$type)
  sig.gene.df$biotype <- as.character(sig.gene.df$biotype)
  sig.gene.df$chrom <- as.integer(as.character(sig.gene.df$chrom))
  sig.gene.df$start <- as.integer(as.character(sig.gene.df$start))
  sig.gene.df$end <- as.integer(as.character(sig.gene.df$end))
  sig.gene.df$model <- as.character(sig.gene.df$model)
  sig.gene.df$description <- as.character(sig.gene.df$description)
  sig.gene.df$alpha <- as.numeric(as.character(sig.gene.df$alpha))
  sig.gene.df$z.score <- as.numeric(as.character(sig.gene.df$z.score))
  sig.gene.df$p.value <- as.numeric(as.character(sig.gene.df$p.value))
  sig.gene.df$q.value <- as.numeric(as.character(sig.gene.df$q.value))
  sig.gene.df$num.model.snps <- as.integer(as.character(sig.gene.df$num.model.snps))
  sig.gene.df$num.used.snps <- as.integer(as.character(sig.gene.df$num.used.snps))
  sig.gene.df$cv.r2 <- as.numeric(as.character(sig.gene.df$cv.r2))
  sig.gene.df$bonferonni <- as.logical(sig.gene.df$bonferonni)
  sig.gene.df$strand <- as.integer(as.character(sig.gene.df$strand))
  sig.gene.df <- mutate(sig.gene.df,utilization=num.used.snps/num.model.snps)
  sig.gene.df <- sig.gene.df[!duplicated(sig.gene.df),]
  saveRDS(sig.gene.df,paste0(rds.dir,"sig.gene.df.RDS"))
  print("Saved RDS file")
}
make_sig_gene.df(0.5)
sig.gene.df <- readRDS(paste0(rds.dir,"sig.gene.df.RDS"))
```

# Profile Significant Genes 

Here, I will explore the top genes from the MetaXcan analyses and consider the affect of applying different restrictions, namely Lasso models for **CrossTissue** and **Tissue-wide** predictors

Summarize genes function to allow comparison of gene sets 


```{r profile1, echo=FALSE,fig.height=8,fig.width=6}

sub_sig <- function(alphalimit=NULL,TwCtOnly=FALSE,bonfer=FALSE){
  if (TwCtOnly==TRUE){
    df <- filter(sig.gene.df,!grepl("TS_",model)) %>%  filter(!grepl("Organ_",model))
  } else {df <- sig.gene.df}
  if (is.null(alphalimit)==FALSE) df <- filter(df,alpha==alphalimit)
  if (bonfer==TRUE){
    if (is.null(alphalimit)==TRUE){
      temp2.df <- readRDS(rds.dir %&% "elasticNet_Bonferonni.RDS")
      keepgenes2 <- unique(as.character(temp2.df$Gene))
      keepgenes <- append(keepgenes2)
      df <- filter(df, gene %in% keepgenes)
    } else if (alphalimit==0.5){
      temp.df <- readRDS(rds.dir %&% "elasticNet_Bonferonni.RDS")
      keepgenes <- unique(as.character(temp.df$Gene))
      df <- filter(df, gene %in% keepgenes)
    } else if (alphalimit==1){
      print("Option for alpha == 1 is not available")
    } else {
      print("Please enter a valid Alpha for Bonferonni option")
    }
  }       
  return(df)
}


get_sum_table <- function(sig.gene.dataframe){
  sum.df <- dplyr::select(sig.gene.dataframe,gene,type)
  sum.df <- sum.df[!duplicated(sum.df$gene),]
  return(table(sum.df$type))
}

reformat_sub_sig <- function(df,designame){
  df <- dplyr::select(df,gene,type)
  c <- rep(designame,dim(df)[1])
  df <- cbind(df,c)
  names(df) <- c("Gene","Type","Criteria")
  df <- df[!duplicated(df$Gene),]
  df$Gene <- as.factor(df$Gene); df$Type <- as.factor(df$Type)
  return(df)
}

prof <- function(myalpha){
  df <- reformat_sub_sig(sub_sig(TwCtOnly=TRUE,alphalimit=NULL),"TW.a"%&%myalpha)
  df$Type <- factor(df$Type,levels=c("T2D.Gene","Known.Region","Unknown"))
  plt <- ggplot(data=df,aes(x=Criteria)) +
    geom_bar(aes(fill=Type),color="black") +
    scale_fill_brewer() + ylab("Count") + 
    xlab("Gene Set") + 
    ggtitle("FDR 5%-significant Gene Sets") +
    scale_y_continuous(breaks=seq(0,230,10)) +
    theme_bw() + 
    theme(plot.title = element_text(face="bold",size=20),
          axis.title.x = element_text(face="bold"),
          axis.title.y = element_text(face="bold"),
          axis.text.x = element_text(angle=-45,vjust=0.55))
  ggsave(paste0(pdf.dir,"ProfiledSignificantGeneSets.pdf"))
  return(plt)
}
prof(0.5)
```


After analyzing the plot created in the previous chunk of code, I will save dataframes for the TW.CT.a1  to carry foward for Gene Set Enrichment Analyses 

```{r save_dfs, echo=FALSE}
sig.gene.TW.a05.df <- sub_sig(TwCtOnly=TRUE,alphalimit=0.5) 
saveRDS(sig.gene.TW.a05.df,paste0(rds.dir,"sig.gene.TW.a05.df.RDS"))
```


Limit to only **Bonferroni-significant** results 


```{r profileBonferOnly, echo=FALSE,fig.height=8,fig.width=6}

prof_bonfer_only <- function(myalpha){
  df <- reformat_sub_sig(sub_sig(TwCtOnly=TRUE,bonfer=TRUE,alphalimit=0.5),"TW.CT.a"%&% myalpha%&%".bonfer")
  df$Type <- factor(df$Type,levels=c("T2D.Gene","Known.Region","Unknown"))
  plt <- ggplot(data=df,aes(x=Criteria)) +
    geom_bar(aes(fill=Type),color="black") +
    scale_fill_brewer() + ylab("Count") + 
    xlab("Gene Set") + 
    ggtitle("Bonferroni-significant Gene Sets") +
    scale_y_continuous(breaks=seq(0,230,10)) +
    theme_bw() + 
    theme(plot.title = element_text(face="bold",size=20),
          axis.title.x = element_text(face="bold"),
          axis.title.y = element_text(face="bold"),
          axis.text.x = element_text(angle=-45,vjust=0.55))
  ggsave(paste0(pdf.dir,"ProfiledSignificantGeneSets.Bonferroni.pdf"))
  return(plt)
}

prof_bonfer_only(0.5)
```

After analyzing the plot created in the previous chunk of code, I will save dataframes for the TW.CT.a1 and t2d.TW.a1 subsets to carry foward for Gene Set Enrichment Analyses 

```{r save_dfs_bonfer, echo=FALSE}
sig.gene.TW.a05.bonfer.df <- sub_sig(TwCtOnly=TRUE,bonfer=TRUE,alphalimit=0.5) 
saveRDS(sig.gene.TW.a05.bonfer.df,paste0(rds.dir,"sig.gene.TW.a05.bonfer.df.RDS"))
```

# Write results table

```{r elasticNet_table,echo=FALSE}

make.profile.table <- function(){
  rds.vec <- c("sig.gene.TW.a05.df.RDS","sig.gene.TW.a05.bonfer.df.RDS")
  Alpha <- c("1","1")
  Threshold <- c("FDR 5%","Bonferroni")
  T2D.Genes <- c()
  Known.Loci <- c()
  Unknown.Loci <- c()
  Total <- c() 
  for (rds in rds.vec){
    df <- readRDS(paste0(rds.dir,rds))
    df <- select(df,gene,type)
    df <- df[!duplicated(df), ]
    tab <- table(df$type)
    kg <- as.numeric(tab[1]); tg <- as.numeric(tab[2]); ug <- as.numeric(tab[3])
    tot <- sum(tg,kg,ug); Total <- append(Total,tot)
    T2D.Genes <- append(T2D.Genes,tg)
    Known.Loci <- append(Known.Loci,kg)
    Unknown.Loci <- append(Unknown.Loci,ug)
  }
  #meta.name <- gsub("DIAGRAM","DIAGRAM",meta.name) # technical correction for just DIAGRAM
  # Will be ignored for GERA 
  out.df <- data.frame(Alpha,Threshold,Total,T2D.Genes,Known.Loci,Unknown.Loci)
  tname <- "Summary of MetaXcan results from " %&% meta.name %&% " study"
  fname <- meta.name %&% ".profile.summary.table"
  make.latex.table(out.df,tname,fname)
}

make.profile.table()
```

