---
title: "3.2_ResultsTables.Rmd"
author: "Jason Torres"
date: "August 4, 2016"
output:
  pdf_document:
    toc: yes
  html_document:
    fig_caption: yes
    highlight: textmate
    number_sections: yes
    theme: journal
    toc: yes
---

This script generates table figures for top gene results from MetaXcan results

# Setup 

```{r setup, echo=FALSE}
library(data.table)
library(dplyr)
library(gridExtra)
library(grid)
library(ggplot2)
if (!require("RColorBrewer")) {
install.packages("RColorBrewer")
library(RColorBrewer)
}
"%&%" <- function(a,b) paste0(a,b)
root.dir <- "/Users/jtorres/Google Drive/Projects/MetaXcan/MetaXcanT2D/"
rds.dir <- root.dir %&% "RDS.gtexV6p/"
fig.dir <- root.dir %&% "figures.gtexV6p/"
```


# Functions 

Bonferonni-significance threshold, per tissue model in Meta-analysis 

```{r bonfer_df, echo=FALSE}

build_bonfer_df <- function(analysis.name){
  if (analysis.name=="DIAGRAM"){
    load("RDS.gtexV6p/diag3/metaXcan.results.Rdata")
  } else if (analysis.name=="GERA-T2D"){
    load("RDS.gtexV6p/gera/metaXcan.results.Rdata")
  } else if (analysis.name=="Meta-analysis"){
    load("RDS.gtexV6p/diag_gera_meta-analysis/metaXcan.results.Rdata")
  } else{
    stop("Enter valid analysis name: DIAGRAM, GERA-T2D, Meta-analysis")
  } 
  p.df <- cbind(select(p.EN.df,one_of("CrossTissue")),
                select(p.EN.df,contains("TW_"))) 
  Model <- names(p.df)
  Bonfer.Correct <- c()
  count.vec <- c()
  for (m in Model){
    #print(dim(select(p.df,one_of(m))))
    vec <- na.omit(as.numeric(as.matrix(select(p.df,one_of(m)))))
    count.vec <- append(count.vec,vec)
    bonfer <- 0.05 / length(vec)
    Bonfer.Correct <- append(Bonfer.Correct,bonfer)
  }
  bonfer.mult <- 0.05 / length(count.vec)
  Model <- append(Model,"MultipleCorrection")
  Bonfer.Correct <- append(Bonfer.Correct,bonfer.mult)
  df <- as.data.frame(cbind(Model,Bonfer.Correct),stringsAsFactors=FALSE)
  df$Bonfer.Correct <- as.numeric(df$Bonfer.Correct)
  return(df)
}

bonfer.df <- build_bonfer_df("Meta-analysis")

```


Bonferonni-significance threshold correcting for multiple models 

```{r mult_correct, echo=FALSE}

get_bonfer_sig <- function(){
  load(rds.dir %&% "/diag3/metaXcan.results.Rdata")
  p.df <- p.EN.df
  p.df <- cbind(select(p.df,one_of("Gene","CrossTissue")),
                select(p.df,contains("TW_")))
  vec <- c()
  for (c in 2:dim(p.df)[2]){
    col.vec <- na.omit(p.df[,c])
    #print(length(col.vec))
    vec <- append(vec,col.vec)
  }
  bonfer <- 0.05 / length(vec)
  return(bonfer)
}

get_bonfer_sig_meta <- function(){
  load(rds.dir %&% "/diag_gera_meta-analysis/metaXcan.results.Rdata")
  p.df <- p.EN.df
  p.df <- cbind(select(p.df,one_of("Gene","CrossTissue")),
                select(p.df,contains("TW_")))
  vec <- c()
  for (c in 2:dim(p.df)[2]){
    col.vec <- na.omit(p.df[,c])
    #print(length(col.vec))
    vec <- append(vec,col.vec)
  }
  bonfer <- 0.05 / length(vec)
  return(bonfer)
}

bonfer <- get_bonfer_sig()
bonfer.meta <- get_bonfer_sig_meta()
```



```{r build_table_core_df, echo=FALSE}

loci.df <- (readRDS("/Users/jtorres/Google Drive/Projects/MetaXcan/" %&% 
                "MetaXcanT2D/RDS.gtexV6p/loci.df.RDS")) %>% select(chrom:nearby.txend)

get_reported_genes <- function(gname){
  gvec <- filter(loci.df,nearby.gene==gname)$reported.gene
  write.vec <- c()
  count = 0 
  for (g in gvec){
    count = count + 1
    write.vec <- append(write.vec,g)
    if (count == 3){
      write.vec <- append(write.vec, "\n")
      count = 0
    }
  }
  gstring <- paste(write.vec,collapse=",")
  gstring <- gsub("\n,","\n",gstring)
  if (gstring==""){
    gstring = "none reported" 
  }
  return(gstring)
}

get_nhgri_ebi_t2d_genes <- function(){
  #server.dir <- "/Volumes/im-lab/nas40t2/jason/projects/ReferenceData/NHGRI/"
  #file.path <- server.dir %&% "nhgri_ebi-trait-genes.txt.gz"
  file.path <- "files_from_tarbell/nhgri_ebi-trait-genes.txt.gz"
  df <- fread("cat " %&% file.path %&% " | zmore")
  df <- filter(df,`DISEASE/TRAIT`=="Type_2_diabetes")
  gene.vec <- c()
  for (g in df$`REPORTED_GENE(S)`){
    s <- strsplit(g,",_")[[1]]
    gene.vec <- append(gene.vec,s)
  }
  gene.vec <- unique(gene.vec)
  return(gene.vec)
}

# 236 184
build_table_core_df <- function(){
  #nghri.path <- "/Volumes/im-lab/nas40t2/jason/projects/ReferenceData/" %&% 
  #  "NHGRI/reported_genes_t2d_11-20-2015.txt"
  #unique(fread(nghri.path,header=FALSE)$V1)
  nhgri.genes <- get_nhgri_ebi_t2d_genes() 
  diag.df <- readRDS(rds.dir %&% "diag3/sig.gene.TW.a05.df.RDS")
  gera.rep.df <- readRDS(rds.dir %&% "diag3/sig.genes.gera-rep.df.RDS")
  diag.bonfer <- c()
  diag.mult.correct <- c() 
  meta.z.score <- c()
  meta.p.value <- c()
  meta.mult.correct <- c()
  nhgri.gene <- c() 
  reported.genes <- c()
  meta.bonfer <- c() 
  load(rds.dir %&% "/diag_gera_meta-analysis/metaXcan.results.Rdata")
  p.meta.df <- p.EN.df
  z.meta.df <- z.EN.df
  for (i in 1:dim(gera.rep.df)[1]){
    g <- gera.rep.df[i,1]
    rep.genes <- get_reported_genes(g)
    reported.genes <- append(reported.genes,rep.genes)
    nhgri <- g %in% nhgri.genes 
    nhgri.gene <- append(nhgri.gene,nhgri)
    m <- gera.rep.df[i,4]
    diag.p <- gera.rep.df[i,6]
    val <- filter(diag.df,gene==g,model==m)$bonferonni[1] # correction since KLHL42 was counted 
    # twice since I manually corrected outdated name KLHDC5 to KLHL42 
    #print(c(g,val))
    diag.bonfer <- append(diag.bonfer,val)
    multsig <- diag.p <= bonfer
    diag.mult.correct <- append(diag.mult.correct,multsig)
    meta.z <- as.numeric(as.matrix(filter(z.meta.df,Gene==g) %>% select(one_of(m))))
    if (length(meta.z)==0){
      meta.z = NA
    }
    meta.p <- as.numeric(as.matrix(filter(p.meta.df,Gene==g) %>% select(one_of(m))))
    if (length(meta.p)==0){
      meta.p = NA
    }
    meta.z.score <- append(meta.z.score,meta.z) ##
    meta.p.value <- append(meta.p.value,meta.p)  ##
    meta.multsig <- (meta.p <= bonfer.meta) 
    if (length(meta.multsig)==0){
      meta.multsig = NA
    }
    meta.mult.correct <- append(meta.mult.correct,meta.multsig) ## 
    meta.tiss.bonfer <- as.numeric(filter(bonfer.df,Model==m) %>% select(one_of("Bonfer.Correct")))
    val <- (meta.p <= meta.tiss.bonfer)
    if (length(val)==0){
      val = NA
    }
    meta.bonfer <- append(meta.bonfer,val) ##
  }
  gera.rep.df <- cbind(nhgri.gene,reported.genes,gera.rep.df,diag.bonfer,diag.mult.correct,
                       meta.z.score,meta.p.value,meta.mult.correct,meta.bonfer)
  gera.rep.df <- select(gera.rep.df,one_of("chrom","gene","type","nhgri.gene",
                                           "reported.genes","model","diag.z.score",
                                           "diag.p.value","diag.bonfer",
                                           "diag.mult.correct","gera.z.score",
                                           "gera.p.value","replicated",
                                           "meta.z.score","meta.p.value",
                                           "meta.mult.correct","meta.bonfer"))
  out.df <- arrange(gera.rep.df,chrom)
  return(out.df)
}

process_core_df <- function(){
  core.df <- build_table_core_df()
  core.df$model <- gsub("TW_","",core.df$model)
  core.df$model <- gsub("-",".",core.df$model)
  core.df$diag.z.score <- prettyNum(core.df$diag.z.score,digits=2)
  core.df$diag.p.value <- prettyNum(core.df$diag.p.value,digits=2)
  core.df$gera.z.score <- prettyNum(core.df$gera.z.score,digits=2)
  core.df$gera.p.value <- prettyNum(core.df$gera.p.value,digits=2)
  core.df$meta.z.score <- prettyNum(core.df$meta.z.score,digits=2)
  core.df$meta.p.value <- prettyNum(core.df$meta.p.value,digits=2)
  core.df$reported.genes <- as.character(core.df$reported.genes)
  core.df$reported.genes <- gsub(",\n$","\n",core.df$reported.genes)
  #core.df$reported.genes <- gsub("\n[:alnum:]*",",\n",core.df$reported.genes)

  # Need to replace NA with Chromosome 12 for KLHDC5 
  klhdc5.df <- filter(core.df,gene=="KLHDC5")
  core.df <- filter(core.df,gene!="KLHDC5")
  klhdc5.df[1,1] <- 12
  core.df <- rbind(core.df,klhdc5.df)
  core.df <- arrange(core.df,chrom)
  return(core.df)
}

# stargazer 
```



```{r table_function, echo=FALSE}


edit_cell <- function(table, row, col, name="core-fg", ...){
  #http://stackoverflow.com/questions/32173928/r-tablegrob-change-format-of-row
  row <- row + 1 # my adjustment 
  col <- col + 1 # my adjustment
  l <- table$layout
  ids <- which(l$t %in% row & l$l %in% col & l$name==name)
  for (id in ids){
  newgrob <- editGrob(table$grobs[id][[1]], ...)
  table$grobs[id][[1]] <- newgrob
  }
  table
}


metaXcan_table_plot <- function(core.df,colorDIAG=TRUE){
  plot.df <- select(core.df,one_of("chrom","gene","reported.genes","model",
                                  "diag.z.score","diag.p.value","gera.z.score",
                                  "gera.p.value","meta.z.score","meta.p.value"))
  plot.df$model <- gsub("\\.","\n",plot.df$model)
  names(plot.df) <- c("Chr","Gene","Reported Genes\n within 1Mb Locus",
                      "Model","DIAGRAM\nZ-score","DIAGRAM\nP-value","GERA\nZ-score",
                      "GERA\nP-value","Meta-analysis\nZ-score","Meta-analysis\nP-value")
  mytheme <- gridExtra::ttheme_default(
    core = list(fg_params=list(cex = 0.7),
                bg_params=list(fill="ghostwhite",alpha=0.3,col="black")),
    colhead = list(fg_params=list(cex = 0.9),
                   bg_params=list(alpha=0.30,col="black")),
    rowhead = list(fg_params=list(cex = 0)))   

  tabgrob <- tableGrob(plot.df,
                      theme=mytheme)
  # Italicize all gene names 
  tabgrob <- edit_cell(tabgrob, row=1:dim(plot.df)[1], col=2, "core-fg", 
                 gp=gpar(fontface="italic"))
  vec <- (1:dim(plot.df)[1])[plot.df[,3]!="none reported"]
  tabgrob <- edit_cell(tabgrob, row=vec, col=3, "core-fg", 
                 gp=gpar(fontface="italic"))
  vec <- (1:dim(plot.df)[1])[plot.df[,3]=="none reported"]
  tabgrob <- edit_cell(tabgrob, row=vec, col=3, "core-bg", 
                 gp=gpar(fill="grey",alpha=0.2, 
                         col = "black", lwd=1))
  # Indicate (fill) unreported genes relative to DIAGRAM top 1000 SNPs and NHGRI 
  eval.vec <- c()
  for (i in 1:dim(plot.df)[1]){
    eval.nhgri <- core.df$nhgri.gene[i]
    g <- plot.df$Gene[i]
    repgenes <- plot.df[,3][i]
    eval.diag <- grepl(g,repgenes)
    eval <- eval.nhgri==FALSE & eval.diag==FALSE
    eval.vec <- append(eval.vec,eval)
  }
  vec <- (1:dim(plot.df)[1])[eval.vec]
  tabgrob <- edit_cell(tabgrob, row=vec, col=2, "core-bg", 
                 gp=gpar(fill="dodgerblue1",alpha=0.5, 
                         col = "dodgerblue4", lwd=3))
  #vec <- (1:dim(plot.df)[1])[!eval.vec]
  #tabgrob <- edit_cell(tabgrob, row=vec, col=2, "core-bg", 
  #               gp=gpar(fill="pink1",alpha=0.6, 
  #                       col = "pink4", lwd=3))
  # Indicate DIAGRAM Signficant P-values (fill red)
  if (colorDIAG==TRUE){
    vec1 <- (1:dim(plot.df)[1])[core.df$diag.mult.correct]
    tabgrob <- edit_cell(tabgrob, row=vec1, col=6, "core-bg", 
                   gp=gpar(fill="firebrick2",alpha=0.5, 
                           col = "firebrick4", lwd=3))
    vec2 <- (1:dim(plot.df)[1])[core.df$diag.bonfer]
    vec2 <- setdiff(vec2,vec1)
    tabgrob <- edit_cell(tabgrob, row=vec2, col=6, "core-bg", 
                   gp=gpar(fill="firebrick1",alpha=0.1, 
                           col = "firebrick4", lwd=3))
  }
  # Indicate Meta-analysis Signficant P-values (fill red)
  vec1 <- (1:dim(plot.df)[1])[core.df$meta.mult.correct]
  tabgrob <- edit_cell(tabgrob, row=vec1, col=10, "core-bg", 
                 gp=gpar(fill="firebrick2",alpha=0.5, 
                         col = "firebrick4", lwd=3))
  vec2 <- (1:dim(plot.df)[1])[core.df$meta.bonfer]
  vec2 <- setdiff(vec2,vec1)
  tabgrob <- edit_cell(tabgrob, row=vec2, col=10, "core-bg", 
                 gp=gpar(fill="firebrick1",alpha=0.1, 
                         col = "firebrick4", lwd=3))
  # GERA, Box Red if pval < 0.05
  vec <- (1:dim(plot.df)[1])[core.df$replicated]
  tabgrob <- edit_cell(tabgrob, row=vec, col=8, "core-bg", 
                 gp=gpar(col = "firebrick4", lwd=6))
  plt <- grid.arrange(tabgrob,nrow=1,ncol=1)
  return(plt)
}

```


Save all results for genes that are FDR < 5% signficant in at least one tissue


```{r Save_metaxcan_table_plots, echo=FALSE}

# Genome-wide FDR <= 5% 

manual_entry <- function(mygene,mod){
  temp.df <- T2Dcore.df
  sub.df <- filter(temp.df,gene==mygene)[1,]
  # DIAGRAM lookup 
  load(rds.dir%&%"diag3/metaXcan.results.Rdata")
  diag.mult <- 0.05/length(na.omit(as.numeric(as.matrix(select(p.EN.df,-contains("Gene"))))))
  diag.bon <- 0.05/length(na.omit(as.numeric(as.matrix(select(p.EN.df,one_of(mod))))))
  diag.z <- as.numeric(filter(z.EN.df,Gene==mygene) %>% select(one_of(mod)))
  diag.p <- as.numeric(filter(p.EN.df,Gene==mygene) %>% select(one_of(mod)))
  diag.mult.eval <- diag.p <= diag.mult
  diag.bon.eval <- diag.p <= diag.bon
  # GERA lookup 
  load(rds.dir%&%"gera/metaXcan.results.Rdata")
  gera.mult <- 0.05/length(na.omit(as.numeric(as.matrix(select(p.EN.df,-contains("Gene"))))))
  gera.bon <- 0.05/length(na.omit(as.numeric(as.matrix(select(p.EN.df,one_of(mod))))))
  gera.z <- as.numeric(filter(z.EN.df,Gene==mygene) %>% select(one_of(mod)))
  gera.p <- as.numeric(filter(p.EN.df,Gene==mygene) %>% select(one_of(mod)))
  gera.mult.eval <- gera.p <= gera.mult
  gera.bon.eval <- gera.p <= gera.bon
  # META lookup 
  load(rds.dir%&%"diag_gera_meta-analysis/metaXcan.results.Rdata")
  meta.mult <- 0.05/length(na.omit(as.numeric(as.matrix(select(p.EN.df,-contains("Gene"))))))
  meta.bon <- 0.05/length(na.omit(as.numeric(as.matrix(select(p.EN.df,one_of(mod))))))
  meta.z <- as.numeric(filter(z.EN.df,Gene==mygene) %>% select(one_of(mod)))
  meta.p <- as.numeric(filter(p.EN.df,Gene==mygene) %>% select(one_of(mod)))
  meta.mult.eval <- meta.p <= meta.mult
  meta.bon.eval <- meta.p <= meta.bon
  # Update entry 
  mymodel <- gsub("TW_","",mod)
  mymodel <- gsub("_","-",mymodel)
  sub.df$model <- mymodel
  sub.df$diag.z.score <- prettyNum(diag.z,digits=2)
  sub.df$diag.p.value <- prettyNum(diag.p,digits=2)
  sub.df$diag.bonfer <- diag.bon.eval
  sub.df$diag.mult.correct <- diag.mult.eval
  sub.df$gera.z.score <- prettyNum(gera.z,digits=2)
  sub.df$gera.p.value <- prettyNum(gera.p,digits=2)
  sub.df$gera.bonfer <- gera.bon.eval
  sub.df$gera.mult.correct <- gera.mult.eval
  sub.df$meta.z.score <- prettyNum(meta.z,digits=2)
  sub.df$meta.p.value <- prettyNum(meta.p,digits=2)
  sub.df$meta.bonfer <- meta.bon.eval
  sub.df$meta.mult.correct <- meta.mult.eval
  sub.df$replicated <- gera.p <= 0.05
  return(sub.df)
}

append_GWfdr05 <- function(T2Dcore.df){
  # Need to manually add these genes to the supplementary table 
  #[1] "TW_Cells_EBV-transformed_lymphocytes:CCNE2" = T2D.Gene 
  #[2] "TW_Lung:HCG27" =  T2D.Gene                   
  #[3] "TW_Muscle_Skeletal:TCP11L1"  = "Unknown"               
  #[4] "TW_Prostate:AP3S2"= T2D.Gene                        
  #[5] "TW_Skin_Not_Sun_Exposed_Suprapubic:FAM229B" = "Unknown"  
  df1 <- manual_entry("CCNE2","TW_Cells_EBV-transformed_lymphocytes")
  df2 <- manual_entry("HCG27","TW_Lung")
  df3 <- manual_entry("TCP11L1","TW_Muscle_Skeletal")
  df4 <- manual_entry("AP3S2","TW_Prostate")
  df5 <- manual_entry("FAM229B","TW_Skin_Not_Sun_Exposed_Suprapubic")
  df5$chrom <- 6
  df5$gene <- "FAM229B"
  df5$type <- "Unknown"
  df5$nhgri.gene <- FALSE
  df5$reported.genes <- "none reported"
  out.df <- rbind(df1,df2)
  out.df <- rbind(out.df,df3)
  out.df <- rbind(out.df,df4)
  out.df <- rbind(out.df,df5)
  out.df <- select(out.df,-contains("gera.bonfer")) %>% select(-contains("gera.mult.correct"))
  out.df <- rbind(out.df,T2Dcore.df)
  return(out.df)
}

save_metaXcan_plots <- function(){
  # Results from GenomeWide-DIAGRAM.Rmd script 
  fdr.genes <- readRDS(rds.dir%&%"genome.wide.FDR05.associations.RDS")
  fdr.genes <- gsub("TW_","",fdr.genes)
  T2Dcore.df <- process_core_df()
  # Adjust data frame entries 
  T2Dcore.df$reported.genes <- gsub("PRC1,VPS33B","AP3S2,PRC1,VPS33B",T2Dcore.df$reported.genes)
  T2Dcore.df$reported.genes <- gsub("AP3S2,AP3S2","AP3S2,PRC1,VPS33B",T2Dcore.df$reported.genes)
  T2Dcore.df$reported.genes <- gsub("MTNR1B,MTNR1B,MTNR1B,\nMTNR1B,MTNR1B,MTNR1B,\nMTNR1B,MTNR1B","MTNR1B",T2Dcore.df$reported.genes)
  T2Dcore.df$reported.genes <- gsub("CCHCR1,CCHCR1,TCF19,\nTCF19,POU5F1,POU5F1,\nHCG27,HCG27", "CCHCR1,TCF19,POU5F1\nHCG27",T2Dcore.df$reported.genes)
 T2Dcore.df$reported.genes <- gsub("DPY19L4,DPY19L4,INTS8,\nINTS8,CCNE2,CCNE2,\nTP53INP1,TP53INP1,NDUFAF6,\nNDUFAF6", "DPY19L4,INTS8,CCNE2\nTP53INP1,NDUFAF6",T2Dcore.df$reported.genes)  
 T2Dcore.df$reported.genes <- gsub("CDC123", "CDC123,CAMK1D",T2Dcore.df$reported.genes)  
  # filter down to FRD 05 (genome-wide) 
  T2Dcore.df$pairs <- paste0(T2Dcore.df$model,":",T2Dcore.df$gene)
  T2Dcore.df <- filter(T2Dcore.df,pairs %in% fdr.genes)
  T2Dcore.df <- select(T2Dcore.df,-contains("pairs"))
  T2Dcore.df <- arrange(T2Dcore.df,chrom,reported.genes)
  # Manually update THADA annotation 
  index <- match("THADA",T2Dcore.df$gene)
  T2Dcore.df$reported.genes[index] <- "THADA"
  # plots
  T2Dcore.df$model <- gsub("_","-",T2Dcore.df$model)
  T2Dcore.df <- append_GWfdr05(T2Dcore.df)
  T2Dcore.df <- arrange(T2Dcore.df,chrom,reported.genes)
  #saveRDS(T2Dcore.df,file=rds.dir%&%"FDR05.sig.df.RDS")
  plt1 <- metaXcan_table_plot(T2Dcore.df[1:40,])
  savename <- fig.dir %&% "fdr05_table_p" %&% 1 %&% ".pdf"
  ggsave(filename=savename,plot=plt1,height=15,width=12)  
  plt2 <- metaXcan_table_plot(T2Dcore.df[41:80,])
  savename <- fig.dir %&% "fdr05_table_p" %&% 2 %&% ".pdf"
  ggsave(filename=savename,plot=plt2,height=15,width=12)   
  plt3 <- metaXcan_table_plot(T2Dcore.df[81:120,])
  savename <- fig.dir %&% "fdr05_table_p" %&% 3 %&% ".pdf"
  ggsave(filename=savename,plot=plt3,height=15,width=12) 
  plt4 <- metaXcan_table_plot(T2Dcore.df[121:160,])
  savename <- fig.dir %&% "fdr05_table_p" %&% 4 %&% ".pdf"
  ggsave(filename=savename,plot=plt4,height=15,width=12) 
  plt5 <- metaXcan_table_plot(T2Dcore.df[161:207,])
  savename <- fig.dir %&% "fdr05_table_p" %&% 5 %&% ".pdf"
  ggsave(filename=savename,plot=plt5,height=15,width=12) 
  #t <- select(T2Dcore.df,one_of("gene","type","nhgri.gene","reported.genes"))[!duplicated(T2Dcore.df$gene),]
  #table(t$type)
  # t <- T2Dcore.df
  #t$gera.p.value <- as.numeric(t$gera.p.value)
  #t <- filter(t,gera.p.value <= 0.05)
  #length(unique(t$gene))
  #t <- select(t,one_of("gene","type","nhgri.gene","reported.genes"))
  #t <- t[!duplicated(t$gene),]
  #table(t$type)
}


#save_metaXcan_plots()
```


Save all results for genes that are Bonferonni-signficant in at least one tissue


```{r Save_metaxcan_bonfer_table_plots, echo=FALSE}




save_metaXcan_bonfer_single_plots <- function(){
  T2Dcore.df <- filter(process_core_df(),diag.bonfer==TRUE)
  T2Dcore.df$reported.genes <- gsub("PRC1,VPS33B","AP3S2,PRC1,VPS33B",T2Dcore.df$reported.genes)
  T2Dcore.df$reported.genes <- gsub("AP3S2,AP3S2","AP3S2,PRC1,VPS33B",T2Dcore.df$reported.genes)
  T2Dcore.df$reported.genes <- gsub("MTNR1B,MTNR1B,MTNR1B,\nMTNR1B,MTNR1B,MTNR1B,\nMTNR1B,MTNR1B","MTNR1B",T2Dcore.df$reported.genes)
  # Manually update THADA annotation 
  #index <- match("THADA",T2Dcore.df$gene)
  #T2Dcore.df$reported.genes[index] <- "THADA"
  # generate plots 
  plt1 <- metaXcan_table_plot(T2Dcore.df[1:30,])
  savename <- fig.dir %&% "bonfer-single-tab_p" %&% 1 %&% ".pdf"
  ggsave(filename=savename,plot=plt1,height=15,width=11)  
  plt2 <- metaXcan_table_plot(T2Dcore.df[31:58,])
  savename <- fig.dir %&% "bonfer-single-tab_p" %&% 2 %&% ".pdf"
  ggsave(filename=savename,plot=plt2,height=15,width=11)   
  plt3 <- metaXcan_table_plot(T2Dcore.df[59:86,])
  savename <- fig.dir %&% "bonfer-single-tab_p" %&% 3 %&% ".pdf"
  ggsave(filename=savename,plot=plt3,height=15,width=11) 
  plt4 <- metaXcan_table_plot(T2Dcore.df[87:110,])
  savename <- fig.dir %&% "bonfer-single-tab_p" %&% 4 %&% ".pdf"
  ggsave(filename=savename,plot=plt4,height=15,width=11)   
  plt5 <- metaXcan_table_plot(T2Dcore.df[110:124,])
  savename <- fig.dir %&% "bonfer-single-tab_p" %&% 5 %&% ".pdf" 
  ggsave(filename=savename,plot=plt5,height=15,width=11) 
  #plt6 <- metaXcan_table_plot(T2Dcore.df[131:150,])
  #savename <- fig.dir %&% "bonfer-single-tab_p" %&% 6 %&% ".pdf" 
  #ggsave(filename=savename,plot=plt6,height=15,width=11) 
  #plt7 <- metaXcan_table_plot(T2Dcore.df[151:164,])
  #savename <- fig.dir %&% "bonfer-single-tab_p" %&% 7 %&% ".pdf" 
  #ggsave(filename=savename,plot=plt7,height=15,width=11) 
  #plt8 <- metaXcan_table_plot(T2Dcore.df[165:180,])
  #savename <- fig.dir %&% "bonfer-single-tab_p" %&% 8 %&% ".pdf" 
  #ggsave(filename=savename,plot=plt8,height=15,width=11) 
}


save_metaXcan_bonfer_single_plots()
```



```{r Save_metaxcan_bonfer_all_plots, echo=FALSE}

save_metaXcan_bonfer_plots <- function(){
  T2Dcore.df <- filter(process_core_df(),diag.mult.correct==TRUE)
  T2Dcore.df$model <- gsub("_","-",T2Dcore.df$model)
  # Adjust data frame entries 
  T2Dcore.df$reported.genes <- gsub("PRC1,VPS33B","AP3S2,PRC1,VPS33B",T2Dcore.df$reported.genes)
  T2Dcore.df$reported.genes <- gsub("AP3S2,AP3S2","AP3S2,PRC1,VPS33B",T2Dcore.df$reported.genes)
  T2Dcore.df$reported.genes <- gsub("MTNR1B,MTNR1B,MTNR1B,\nMTNR1B,MTNR1B,MTNR1B,\nMTNR1B,MTNR1B","MTNR1B",T2Dcore.df$reported.genes)
  T2Dcore.df$reported.genes <- gsub("CCHCR1,CCHCR1,TCF19,\nTCF19,POU5F1,POU5F1,\nHCG27,HCG27", "CCHCR1,TCF19,POU5F1\nHCG27",T2Dcore.df$reported.genes)
  T2Dcore.df$reported.genes <- gsub("DPY19L4,DPY19L4,INTS8,\nINTS8,CCNE2,CCNE2,\nTP53INP1,TP53INP1,NDUFAF6,\nNDUFAF6", "DPY19L4,INTS8,CCNE2\nTP53INP1,NDUFAF6",T2Dcore.df$reported.genes)  
  T2Dcore.df$reported.genes <- gsub("CDC123", "CDC123,CAMK1D",T2Dcore.df$reported.genes) 
 # Arrange 
  T2Dcore.df <- arrange(T2Dcore.df,chrom,reported.genes)
  plt.manu <- metaXcan_table_plot(T2Dcore.df[1:49,],colorDIAG=FALSE)
  savename <- fig.dir %&% "bonfer-all-manutab" %&% 1 %&% ".pdf"
  ggsave(filename=savename,plot=plt.manu,height=15,width=11)  
}

save_metaXcan_bonfer_plots()
```





```{r save}

#T2Dcore.df <- filter(process_core_df(),diag.bonfer==TRUE)
T2Dcore.df <- filter(process_core_df())
T2Dcore.df <- process_core_df()
saveRDS(T2Dcore.df,file=rds.dir %&% "T2Dcore.df.RDS")

```

