---
title: "11.1_AnalyzeLoci.Rmd"
author: "Jason Torres"
date: "August 4, 2016"
output:
  pdf_document:
    toc: yes
  html_document:
    fig_caption: yes
    highlight: textmate
    number_sections: yes
    theme: journal
    toc: yes
---

# Set up project 

**Load libraries, assign paths, and read data**

Prerequisite: 

* Run JTbuild.t2d-loci-tables.extended.predFDR05.ImpG.py
* Run 03.0_FillMissingT2DlociGenes.Rmd

This generates an annotation file of all genes within 1Mb of the top 89 T2D genes implicated by the top 1000 SNPs from DIAGRAM (AMP-T2D)



```{r setup, echo=FALSE}

library("dplyr")
library("data.table")
"%&%" <- function(a,b) paste0(a,b) 

load("/Users/jtorres/Google Drive/Projects/MetaXcan/MetaXcanT2D/RDS/diag3/metaXcan.results.Rdata")
z.df <- cbind(select(z.L.df,one_of("Gene","CrossTissue")),
                select(z.L.df,contains("TW_")))

result.dir <- "/Volumes/im-lab/nas40t2/jason/projects/MetaXcan/results/"
#table.file <- result.dir %&% #"loci-tables/DIAGRAM-t2d-loci.table.extended.alpha1.predFDR05.txt.gz"
table.file <- result.dir %&% "loci-tables/DIAGRAM-t2d-loci.table.extended.alpha1.predFDR05.ImpG.APPENDED.txt.gz"
file.arg <- "cat " %&% table.file %&% " | zmore"

is.empty <- function(x) {
    return(length(x)==0)
}


```


Modify and save loci.df 
**Need to harmonize loci.df with data in p.L.df. Importantly, values corresponding to genes with cross-validation r2 < 1% in a given tissu model had values set to NA. Need to esure that data in loci.df is consistent.** 

```{r loci_df,  echo=FALSE}

subdf <- function(df,meta.name,type="TW") { 
  df <- tbl_df(df) %>% select(chrom:nearby.txend,contains(meta.name))
  tw.df <- df %>% select(chrom:nearby.txend,contains("TW_"),contains("DGN"),contains("CrossTissue"))
  org.df <- df %>% select(chrom:nearby.txend,contains("Organ_"),contains("CrossTissue"))
  ts.df <- df %>% select(chrom:nearby.txend,contains("TS_"),contains("CrossTissue"))
  names(tw.df) <- gsub(meta.name %&% "_","",names(tw.df))
  names(org.df) <- gsub(meta.name %&% "_","",names(org.df))
  names(ts.df) <- gsub(meta.name %&% "_","",names(ts.df))
  if (type=="TW"){
    return(tw.df)
  } else if (type=="Organ"){
    return(org.df)
  } else if (type=="TS"){
    return(ts.df)
  } else{
    stop("Need a valid type: TW, Organ, or TS")
  }
}


harmonize_loci_df <- function(){
  loci.df <- fread(file.arg)
  loci.df <- subdf(loci.df,"DIAGRAM3","TW")
  names(z.df) <- gsub("-",".",names(z.df))
  loci.mod.vec <- names(loci.df)[14:length(names(loci.df))]
  # Iterate over rows in loci.df
  for (i in 1:dim(loci.df)[1]){
    reported.gene <- as.character(loci.df[i,8])
    gene <- as.character(loci.df[i,11])
    #print(c(reported.gene,gene))
    # Iterate over tissue model columns; first model starts at column 14 (TW_Adipose_Subcutaneous)
    for (j in 14:length(names(loci.df))){
      model <- loci.mod.vec[j-13] # 13 is a correction value to ensure that model referencing 
      # in loci.mod.vec begins at 1
      locidf.val <- as.numeric(loci.df[i,j])
      zdf.val <- as.numeric(as.matrix(filter(z.df,Gene==gene) %>% select(one_of(model))))
      if (is.empty(zdf.val)){
        zdf.val <- NA
      }
      #print(c(locidf.val,zdf.val)) # Checking 
      if (!is.na(locidf.val) & is.na(zdf.val)){
        loci.df[i,j] <- zdf.val
      }
    }
  }
  savename <- "/Users/jtorres/Google Drive/Projects/MetaXcan/" %&% 
                "MetaXcanT2D/RDS/loci.df.RDS"
  print("Saving RDS file")
  saveRDS(loci.df,file=savename)
  return(loci.df)
}

harmonize_loci_df()

loci.df <- readRDS("/Users/jtorres/Google Drive/Projects/MetaXcan/" %&% 
                "MetaXcanT2D/RDS/loci.df.RDS")

```



# Functions 

Build bonferonni reference data frame 

```{r bonfer_df, echo=FALSE}

build_bonfer_df <- function(analysis.name){
  if (analysis.name=="DIAGRAM"){
    load("RDS/diag3/metaXcan.results.Rdata")
  } else if (analysis.name=="GERA-T2D"){
    load("RDS/gera/metaXcan.results.Rdata")
  } else if (analysis.name=="Meta-analysis"){
    load("RDS/diag_gera_meta-analysis/metaXcan.results.Rdata")
  } else{
    stop("Enter valid analysis name: DIAGRAM, GERA-T2D, Meta-analysis")
  } 
  p.df <- cbind(select(p.L.df,one_of("CrossTissue")),
                select(p.L.df,contains("TW_"))) 
  Model <- names(p.df)
  Bonfer.Correct <- c()
  count.vec <- c()
  for (m in Model){
    #print(dim(select(p.df,one_of(m))))
    vec <- na.omit(as.numeric(as.matrix(select(p.df,one_of(m)))))
    count.vec <- append(count.vec,vec)
    bonfer <- 0.05 / length(vec)
    Bonfer.Correct <- append(Bonfer.Correct,bonfer)
  }
  bonfer.mult <- 0.05 / length(count.vec)
  Model <- append(Model,"MultipleCorrection")
  Bonfer.Correct <- append(Bonfer.Correct,bonfer.mult)
  df <- as.data.frame(cbind(Model,Bonfer.Correct),stringsAsFactors=FALSE)
  df$Bonfer.Correct <- as.numeric(df$Bonfer.Correct)
  return(df)
}

bonfer.df <- build_bonfer_df("DIAGRAM")

```


```{r locus_functions, echo=FALSE}


locus.subset <- function(genename){
  df <- arrange(loci.df, nearby.txstart) %>% 
    filter(reported.gene==genename) %>%   select(-contains("length"),
                                                 -contains("tx"),-starts_with("win"))
  chrom <- df$chrom[1]; snp <- df$most.sig.snp[1]; repgene <- df$reported.gene[1]
  df <- select(df,nearby.gene:CrossTissue); names(df) <- gsub("TW_","",names(df))
  names(df) <- gsub("TS_","",names(df)); names(df) <- gsub("Organ_","",names(df))
  return(df)
}

get.window.pos <- function(genename){
  df <- filter(loci.df,reported.gene==genename, 
               nearby.gene==genename) %>% select(chrom:win.end)
  chrom <- as.numeric(df[1,1]); win.start <- as.numeric(df[1,2])
  win.end <- as.numeric(df[1,3]); winvec <- c(chrom,win.start,win.end)
  return(winvec)
}

calc_p <- function(z){
  return(2 * (1 - pnorm(abs(z))))
}

analyze_locus <- function(genename){
  l.df <- as.data.frame(locus.subset(genename),stringsAsFactors=FALSE)
  gvec <- l.df$nearby.gene
  mvec <- bonfer.df$Model
  bonfer.mult <- filter(bonfer.df,Model=="MultipleCorrection")[1,2]
  Locus <- c()
  Nearby.Gene <- c()
  Zscore <- c()
  PValue <- c()
  MODEL <- c()
  Model.Wide.Significant <- c() 
  for (g in gvec){
    for (m in mvec[1:42]){ # NOTE: mvec[43] is "MultipleCorrection" and not valid here 
      mthresh <- filter(bonfer.df,Model==m)[1,2]
      m <- gsub("TW_","",m); m <- sub("-",".",m)
      m <- sub("-",".",m)
      g.z <- as.numeric(as.matrix(filter(l.df,nearby.gene==g) %>% select(contains(m))))
      pval <- calc_p(g.z)
      if (!is.na(pval) & pval <= mthresh){
        Locus <- append(Locus,genename)
        Nearby.Gene <- append(Nearby.Gene,g)
        Zscore <- append(Zscore,g.z)
        PValue <- append(PValue,pval)
        MODEL <- append(MODEL,m)
        Model.Wide.Significant <- append(Model.Wide.Significant,(pval <= bonfer.mult))
      }
    }
  }
  df <- as.data.frame(cbind(Locus,Nearby.Gene,Zscore,PValue,MODEL,Model.Wide.Significant),stringsAsFactors=FALSE)
  return(df)
}

get_bonfer_mult_locus <- function(genename){
  l.df <- as.data.frame(locus.subset(genename),stringsAsFactors=FALSE)
  vec <- na.omit(as.numeric(as.matrix(l.df)))
  val <- 0.05/length(vec)
  return (val)
}

analyze_locus_v2 <- function(genename){
  l.df <- as.data.frame(locus.subset(genename),stringsAsFactors=FALSE)
  gvec <- l.df$nearby.gene
  mvec <- bonfer.df$Model
  bonfer.mult <- filter(bonfer.df,Model=="MultipleCorrection")[1,2]
  bonfer.mult.locus <- get_bonfer_mult_locus(genename)
  Locus <- c()
  Nearby.Gene <- c()
  Zscore <- c()
  PValue <- c()
  MODEL <- c()
  Genome.Wide.Significant <- c()
  Model.Wide.Significant <- c() 
  Locus.Significant <- c() 
  for (g in gvec){
    for (m in mvec[1:42]){ # NOTE: mvec[43] is "MultipleCorrection" and not valid here 
      mthresh <- filter(bonfer.df,Model==m)[1,2]
      m <- gsub("TW_","",m); m <- sub("-",".",m)
      m <- sub("-",".",m)
      g.z <- as.numeric(as.matrix(filter(l.df,nearby.gene==g) %>% select(contains(m))))
      pval <- calc_p(g.z)
#      if (!is.na(pval) & pval <= mthresh){
      if (!is.na(pval) & pval <= bonfer.mult.locus){
        Locus <- append(Locus,genename)
        Nearby.Gene <- append(Nearby.Gene,g)
        Zscore <- append(Zscore,g.z)
        PValue <- append(PValue,pval)
        MODEL <- append(MODEL,m)
        Model.Wide.Significant <- append(Model.Wide.Significant,(pval <= bonfer.mult))
        Genome.Wide.Significant <- append(Genome.Wide.Significant,(pval <= mthresh))
        Locus.Significant <- append(Locus.Significant,(pval <= bonfer.mult.locus))
      }
    }
  }
  df <- as.data.frame(cbind(Locus,Nearby.Gene,Zscore,PValue,MODEL,Model.Wide.Significant,
                            Genome.Wide.Significant,Locus.Significant),stringsAsFactors=FALSE)
  return(df)
}



```


Build signficant Loci data frame 

```{r build_sig_df, echo=FALSE}

gene.vec <- unique(loci.df$reported.gene)

build_sig_locidf <- function(genevector){
  out.df <- c()
  count <- 0
  for (g in genevector){
    count <- count + 1 
    print(c(count, g))
    df <- analyze_locus_v2(g)
    #df <- analyze_locus(g)
    out.df <- rbind(out.df,df)
  }
  out.df$Zscore <- as.numeric(out.df$Zscore)
  out.df$PValue <- as.numeric(out.df$PValue)
  out.df$Model.Wide.Significant <- as.logical(out.df$Model.Wide.Significant)
  out.df$Genome.Wide.Significant <- as.logical(out.df$Genome.Wide.Significant)
  out.df$Locus.Significant <- as.logical(out.df$Locus.Significant)
  return(out.df)
}

loci.sig.df <- build_sig_locidf(gene.vec)
saveRDS(loci.sig.df,"/Users/jtorres/Google Drive/Projects/MetaXcan/" %&% 
                "MetaXcanT2D/RDS/loci.sig.df.RDS")
loci.sig.df <- readRDS("/Users/jtorres/Google Drive/Projects/MetaXcan/" %&% 
                "MetaXcanT2D/RDS/loci.sig.df.RDS")

```


Profile results 

```{r profile, echo=FALSE}

profile <- function(multsig=FALSE){
  if (multsig==TRUE){
    df <- filter(loci.sig.df,Model.Wide.Significant==TRUE)
  } else {
    df <- loci.sig.df
  }
  loci.vec <- unique(df$Locus)
  LOCUS <- c()
  Count <- c()
  Genes <- c()
  ReportedGene.Sig <- c()
  for (loc in loci.vec){
    temp.df <- filter(df,Locus==loc)
    g.vec <- unique(temp.df$Nearby.Gene)
    count <- length(g.vec)
    gstring <- paste0(g.vec,collapse=",")
    repsig <- grepl(loc,gstring)
    LOCUS <- append(LOCUS,loc)
    Count <- append(Count,count)
    Genes <- append(Genes,gstring)
    ReportedGene.Sig <- append(ReportedGene.Sig,repsig)
  }
  out.df <- as.data.frame(cbind(LOCUS,Count,Genes,ReportedGene.Sig),stringsAsFactors=FALSE)  
  out.df$Count <- as.integer(out.df$Count)
  out.df$ReportedGene.Sig <- as.logical(out.df$ReportedGene.Sig) 
  return(out.df)
}

profile_v2 <- function(sigthresh){
  if (sigthresh=="multi"){
    df <- filter(loci.sig.df,Model.Wide.Significant==TRUE)
  } else if (sigthresh=="genome") {
    df <- filter(loci.sig.df,Genome.Wide.Significant==TRUE)
  } else if (sigthresh=="locus"){
    df <- filter(loci.sig.df,Locus.Significant==TRUE)
  } else {
    stop("Need to enter valid significant threshold: multi, genome, locus")
  }
  loci.vec <- unique(df$Locus)
  LOCUS <- c()
  Count <- c()
  Genes <- c()
  ReportedGene.Sig <- c()
  for (loc in loci.vec){
    temp.df <- filter(df,Locus==loc)
    g.vec <- unique(temp.df$Nearby.Gene)
    count <- length(g.vec)
    gstring <- paste0(g.vec,collapse=",")
    repsig <- grepl(loc,gstring)
    LOCUS <- append(LOCUS,loc)
    Count <- append(Count,count)
    Genes <- append(Genes,gstring)
    ReportedGene.Sig <- append(ReportedGene.Sig,repsig)
  }
  out.df <- as.data.frame(cbind(LOCUS,Count,Genes,ReportedGene.Sig),stringsAsFactors=FALSE)  
  out.df$Count <- as.integer(out.df$Count)
  out.df$ReportedGene.Sig <- as.logical(out.df$ReportedGene.Sig) 
  return(out.df)
}


```


# Analyze Results

```{r analyze_results, echo=FALSE}


print_report_nomultsig <- function(){
  result.sig.df <- profile_v2(sigthresh="genome")
  gene.vec <- unique(loci.df$reported.gene)
  sig.gene.vec <- unique(result.sig.df$LOCUS)
  totlength <- length(sig.gene.vec)
  rep.gene.sig.vec <- unique(filter(result.sig.df,ReportedGene.Sig==TRUE)$LOCUS)
  rep.gene.not.sig.vec <- unique(filter(result.sig.df,ReportedGene.Sig==FALSE)$LOCUS)
  rep.gene.sig.only.vec <- unique(filter(result.sig.df,Count==1,ReportedGene.Sig==TRUE)$LOCUS)  
  rep.gene.sig.not.only.vec <- unique(filter(result.sig.df,Count>1,ReportedGene.Sig==TRUE)$LOCUS) 
  rep.gene.not.sig.only.vec <- unique(filter(result.sig.df,Count==1,ReportedGene.Sig==FALSE)$LOCUS)  
  rep.gene.not.sig.not.only.vec <-   unique(filter(result.sig.df,Count>1,ReportedGene.Sig==FALSE)$LOCUS)
  prop1 <- length(rep.gene.sig.only.vec) / totlength
  prop2 <- length(rep.gene.sig.not.only.vec) / totlength
  prop3 <- length(rep.gene.not.sig.only.vec) / totlength
  prop4 <- length(rep.gene.not.sig.not.only.vec) / totlength
  print("There is a total of " %&% length(gene.vec) %&% " T2D Loci Analyzed")
  print("Number of Loci with MetaXcan-significant genes in at least one tissue: " %&% length(sig.gene.vec)) 
  print("Number of Loci with reported gene as signficant: " %&% length(rep.gene.sig.vec))
  print("Number of Loci with reported gene as NOT signficant: " %&% length(rep.gene.not.sig.vec))
  print("Number of Loci with ONLY reported gene as signficant: " %&% length(rep.gene.sig.only.vec) %&% " (" %&% prop1 %&% ")")   
  print("Number of Loci with multiple genes INCLUDING reported gene as signficant: " %&% length(rep.gene.sig.not.only.vec)  %&% " (" %&% prop2 %&% ")")
  print("Number of Loci with reported gene as NOT signficant and ONLY one significant gene: " %&% length(rep.gene.not.sig.only.vec) %&% " (" %&% prop3 %&% ")")
  print("Number of Loci with reported gene as NOT signficant and MULTIPLE significant genes: " %&% length(rep.gene.not.sig.not.only.vec) %&% " (" %&% prop4 %&% ")")
  save.list <- list(rep.gene.sig.only.vec,rep.gene.sig.not.only.vec,
                      rep.gene.not.sig.only.vec,rep.gene.not.sig.not.only.vec)
  print("Saving list as RDS")
  saveRDS(save.list,"/Users/jtorres/Google Drive/Projects/MetaXcan/" %&% 
                "MetaXcanT2D/RDS/loci.profile.list.RDS")
}

print_report_multsig <- function(){
  result.sig.df <- profile_v2(sigthresh="multi")
  gene.vec <- unique(loci.df$reported.gene)
  sig.gene.vec <- unique(result.sig.df$LOCUS)
  totlength <- length(sig.gene.vec)
  rep.gene.sig.vec <- unique(filter(result.sig.df,ReportedGene.Sig==TRUE)$LOCUS)
  rep.gene.not.sig.vec <- unique(filter(result.sig.df,ReportedGene.Sig==FALSE)$LOCUS)
  rep.gene.sig.only.vec <- unique(filter(result.sig.df,Count==1,ReportedGene.Sig==TRUE)$LOCUS)  
  rep.gene.sig.not.only.vec <- unique(filter(result.sig.df,Count>1,ReportedGene.Sig==TRUE)$LOCUS) 
  rep.gene.not.sig.only.vec <- unique(filter(result.sig.df,Count==1,ReportedGene.Sig==FALSE)$LOCUS)  
  rep.gene.not.sig.not.only.vec <-   unique(filter(result.sig.df,Count>1,ReportedGene.Sig==FALSE)$LOCUS)
  prop1 <- length(rep.gene.sig.only.vec) / totlength
  prop2 <- length(rep.gene.sig.not.only.vec) / totlength
  prop3 <- length(rep.gene.not.sig.only.vec) / totlength
  prop4 <- length(rep.gene.not.sig.not.only.vec) / totlength
  print("There is a total of " %&% length(gene.vec) %&% " T2D Loci Analyzed")
  print("Number of Loci with MetaXcan-significant genes, corrected for multiple tissues: " %&% length(sig.gene.vec)) 
  print("Number of Loci with reported gene as signficant: " %&% length(rep.gene.sig.vec))
  print("Number of Loci with reported gene as NOT signficant: " %&% length(rep.gene.not.sig.vec))
  print("Number of Loci with ONLY reported gene as signficant: " %&% length(rep.gene.sig.only.vec) %&% " (" %&% prop1 %&% ")")   
  print("Number of Loci with multiple genes INCLUDING reported gene as signficant: " %&% length(rep.gene.sig.not.only.vec)  %&% " (" %&% prop2 %&% ")")
  print("Number of Loci with reported gene as NOT signficant and ONLY one significant gene: " %&% length(rep.gene.not.sig.only.vec) %&% " (" %&% prop3 %&% ")")
  print("Number of Loci with reported gene as NOT signficant and MULTIPLE significant genes: " %&% length(rep.gene.not.sig.not.only.vec) %&% " (" %&% prop4 %&% ")")
  save.list <- list(rep.gene.sig.only.vec,rep.gene.sig.not.only.vec,
                      rep.gene.not.sig.only.vec,rep.gene.not.sig.not.only.vec)
  print("Saving list as RDS")
  saveRDS(save.list,"/Users/jtorres/Google Drive/Projects/MetaXcan/" %&% 
                "MetaXcanT2D/RDS/loci.profile.multcorrect.list.RDS")
}


print_report_loci <- function(){
  result.sig.df <- profile_v2(sigthresh="locus")
  gene.vec <- unique(loci.df$reported.gene)
  sig.gene.vec <- unique(result.sig.df$LOCUS)
  totlength <- length(sig.gene.vec)
  rep.gene.sig.vec <- unique(filter(result.sig.df,ReportedGene.Sig==TRUE)$LOCUS)
  rep.gene.not.sig.vec <- unique(filter(result.sig.df,ReportedGene.Sig==FALSE)$LOCUS)
  rep.gene.sig.only.vec <- unique(filter(result.sig.df,Count==1,ReportedGene.Sig==TRUE)$LOCUS)  
  rep.gene.sig.not.only.vec <- unique(filter(result.sig.df,Count>1,ReportedGene.Sig==TRUE)$LOCUS) 
  rep.gene.not.sig.only.vec <- unique(filter(result.sig.df,Count==1,ReportedGene.Sig==FALSE)$LOCUS)  
  rep.gene.not.sig.not.only.vec <-   unique(filter(result.sig.df,Count>1,ReportedGene.Sig==FALSE)$LOCUS)
  prop1 <- length(rep.gene.sig.only.vec) / totlength
  prop2 <- length(rep.gene.sig.not.only.vec) / totlength
  prop3 <- length(rep.gene.not.sig.only.vec) / totlength
  prop4 <- length(rep.gene.not.sig.not.only.vec) / totlength
  print("There is a total of " %&% length(gene.vec) %&% " T2D Loci Analyzed")
  print("Number of Loci with MetaXcan-significant genes, correcting for number of tests at LOCUS: " %&% length(sig.gene.vec)) 
  print("Number of Loci with reported gene as signficant: " %&% length(rep.gene.sig.vec))
  print("Number of Loci with reported gene as NOT signficant: " %&% length(rep.gene.not.sig.vec))
  print("Number of Loci with ONLY reported gene as signficant: " %&% length(rep.gene.sig.only.vec) %&% " (" %&% prop1 %&% ")")   
  print("Number of Loci with multiple genes INCLUDING reported gene as signficant: " %&% length(rep.gene.sig.not.only.vec)  %&% " (" %&% prop2 %&% ")")
  print("Number of Loci with reported gene as NOT signficant and ONLY one significant gene: " %&% length(rep.gene.not.sig.only.vec) %&% " (" %&% prop3 %&% ")")
  print("Number of Loci with reported gene as NOT signficant and MULTIPLE significant genes: " %&% length(rep.gene.not.sig.not.only.vec) %&% " (" %&% prop4 %&% ")")
  save.list <- list(rep.gene.sig.only.vec,rep.gene.sig.not.only.vec,
                      rep.gene.not.sig.only.vec,rep.gene.not.sig.not.only.vec)
  print("Saving list as RDS")
  saveRDS(save.list,"/Users/jtorres/Google Drive/Projects/MetaXcan/" %&% 
                "MetaXcanT2D/RDS/loci.profile.multcorrectLoci.list.RDS")
}

print_report_nomultsig()
print_report_multsig()
print_report_loci()
```

