---
title: "04.1_DelineateWindows.Rmd"
author: "Jason Torres"
date: "August 5, 2016"
output:
  pdf_document:
    toc: yes
  html_document:
    fig_caption: yes
    highlight: textmate
    number_sections: yes
    theme: journal
    toc: yes
---

```{r setup, echo=FALSE}

library(dplyr)
library(data.table)
library(ggplot2)
library(gridExtra)

"%&%" <- function(a,b) paste0(a,b)

root.dir <- "/Users/jtorres/Google Drive/Projects/MetaXcan/MetaXcanT2D/"
rds.dir <- root.dir %&% "RDS.gtexV6p/"

loci.df <- (readRDS(root.dir %&% "RDS.gtexV6p/loci.df.RDS")) %>% select(chrom:nearby.txend)

```


Here are what the windows look like: 

```{r show_T2D_gene_loci_windows, echo=FALSE, fig.height=8,fig.width=14}

win.df <- select(loci.df, one_of("chrom","win.start","win.end"))
win.df <- win.df[!duplicated(win.df),]
win.df <- group_by(win.df,chrom,win.start) %>% arrange(chrom, win.start)
win.df <- as.data.frame(win.df)
index <- 1:dim(win.df)[1]; win.df <- cbind(index,win.df)

plt <- ggplot(data=win.df,aes(x=index)) +
  geom_errorbar(aes(ymin=(win.start/10^6),ymax=(win.end/10^6))) + 
  theme_bw() + ggtitle("T2D Loci Windows") + xlab("Loci Index") +
  ylab("Genomic Position (Mb)") + facet_grid(. ~ chrom) + coord_flip() + theme(axis.text.x=element_text(size=6,angle=-45))
plt

```


Now, will determine the number and identify of non-overlapping windows 

```{r delineate, echo=FALSE}

delineate_per_chrom_single <- function(win.df){
  start1 <- win.df$win.start[1]
  end1 <- win.df$win.end[1]
  loc.vec <- c(1)
  win.start.vec <- c(start1)
  win.end.vec <- c(end1)
  out.df <- as.data.frame(cbind(loc.vec,win.start.vec,win.end.vec))
  names(out.df) <- c("locus","locus.start","locus.end")
  return(out.df)
}

delineate_per_chrom_double <- function(win.df){
  start1 <- win.df$win.start[1]
  end1 <- win.df$win.end[1]
  start2 <- win.df$win.start[2]
  end2 <- win.df$win.end[2]
  # test for overlap
  if (start2 > end1){
    loc.vec <- c(1,2)
    win.start.vec <- c(start1,start2)
    win.end.vec <- c(end1,end2)
  } else if (start2 > start1 & start2 < end1){
    loc.vec <- c(1)
    win.start.vec <- c(start1)
    win.end.vec <- c(end2) 
  } else{
    print("ERROR: Please Inspect")
  }
  out.df <- as.data.frame(cbind(loc.vec,win.start.vec,win.end.vec))
  names(out.df) <- c("locus","locus.start","locus.end")
  return(out.df)
}

delineate_per_chrom_multi <- function(win.df){
  loc.vec <- c()
  win.start.vec <- c() 
  win.end.vec <- c() 
  locnum <- 0 
  interval <- FALSE
  # The use of replace vec is to deal with rare case when end2 < end1 in an interval 
  replace.vec1 <- c() 
  replace.vec2 <- c() 
  for (i in 1:(length(win.df$index)-1)){
    start1 <- win.df$win.start[i]
    end1 <- win.df$win.end[i]
    start2 <- win.df$win.start[i+1]
    end2 <- win.df$win.end[i+1]
    # test for overlap
    if (end2 < end1){
      print(c(end2,end1))
      replace.vec1 <- append(replace.vec1,end2)
      replace.vec2 <- append(replace.vec2,end1)
      print(length(replace.vec1))
    }
    if (interval==FALSE & start2 > end1){
      locnum <- locnum + 1
      loc.vec <- append(loc.vec,locnum)
      win.start.vec <- append(win.start.vec,start1)
      win.end.vec <- append(win.end.vec,end1)
    } else if (interval==FALSE & start2 > start1 & start2 < end1){
      interval <- TRUE
      locnum <- locnum + 1
      loc.vec <- append(loc.vec, locnum)
      win.start.vec <- append(win.start.vec,start1)
    } else if (interval==TRUE & start2 > start1 & start2 < end1){
      interval <- TRUE 
    } else if (interval==TRUE & start2 > end1){
      win.end.vec <- append(win.end.vec,end1)
      interval <- FALSE
    } else {
      print("please inspect")
    }
    if (i == (length(win.df$index)-1)){
      win.end.vec <- append(win.end.vec,end2)
    }
  }
  if (length(win.end.vec)==(length(win.start.vec)+1)){
    loc.vec <- append(loc.vec,locnum+1)
    win.start.vec <- append(win.start.vec,start2)
  }
  out.df <- as.data.frame(cbind(loc.vec,win.start.vec,win.end.vec))
  names(out.df) <- c("locus","locus.start","locus.end")
  print(length(replace.vec1))
  if (length(replace.vec1) > 0){
    for (i in 1:length(replace.vec1)){
      out.df$locus.end <- gsub(replace.vec1[i],replace.vec2[i],out.df$locus.end)  
    }
    out.df$locus.end <- as.integer(out.df$locus.end)
  }
  return(out.df)
}

#delineate_per_chrom_multi(temp.df)



delineate_all <- function(full.win.df){
  chrom.vec <- sort(unique(full.win.df$chrom))
  chrom <- c() 
  out.df <- c() 
  for (c in chrom.vec){
    temp.df <- filter(full.win.df,chrom==c)
    dimen <- dim(temp.df)[1]
    if (dimen == 1){
      stack.df <- delineate_per_chrom_single(temp.df)
    } else if (dimen == 2){
      stack.df <- delineate_per_chrom_double(temp.df)
    } else{
      stack.df <- delineate_per_chrom_multi(temp.df)
    }
    chrom <- append(chrom,rep(c,dim(stack.df)[1]))
    out.df <- rbind(out.df,stack.df)
  }
  locus.index <- 1:dim(out.df)[1]
  out.df <- cbind(chrom,locus.index,out.df)
  names(out.df) <- c("chrom","locus","chrom.index","locus.start","locus.end")
  return(out.df)
}

del.df <- delineate_all(win.df)

```

Now, to append this locus information to window data frame

```{r append_info, echo=FALSE}

append_delinfo <- function(win.df,del.df){
  locus<- c()
  chrom.index <- c() 
  locus.start <- c()
  locus.end <- c()
  for (i in 1:dim(win.df)[1]){
    chromo <- win.df[i,2]
    start <- win.df[i,3]
    end <- win.df[i,4]
    temp.df <- filter(del.df, chrom==chromo,locus.start<=start,locus.end>=end)
    locus <- append(locus,temp.df[1,2])
    chrom.index <- append(chrom.index,temp.df[1,3])
    locus.start <- append(locus.start,temp.df[1,4])
    locus.end <- append(locus.end,temp.df[1,5])
  }
  join.df <- as.data.frame(cbind(locus,chrom.index,locus.start,locus.end))
  out.df <- cbind(win.df,join.df)
  out.df$locus.start <- as.integer(out.df$locus.start)
  out.df$locus.end <- as.integer(out.df$locus.end)
  return(out.df)
}

loc.win.df <- append_delinfo(win.df,del.df)

```


Here, I visualize the locus windows 

```{r show_non_overlapping, echo=FALSE, fig.height=12,fig.width=14}


pltA <- ggplot(data=loc.win.df,aes(x=index)) +
  geom_errorbar(aes(ymin=(win.start/10^6),ymax=(win.end/10^6),
                    color=as.factor(chrom.index))) + 
  theme_bw() + ggtitle("T2D Loci Windows\nPer Reported Gene") + xlab("Locus Index") +
  ylab("Genomic Position (Mb)") + coord_flip() +  facet_grid(. ~ chrom) + 
  scale_color_discrete(name="Chromosome Index") + theme(axis.text.x=element_text(size=6,angle=-45))


pltB <- ggplot(data=del.df,aes(x=locus)) +
  geom_errorbar(aes(ymin=(locus.start/10^6),ymax=(locus.end/10^6))) + 
  facet_grid(. ~ chrom) + 
  theme_bw() + ggtitle("T2D Loci Windows\nNon-overlapping") + xlab("Locus Index") +
  ylab("Genomic Position (Mb)") + coord_flip() + theme(axis.text.x=element_text(size=6,angle=-45))


grid.arrange(pltA,pltB,nrow=2,ncol=1)
```

Now, will add gene annotations and window distances to loc.win.df:

```{r add_gene_info, echo=FALSE}

add_gene_info <- function(){
  chrom <- c()
  win.rep.genes <- c()
  win.genes <-c()
  locus.rep.genes <- c()
  locus.genes <- c()
  win.distance <- c()
  locus.distance <- c() 
  for (i in 1:dim(loc.win.df)[1]){
    chr <- loc.win.df[i,2] 
    wstart <- loc.win.df[i,3]
    wend <- loc.win.df[i,4]
    wdist <- wend - wstart
    lstart <- loc.win.df[i,7]
    lend <- loc.win.df[i,8]
    ldist <- lend - lstart
    win.distance <- append(win.distance,wdist)
    locus.distance <- append(locus.distance,ldist)
    temp1.df <- filter(loci.df,chrom==chr,win.start==wstart,win.end==wend)
    #chr <- temp1.df$chrom[1]
    chrom <- append(chrom,chr)
    wrg <- paste(unique(temp1.df$reported.gene),collapse=",")
    wg <- paste(unique(temp1.df$nearby.gene),collapse=",")
    win.rep.genes <- append(win.rep.genes,wrg)
    win.genes <- append(win.genes,wg)
    temp2.df <- filter(loci.df,chrom==chr,win.start>=lstart,win.end<=lend)
    lrg <- paste(unique(temp2.df$reported.gene),collapse=",")
    lg <- paste(unique(temp2.df$nearby.gene),collapse=",")
    locus.rep.genes <- append(locus.rep.genes,lrg)
    locus.genes <- append(locus.genes,lg)    
  }
  out.df <- cbind(loc.win.df,win.rep.genes,win.genes,locus.rep.genes,locus.genes,win.distance,locus.distance)
  return(out.df)
}

uniq.loci.df <- add_gene_info()
saveRDS(uniq.loci.df,file=rds.dir %&% "uniq.loci.df.RDS")

```


Let's look at histograms of the window and loci distances:

```{r dist_hist, echo=FALSE, fig.height=10,fig.width=5}

par(mfrow=c(2,1))
hist(unique(uniq.loci.df$win.distance/10^6),breaks=20,col="dodgerblue1",
     main="Distances of T2D Loci Windows",xlab=("Genomic Distance (Mb)"),xlim=c(2,3.5))
hist(unique(uniq.loci.df$locus.distance/10^6),breaks=50,col="dodgerblue4",
     main="Distances of Non-overlapping\nT2D Loci Windows",xlab=("Genomic Distance (Mb)"),xlim=c(2,3.5))


```


Here are the non-overlapping loci by reported genes:

```{r nonover_tab, echo=FALSE}

tab.df <- select(uniq.loci.df,one_of("chrom","locus","locus.rep.genes","locus.start","locus.end","locus.genes","locus.distance"))
tab.df <- tab.df[!duplicated(tab.df),]
print(tab.df)
saveRDS(tab.df,file=rds.dir%&%"non.overlapping.t2dLoci.ref.df.RDS")

```


