---
title: "06.6_Mongenicity.Rmd"
author: "Jason Torres"
date: "September 22, 2016"
output:
  pdf_document:
    toc: yes
  html_document:
    fig_caption: yes
    highlight: textmate
    number_sections: yes
    theme: journal
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Heuristic for assessing "Polygenicity" 

```{r heuristic} 

vcf.path <- "/group/im-lab/nas40t2/Data/dbGaP/GTEx/41400/gtex/exchange/" %&% "GTEx_phs000424/exchange/" %&% "analysis_releases/GTEx_Analysis_2014-06-13/genotypes/OMNI_arrays" %&% "GTEx_Analysis_2014-06-13_OMNI_2.5M_5M_451Indiv_allchr_genot_imput_info04_maf01_HEW1E6.vcf.gz"

library("ceu1kg")

server.dir <- "/Volumes/im-lab/nas40t2/jason/projects/ReferenceData/UCSC/"
gencode.path <- server.dir %&%  "GENCODE.Genes.V19.Basic.annotated.txt.gz"
gencode.ref.df <- fread("cat " %&% gencode.path %&% " | zmore")

gene2ensid <- function(mygene){
  genename <- filter(gencode.ref.df,hg19.wgEncodeGencodeAttrsV19.geneName==mygene)$hg19.wgEncodeGencodeAttrsV19.geneId[1]
  return(genename)
}

beta.dir <- "/Volumes/im-lab/nas40t2/jason/projects/MetaXcan/db_files/gtex_v6p/alpha_0.5/allBetas/"

get_model_snps <- function(mygene,model){
  print("Looking up model SNPs: " %&% model %&% " ; " %&% mygene)
  file.path <- beta.dir %&% model %&% ".allBetas.txt.gz"
  temp.df <- fread("cat " %&% file.path %&% " | zmore")
  ensid <- filter(ggd.df,symbol==mygene)$ensgene
  temp.df <- filter(temp.df,grepl(ensid, temp.df$gene))
  return(temp.df)
}

# Test 
mod.snps <- get_model_snps("TCF7L2","Artery_Aorta")$rsid

get_ld_matrix <- function(chrom,mod.snps){
  pat <- "chr" %&% chrom %&% ".rda"
  dir.vec <- dir(system.file("parts",package="ceu1kg"))
  rda.index <- grep(pat,dir.vec)
  lk = load(dir(system.file("parts", package="ceu1kg"),full=TRUE)[rda.index])  
  snp.matrix <- get(lk)
  snp.vec <- attributes(snp.matrix)[[3]][[2]]
  sub.matrix <- snp.matrix[, (snp.vec %in% mod.snps)]
  ld.matrix <- ld(sub.matrix,depth=length(mod.snps),stats="R.squared")
  return(ld.matrix)
}

get_cov_matrix <- function(mygene,model){
  cov.file <- "/Volumes/im-lab/nas40t2/jason/projects/MetaXcan/" %&% "gtex_covariance/unscaled/1/TW_" %&% model %&% "/covariance.txt.gz"
  df <- fread("cat " %&% cov.file %&% " | zmore")
  ensid <- gene2ensid(mygene)
  df <- na.omit(filter(df,GENE==ensid))
  snps <- unique(df$RSID1)
  mat <- matrix(data=NA,nrow=length(snps),ncol=length(snps),
                dimnames=list(snps,snps))
  for (i in 1:length(snps)){
    for (j in 1:length(snps)){
      snp1 <- snps[i]
      snp2 <- snps[j]
      valA <- filter(df,RSID1==snp1,RSID2==snp2)$VALUE
      valB <- filter(df,RSID2==snp1,RSID1==snp2)$VALUE
      val<- (na.omit(c(valA,valB)))[1]
      mat[i,j] <- val
    }
  }
  return(mat)
}

get_metaxcan_results <- function(mygene,model){
  file.path <- "/Volumes/im-lab/nas40t2/jason/projects/MetaXcan/results/" %&% 
    "DIAGRAM3_ImpG_0.80/alpha_1/TW_" %&% model %&% ".zscores.csv"
  df <- filter(fread(file.path),gene_name==mygene)
  return(df)
}

heuristic <- function(mygene,model,gwas.df,ld.cutoff=0.8){
  # get metaxcan results 
  print("Loading MetaXcan result file...")
  mxn.df <- get_metaxcan_results(mygene,model)
  mxn.z <- mxn.df$zscore
  sigma_g <- sqrt(mxn.df$VAR_g)
  # get chromsome 
  chrom <- as.integer(get_chrom(mygene))
  # get model snps 
  mod.df <- get_model_snps(mygene,model)
  # load covariance matrix 
  print("Loading covariance matrix...")
  cov.mat <- get_cov_matrix(mygene,model)
  snps <- row.names(cov.mat)
  index <- match(snps,mod.df$rsid)
  mod.df <- mod.df[index,]
  beta.vec <- matrix(mod.df$beta,ncol=1)
  ##sigma.g <- t(beta.vec) %*% cov.mat %*% beta.vec
  # get ld matrix 
  ld.mat <- as.matrix(get_ld_matrix(chrom,mod.df$rsid))  
  # calculate terms for each SNP in the model
  print("Generating heuristic...")
  snp.vec <- c() 
  term.vec <- c() 
  for (i in 1:dim(mod.df)[1]){
    snp <- mod.df$rsid[i]
    w_lg <- mod.df$beta[i]
    alt_allele <- mod.df$alt[i]
    gwas_effect_allele <- filter(gwas.df,SNP==snp)$A2
    if (gwas_effect_allele != alt_allele){
      w_lg <- (-1)*w_lg
    }
    index <- match(snp,snps)
    sigma_l <- cov.mat[index,index]
    gwas_z <- filter(gwas.df,SNP==snp)$Z
    term <- w_lg * (sigma_l / sigma_g) * gwas_z
    snp.vec <- append(snp.vec,snp)
    term.vec <- append(term.vec,term)
  }
  print("MetaXcan Z: " %&% mxn.z %&% "; Sum of terms: " %&% sum(term.vec))
  df <- as.data.frame(cbind(snp.vec,term.vec),
                      stringsAsFactors=FALSE)
  names(df) <- c("snp","term")
  df$term <- as.numeric(df$term)
  # determine sign of metaxcan z score and limit to SNPs with same sign 
  zsign <- sign(mxn.z)
  df <- filter(df,sign(term)==zsign)
  prop <- abs(df$term) / sum(abs(df$term)) # proportion explained by each SNP
  df <- cbind(df,prop); df <- arrange(df,desc(prop))
  lead.snp <- df$snp[1]
  # append pairwise LD with the lead SNP 
  ld.vec <- c()
  flag.vec <- c()
  same <- c()
  diff <- c() 
  for (snp in df$snp){
    val1 <- ld.mat[grep(snp,row.names(ld.mat)),
                   grep(lead.snp,row.names(ld.mat))]
    val2 <- ld.mat[grep(lead.snp,row.names(ld.mat)),
                   grep(snp,row.names(ld.mat))]    
    ld <- max(val1,val2)
    if (snp == lead.snp){
      ld <- 1 
    }
    if (ld >= ld.cutoff){
      flag <- "same"
      same <- append(same,snp)
    } else{
      flag <- "different"
      diff <- append(diff,snp)
    }
    ld.vec <- append(ld.vec,ld)
    flag.vec <- append(flag.vec,flag)
  }
  df <- cbind(df,ld.vec,flag.vec)
  names(df)[c(dim(df)[2]-1,dim(df)[2])] <- c("ld","flag")
  df$flag <- as.character(df$flag)
  heur <- sum(filter(df,flag=="same")$prop)
  print("Process complete")
  return(list(heur,same,diff))  
}

build_heur_df <- function(gwas.df,ld.cutoff){
  gene.vec <- c()
  mod.vec <- c()
  heur.vec <- c() 
  same.vec <- c()
  diff.vec <- c()
  bad.genes <- c("OR2H1","HCG27")
  for (i in 1:dim(sig.df)[1]){
    print(i)
    mygene <- sig.df$gene[i]
    mymod <- gsub("\\.","-",sig.df$model[i])
    if (!(mygene %in% bad.genes)){
      heur.list <- heuristic(mygene,mymod,gwas.df,ld.cutoff)
      heur <- heur.list[[1]]
      same.snps <- paste(heur.list[[2]],collapse=",")
      diff.snps <- paste(heur.list[[3]],collapse=",")
      gene.vec <- append(gene.vec,mygene)
      mod.vec <- append(mod.vec,mymod)
      heur.vec <- append(heur.vec,heur) 
      same.vec <- append(same.vec,same.snps)
      diff.vec <- append(diff.vec,diff.snps)
    }
  }
  out.df <- as.data.frame(cbind(gene.vec,mod.vec,heur.vec,
                                same.vec,diff.vec),
                          stringsAsFactors=FALSE)
  names(out.df) <- c("gene","model","heur.score","same.snps","different.snps")
  out.df$heur.score <- as.numeric(out.df$heur.score)
  out.df <- arrange(out.df,desc(heur.score))
}

heur80.df <- build_heur_df(diag.df,ld.cutoff=0.80)
saveRDS(heur80.df,file=root.dir%&%"RDS/heur.ld80.df.RDS")

heur50.df <- build_heur_df(diag.df,ld.cutoff=0.50)
saveRDS(heur50.df,file=root.dir%&%"RDS/heur.ld50.df.RDS")

heur30.df <- build_heur_df(diag.df,ld.cutoff=0.30)
saveRDS(heur30.df,file=root.dir%&%"RDS/heur.ld30.df.RDS")

heur80.df <- readRDS(root.dir%&%"RDS/heur.ld80.df.RDS")
heur50.df <- readRDS(root.dir%&%"RDS/heur.ld50.df.RDS")
heur30.df <- readRDS(root.dir%&%"RDS/heur.ld50.df.RDS")

#miami("WFS1","Nerve-Tibial") # good example of possible "monogenic" 
#miami("HMG20A","Pituitary") # good example of possible "polygenic" 

```

