---
title: "Fill-Missing-T2DLociGenes.Rmd"
author: "Jason Torres"
date: "August 4, 2016"
output: pdf_document
---

# Setup 

```{r setup, echo=FALSE}

library("dplyr")
library("data.table")
"%&%" <- function(a,b) paste0(a,b)

nghri.path <- "/Volumes/im-lab/nas40t2/jason/projects/ReferenceData/" %&% 
  "NHGRI/reported_genes_t2d_11-20-2015.txt"
res.dir <- "/Volumes/im-lab/nas40t2/jason/projects/MetaXcan/results/" 
ref.dir <- "/Volumes/im-lab/nas40t2/jason/projects/ReferenceData/"
rds.dir <- "/Users/jtorres/Google Drive/Projects/MetaXcan/MetaXcanT2D/RDS/"
fig.dir <- "/Users/jtorres/Google Drive/Projects/MetaXcan/MetaXcanT2D/figures/"
load(rds.dir %&% "diag3/metaXcan.results.Rdata")
z.df <- cbind(select(z.L.df,one_of("Gene","CrossTissue")),
              select(z.L.df,contains("TW_")))
gencode.dir = "/Volumes/im-lab/nas40t2/hwheeler/PrediXcan_CV/"
gencode.file = gencode.dir %&% "gencode.v12.V1.summary.protein.nodup.genenames"

```


Gene sets 

```{r gene_sets, echo=FALSE}

#nhgri.genes <- unique(fread(nghri.path,header=FALSE)$V1)

ref.df <- fread("cat " %&% gencode.file %&% " | zmore")
ref.genes <- unique(ref.df$V6)

mtxn.genes <- unique(z.df$Gene)

t2d.portal.df<- fread(ref.dir %&% "diagramT2D/t2d-diagram3-results.csv")
t2d.portal.genes <- unique(t2d.portal.df$`nearest gene`)
t2d.portal.snps <- unique(t2d.portal.df$`rsid`)

loci.df <- fread(res.dir %&% "t2d-gene-info.incomplete.txt")
loci.genes <- unique(loci.df$reported.gene)
diffgenes <- setdiff(t2d.portal.genes,loci.genes)

```


# Build data frame of missing gene information 


```{r build, echo=FALSE}

reported.gene <- c("NDUFAF6","KLHL42","NCR3LG1","CCDC41","ALX1","HLA-C","TCF19")
chrom <- c(8,12,11,12,12,6,6)
reported.txstart <- c(96037213,27933186,17373308,94700225,85673885,31236526,31126319)
reported.txend <- c(96070944,27955973,17398868,94853764,85695562,31239907,31134936)
most.sig.snp <- c("rs13255935","rs10842993","rs7928810","rs3761675","rs7953666","rs115417228","rs114799468")
sig.snp.pos <- c(95965435,27964398,17372693,94885799,85791582,31236526,31126319)
min.gwas.pval <- c(2.50e-9,5.60e-9,3.60e-8,2.90e-7,0.00000170,0.00000190,0.00000200)
win.start <- sig.snp.pos - 1e6
win.end <- sig.snp.pos + 1e6
win.length <- win.end - win.start

miss.df <- as.data.frame(cbind(chrom,win.start,win.end,win.length,most.sig.snp,sig.snp.pos,min.gwas.pval,
                 reported.gene,reported.txstart,reported.txend),stringsAsFactors=FALSE)

```


Need to manually append a few entries to t2d-info df; of critical importance is *_PLEKHS1_*

```{r append_missing_entries_manually, echo=FALSE}

manually_append_PLEKHS1 <- function(df){
  # Gencode V19, Basic Annotation, GRCh37/hg19
  append.vec <- c(10,113732882,115823426,2090544,"rs4506565",114756041,
                  9.6e-77,"TCF7L2",114710009,114927437,"PLEKHS1",115511213,115542191)
  df$nearby.gene <- as.character(df$nearby.gene)
  return.df <- rbind(df,append.vec)
  return(return.df) 
}

```




```{r build_function,echo=FALSE}

sync_write_df <- function(df){
  df <- data.frame(df)
  out.df <- df
  for (r in 1:dim(df)[1]){ # iterate through each row
    # determine if a reported.gene:nearest.gene pair is incongruent 
    # i.e. the reverse is not included where appropriate 
    r.gene <- df[r,8]; r.start <- df[r,9]; r.end <- df[r,10]
    n.gene <- df[r,11]; n.start <- df[r,12]; n.end <- df[r,13]
    if (n.gene %in% df$reported.gene){ # for each index gene 
      #determine if nearest.gene is in the set of reported genes 
      temp.df <- filter(df, reported.gene==n.gene)
      if ((r.gene %in% temp.df$nearby.gene)==FALSE){ # determine if index gene 
        # is NOT included among the nearest genes for the secondary reference gene (e.g. KCNJ11)
        win.start <- temp.df[1,2]; win.end <- temp.df[1,3]
        if (r.start >= win.start & r.end <= win.end){ # check if index gene should be included 
          #print(df[r,])
          new.df <- temp.df[1,1:10] 
          nearby.gene <- r.gene
          nearby.txstart <- r.start
          nearby.txend <- r.end
          new.df <- cbind(new.df,nearby.gene,nearby.txstart,nearby.txend)
          out.df <- rbind(out.df,new.df)
        }
      }
    }
  }
  return(out.df)
}


extend_t2d_info <- function(){
  names(ref.df) <- c("chrom","type","tx.start","tx.end","ensid","gene","gene.type")
  ref.df$chrom <- as.integer(gsub("chr","",ref.df$chrom))
  diffgenes <- c("NDUFAF6","KLHL42","NCR3LG1","CCDC41","ALX1","HLA-C","TCF19")
  
  out.df <- c() 
  # Iterate over each of the missing genes
  for (g in diffgenes){
    # lookup window for missing gene 
    temp.df <- filter(miss.df,reported.gene==g) %>% select(one_of("chrom","win.start","win.end"))
    chromo <- as.integer(temp.df[1,1]) 
    win.start <- as.numeric(temp.df[1,2])
    win.end <- as.numeric(temp.df[1,3])
    # Look up genes in the window 
    win.df <- filter(ref.df,chrom==chromo,tx.start>=win.start,tx.end<=win.end)
    win.df <- filter(win.df,gene %in% mtxn.genes)
    window.genes <- unique(win.df$gene)
    # Iterate over window genes 
    for (w in window.genes){
      nearby.gene <- w
      nearby.txstart <- as.integer(filter(win.df,gene==nearby.gene) %>% select(one_of("tx.start")))
      nearby.txend <- as.integer(filter(win.df,gene==nearby.gene) %>% select(one_of("tx.end")))
      #row.vec <- as.vector(as.matrix(filter(miss.df,reported.gene==g)))
      row.df <- (filter(miss.df,reported.gene==g))[1,]
      row.df <- cbind(row.df,nearby.gene,nearby.txstart,nearby.txend)
      # build out df with completed rows of missing gene loci window gene information 
      out.df <- rbind(out.df,row.df)
    }
    # Be sure to add the missing (reported) gene itself; this is extra step to ensure that this is done
    # will remove any possible duplicates before final output 
    row.df <- (filter(miss.df,reported.gene==g))[1,] 
    nearby.gene <- row.df$reported.gene
    nearby.txstart <- row.df$reported.txstart
    nearby.txend <- row.df$reported.txend
    row.df <- cbind(row.df,nearby.gene,nearby.txstart,nearby.txend)
    out.df <- rbind(out.df,row.df)  
  }
  # Manually append PLEKHS1 information (TCF7L2 locus) 
  out.df <- manually_append_PLEKHS1(out.df)
  # adjust variable types  
  out.df$chrom <- as.integer(out.df$chrom)
  out.df$win.start <- as.integer(out.df$win.start)
  out.df$win.end <- as.integer(out.df$win.end)
  out.df$win.length <- as.integer(out.df$win.length)
  out.df$most.sig.snp <- as.character(out.df$most.sig.snp)
  out.df$sig.snp.pos <- as.integer(out.df$sig.snp.pos)
  out.df$min.gwas.pval <- as.numeric(out.df$min.gwas.pval)
  out.df$reported.gene <- as.character(out.df$reported.gene)
  out.df$reported.txstart <- as.integer(out.df$reported.txstart)
  out.df$reported.txend <- as.integer(out.df$reported.txend)
  out.df$nearby.gene <- as.character(out.df$nearby.gene)
  out.df$nearby.txstart <- as.integer(out.df$nearby.txstart)
  out.df$nearby.txend <- as.integer(out.df$nearby.txend)
  out.df <- out.df[!duplicated(out.df),]
  # write extended t2d info file 
  print("Saving extended information file")
  savename <- (res.dir %&% "t2d-gene-info.appended.txt") # need to rename before re-running other R scripts
  write.df <- rbind(loci.df,out.df)
  write.df <- sync_write_df(write.df) # correct incongruent entries 
  write.table(write.df,file=savename,quote=FALSE,sep="\t",row.names=FALSE)
  out.df <- setdiff(write.df, loci.df) # modification to allow for the additional entries created with sync_write_df
  return(out.df)
}


```


Expand extended locus file 

```{r expand_locus, echo=FALSE}

expand_loci_file <- function(){
  #locus.tab.path <- res.dir %&% "loci-tables/DIAGRAM-t2d-loci.table.extended.alpha1.txt.gz"
  locus.tab.path <- res.dir %&%   "loci-tables/" %&% "DIAGRAM-t2d-loci.table.extended.alpha1.predFDR05.ImpG.txt.gz"
  
  locus.tab.df <- fread("cat " %&% locus.tab.path %&% " | zmore")
  fields <- names(locus.tab.df) 
  append.df <- extend_t2d_info()
  out.df <- c() 
  # Iterate over rows (nearby.gene) in append.df 
  for (i in 1:dim(append.df)[1]){
    row.df <- append.df[i,]
    gname <- row.df[1,11]
    # Iterate over fields 
    for (f in fields){
      if (f %in% names(row.df) == FALSE){
        mod <- gsub("DIAGRAM3_","",f); mod <- gsub("\\.","-",mod)
        if (mod %in% names(z.df)){
          z.val <- as.numeric(filter(z.df,Gene==gname) %>% select(one_of(mod)))
        } else{
          z.val <- NA 
        }
        row.df <- cbind(row.df,z.val)
      }
    }
    names(row.df) <- names(locus.tab.df)
    out.df <- rbind(out.df,row.df)
  }
  #append.locus.tab.df <- rbind(locus.tab.df,out.df)
  # write extended t2d locus file 
  print("Saving extended Loci information file")
  savename <- (res.dir %&% "loci-tables/DIAGRAM-t2d-loci.table.extended.alpha1.predFDR05.ImpG.APPENDED.txt.gz") # need to rename before re-running other R scripts
  write.df <- rbind(locus.tab.df,out.df)
  gz1 <- gzfile(savename, "w")
  write.table(write.df,file=gz1,quote=FALSE,sep="\t",row.names=FALSE)
  close(gz1) 
  return(write.df)
}

extend.df <- expand_loci_file()

```


