---
title: "GenomeWide-DIAGRAM.Rmd"
author: "Jason Torres"
date: "October 2, 2016"
output:
  pdf_document:
    toc: yes
  html_document:
    fig_caption: yes
    highlight: textmate
    number_sections: yes
    theme: journal
    toc: yes
---

# Setup

```{r setup, echo=FALSE}
library("dplyr")
library("ggplot2")
library("qqman")
library("gridExtra")
library("data.table")
"%&%" <- function(a,b) paste0(a,b) 

root.dir <- "/Users/jtorres/Google Drive/Projects/MetaXcan/MetaXcanT2D/"

rds.dir <- root.dir %&% "RDS.gtexV6p/diag3/" #" gera/
fig.dir <- root.dir %&% "figures.gtexV6p/" 
ggd.ref.df <- readRDS(paste0(rds.dir,"ggd.ref.df.RDS"))
load(paste0(rds.dir,"metaXcan.results.Rdata"))
meta.name = "DIAGRAM" # "GERA-T2D"

```

Prepare DIAGRAM GWAS 

```{r diagram_gwas,echo=FALSE}

diag.path <- "/Volumes/im-lab/nas40t2/jason/projects/MetaXcan/meta_files/" %&% "ImpGformat/merged_output_files/impG_DIAGRAM/gwas_file_v6p/" %&% "diag3.z.impG_0.8.gtexV6p.txt.gz"
diag.df <- fread("cat " %&% diag.path %&% " | zmore")
library("SNPlocs.Hsapiens.dbSNP144.GRCh37")
snps144 = SNPlocs.Hsapiens.dbSNP144.GRCh37
snps.Gpos = snpsById(snps144,ids=diag.df$SNP,ifnotfound="drop") 
rs.vec <- mcols(snps.Gpos)$RefSNP_id # rs ids 
chrom.vec <- as.character(seqnames(snps.Gpos)) # chromosomes 
pos.vec <- start(ranges(snps.Gpos)) # positions 
snp.df <- as.data.frame(cbind(chrom.vec,rs.vec,pos.vec),stringsAsFactors=FALSE)
names(snp.df) <- c("CHR","SNP","POS")
snp.df$CHR <- gsub("ch","",snp.df$CHR)
snp.df$CHR <- as.integer(snp.df$CHR)
snp.df$POS <- as.integer(snp.df$POS)
diagram.df <- full_join(diag.df,snp.df,by="SNP")
diagram.df$POS <- as.integer(diagram.df$POS)
diagram.df$CHR <- gsub("ch","",diagram.df$CHR)
diagram.df$CHR <- as.integer(diagram.df$CHR)
diagram.df$P <- 2*pnorm(-abs(diagram.df$Z),lower.tail=TRUE)
gwas.df <- select(diagram.df,one_of("SNP","CHR","POS","P"))
names(gwas.df)[3] <- "BP"

```



# Data Analyses on all Tissue-wide 

* p-value histogram 

```{r phist, echo=FALSE,fig.height=5,fig.width=6}
phist <- function(p.vec,meta.name,method.name,tiss.name){
  #title <- "MetaXcan Pvalues for " %&% meta.name %&% "\nModel: " %&% 
   # tiss.name %&% "      Method: " %&% method.name 
  title <- tiss.name
  hist(p.vec,breaks=100,main=title,xlab="p-values",cex.main=1)
}

```


* QQ-plot

```{r qqplot,echo=FALSE,fig.height=6,fig.width=6}
qqunif = 
function(p,BH=T,CI=T,...)
{
  nn = length(p)
  xx =  -log10((1:nn)/(nn+1))
  plot( xx,  -sort(log10(p)),
     xlab=expression(Expected~~-log[10](italic(p))),
        ylab=expression(Observed~~-log[10](italic(p))),
       cex.lab=1.4,mgp=c(2,1,0),
       ... )
  abline(0,1,col='gray')
  if(BH)
    {
      abline(-log10(0.05),1, col='red',lty=1)
      abline(-log10(0.10),1, col='orange',lty=2)
      abline(-log10(0.25),1, col='yellow',lty=3)
      #legend('bottomright', c("FDR = 0.05","FDR = 0.10","FDR = 0.25"),
      #       col=c('red','orange','yellow'),lty=1:3, cex=1)
      abline(h=-log10(0.05/nn)) ## bonferroni
    }
  if(CI)
  {
    ## create the confidence intervals
    c95 <- rep(0,nn)
    c05 <- rep(0,nn)
    ## the jth order statistic from a
    ## uniform(0,1) sample
    ## has a beta(j,n-j+1) distribution
    ## (Casella & Berger, 2002,
    ## 2nd edition, pg 230, Duxbury)
    ## this portion was posted by anonymous on
    ## http://gettinggeneticsdone.blogspot.com/2009/11/qq-plots-of-p-values-in-r-using-ggplot2.html
    
    for(i in 1:nn)
    {
      c95[i] <- qbeta(0.95,i,nn-i+1)
      c05[i] <- qbeta(0.05,i,nn-i+1)
    }

    lines(xx,-log10(c95),col='gray')
    lines(xx,-log10(c05),col='gray')
  }
}

qqpoints=
function(p,BH=T,...)
{
  nn = length(p)
  xx =  -log10((1:nn)/(nn+1))
  points( xx,  -sort(log10(p)), ... )
}

make_qq <- function(p.vec,meta.name,method.name,tiss.name){
  title <- tiss.name
  qqunif(p.vec,main=title,cex.main=1)

}

```


Alternative code for Manhattan Plot 

```{r alt_man, echo=FALSE}

manhat <- function(dataframe, title=NULL, max.y="max", suggestiveline=0, genomewideline=-log10(5e-8), size.x.labels=9, size.y.labels=10, annotate=F, SNPlist=NULL, rev=FALSE,col1,col2,trunc.y=NA,START=1,CAP=0,BY=1) {
  #BY is the interval for the Y-axis 
  
	#if (annotate & is.null(SNPlist)) stop("You requested annotation but provided no SNPlist!")
	d=dataframe
	
	#limit to only chrs 1-23?
	d=d[d$CHR %in% 1:23, ]
	#d=d[sample(1:dim(d)[1],5000,replace = FALSE),] # To speed up testing ONLY !!! 
  d <- arrange(d,CHR,BP) 
	
	if ("CHR" %in% names(d) & "BP" %in% names(d) & "P" %in% names(d) ) {
	
		d=na.omit(d)
		d=d[d$P>0 & d$P<=1, ]
		d$logp = -log10(d$P)
		
		d$pos=NA
		ticks=NULL
		lastbase=0
		
		#new 2010-05-10
		numchroms=length(unique(d$CHR))
		if (numchroms==1) {
			d$pos=d$BP
		} else {
		
			for (i in unique(d$CHR)) {
				if (i==1) {
					d[d$CHR==i, ]$pos=d[d$CHR==i, ]$BP
				}	else {
					lastbase=lastbase+tail(subset(d,CHR==i-1)$BP, 1)
					d[d$CHR==i, ]$pos=d[d$CHR==i, ]$BP+lastbase
				}
				ticks=c(ticks, d[d$CHR==i, ]$pos[floor(length(d[d$CHR==i, ]$pos)/2)+1])
			}
			ticklim=c(min(d$pos),max(d$pos))

		}
		
		mycols=rep(c("gray10","gray60"),max(d$CHR))
		
		if (max.y=="max") maxy=ceiling(max(d$logp)) else maxy=max.y
		if (maxy<8) maxy=8
		if (!is.na(trunc.y)) maxy=trunc.y

#		if (annotate) d.annotate=d[as.numeric(substr(d$SNP,3,100)) %in% SNPlist, ]
		
		if (numchroms==1) {
			plot=qplot(pos,logp,data=d,ylab=expression(-log[10](italic(p))), xlab=paste("Chromosome",unique(d$CHR),"position"))
		}	else {
			plot <- ggplot(data=d,aes(x=pos,y=logp,color=factor(CHR))) + 
			  geom_point() + ylab(expression(-log[10](italic(p)))) + 
			  scale_x_continuous(name="Chromosome", breaks=ticks, labels=(unique(d$CHR))) + 
			  scale_y_continuous(limits=c(0,maxy), breaks=seq(START,maxy+CAP,by=BY), labels=seq(START,maxy+CAP,by=BY)) +
			  theme(legend.position = "none") + theme(title=title) +
			  scale_colour_manual(values = rep(c(col1,col2),11)) + 
			  theme(panel.background=element_blank(), 
			         panel.grid.minor=element_blank(),
			         axis.text.x=element_text(size=size.x.labels, colour="black"), 
			         axis.text.y=element_text(size=size.y.labels, colour="black")
			         )
			if (rev==TRUE){
			  plot <- plot + scale_y_reverse(breaks=seq(START,maxy+CAP,by=BY),labels=seq(START,maxy+CAP,by=BY),limits=c(maxy,0)) 
			}
		}
		if (suggestiveline) plot=plot+geom_hline(yintercept=suggestiveline,colour="gray", alpha=I(1/3))
		if (genomewideline) plot=plot+geom_hline(yintercept=genomewideline,colour="red")
		plot
	}	else {
		stop("Make sure your data frame contains columns CHR, BP, and P")
	}
}

```


* Manhattan Plot 

```{r manhat,echo=FALSE,fig.height=5,fig.width=8}

build_manhat_vecs <- function(p.df){
  print("Getting Gene info...")
  gene.vec <- p.df$Gene
  chrom.vec <- c()
  pos.vec <- c()
  #gene <- "AL157931.1"
  for (gene in gene.vec){
    temp.df <- filter(ggd.ref.df,symbol==gene) %>% dplyr::select(chr,start)
    c <- as.integer(temp.df[1,1]); s <- as.numeric(temp.df[1,2])
    chrom.vec <- append(chrom.vec,c); pos.vec <- append(pos.vec,s)
  }
  return(list(gene.vec,chrom.vec,pos.vec))
}

make_manhat_df <- function(p.df){
  print("Building data frame for Manhattan plot...")
  vec.list <- build_manhat_vecs(p.df)
  block.df <- as.data.frame(cbind(vec.list[[1]],vec.list[[2]],vec.list[[3]]),stringsAsFactors=FALSE)
  names(block.df) <- c("SNP","CHR","BP")
  block.df$CHR <- as.integer(block.df$CHR)
  block.df$BP <- as.numeric(block.df$BP)
  out.df <- c()
  for (c in 2:dim(p.df)[2]){
    P <- p.df[,c]
    chunk.df <- cbind(block.df,P)
    out.df <- rbind(out.df,chunk.df)
  }
  return(out.df)
}

manhat_plot <- function(p.df,meta.name,method.name,tiss.name){
  title <- tiss.name
  man.df <- make_manhat_df(p.df)
  bonfer <- 0.05 / length(na.omit(as.numeric(as.matrix(dplyr::select(p.df,contains(tiss.name))))))
  if (tiss.name == "All"){
    bonfer <- 0.05 / length(na.omit(as.numeric(as.matrix(select(p.df,contains("TW_"))))))
  }
  multcorrect <- 0.05 / length(na.omit(as.numeric(as.matrix(select(p.df,contains("TW_"))))))
  manhat(dataframe=na.omit(man.df), title=NULL, max.y="max", 
         suggestiveline=0, genomewideline=-log(multcorrect,base=10), size.x.labels=9, size.y.labels=10,
         rev=FALSE,col1="black",col2="dodgerblue1")
}

```

```{r test}

plt1 <- manhat(dataframe=na.omit(man.df), title=NULL, max.y="max", 
         suggestiveline=0, genomewideline=-log(multcorrect,base=10), size.x.labels=9, size.y.labels=10,
         rev=FALSE,col1="black",col2="dodgerblue3",START=0,CAP=1,BY=2)
plt2 <- manhat(dataframe=na.omit(gwas.df), title=NULL, max.y="max", 
         suggestiveline=0, size.x.labels=9, size.y.labels=10,
         rev=TRUE,col1="black",col2="gray",trunc.y=50,START=0,CAP=0,BY=5)

grid.arrange(plt1,plt2,nrow=2)

```




# Save PDF Figures 

```{r save_func, echo=FALSE}

save_figures <- function(alpha,model){
  mod.name <- gsub("TW_","",model)
  if (alpha==0.5) {
    method.name <- "Elastic Net"
    rds.name <- "sig.gene.TW.a05.df.RDS"
    p.df <- p.EN.df
  } else if (alpha==1) {
    method.name <- "Lasso"
    rds.name <- "sig.gene.TW.a1.df.RDS"
    p.df <- p.L.df
  }else {
    print("alpha must be 1 or 0.5")
    break
  }
  sig.df <- readRDS(paste0(rds.dir,rds.name))
  p.df <- cbind(dplyr::select(p.df,contains("Gene")),#dplyr::select(p.df,contains("CrossTissue")),
                dplyr::select(p.df,contains("TW_")))
  if (model!="All"){
    p.df <- dplyr::select(p.df,one_of(c("Gene",model)))
  }
  p.vec <- na.omit(as.numeric(as.matrix(p.df)[,2:dim(p.df)[2]]))
  savename1 <- fig.dir  %&% meta.name %&% "." %&% 
                alpha %&% "." %&% mod.name %&% ".phist.pdf"
  savename2 <- fig.dir  %&% meta.name %&% "." %&% 
                alpha %&% "." %&% mod.name %&% ".qq.pdf" 
  savename3 <- fig.dir  %&% meta.name %&% "." %&% 
                alpha %&% "." %&% mod.name %&% ".manhat.pdf"
  print("Saving figures for " %&% meta.name %&% "." %&% alpha %&% "." %&% mod.name )
  #print("histogram")
  #jpng(savename1,height=5,width=6,units="in",res=2000)
  #phist(p.vec,meta.name,method.name,mod.name)
  #dev.off() 
  print("qq-plot")
  pdf(savename2,height=5,width=6)  
  make_qq(p.vec,meta.name,method.name,mod.name)
  dev.off()
  print("manhattan plot")
  pdf(savename3,height=4,width=8)
  manhat_plot(p.df,meta.name,method.name,mod.name)
  dev.off()
}


```

**Elastic Net Models**

```{r save_runs_lasso, echo=FALSE}

save_figures(0.5,"All")
#save_figures(1,"TW_Adipose-Subcutaneous")
#save_figures(1,"TW_Liver")
#save_figures(1,"TW_Pancreas")
#save_figures(1,"TW_Muscle-Skeletal")
```


# Double check signficant associations 

```{r check_assoc_numbers, echo=FALSE}

library(qvalue)
p.vec <- na.omit(as.numeric(as.matrix(select(p.EN.df,contains("TW_")))))
bonfer.mult <- 0.05 / length(p.vec)
num.bonfer.assoc <- table(p.vec <= (0.05/length(p.vec)))
q.vec <- qvalue(p.vec)$qvalue
num.fdr05.assoc <- table(q.vec <= 0.05)

# Determine number of unique genes 

gene.vec <- p.EN.df$Gene
mod.vec <- names(select(p.EN.df,contains("TW_")))
mat <- as.matrix(select(p.EN.df,contains("TW_")))
named.p.vec <- c() # track genes 
named.p.vec2 <- c() # track models 
for (c in 1:dim(mat)[2]){
  mod <- mod.vec[c]
  temp.vec <- as.numeric(mat[,c])
  names(temp.vec) <- gene.vec
  named.p.vec <- append(named.p.vec,temp.vec)
  names(temp.vec) <- paste0(mod,":",gene.vec)
  named.p.vec2 <- append(named.p.vec2,temp.vec)
}
named.p.vec <- na.omit(named.p.vec2)
named.q.vec <- qvalue(named.p.vec2)$qvalue
fdr05.vec <- named.q.vec[named.q.vec <= 0.05]
fdr.genes <- na.omit(unique(names(fdr05.vec)))

#saveRDS(fdr.genes,file=rds.dir%&%"genome.wide.FDR05.associations.RDS")
# Need to manually add these genes to the supplementary table 
#[1] "TW_Cells_EBV-transformed_lymphocytes:CCNE2" = T2D.Gene 
#[2] "TW_Lung:HCG27" =  T2D.Gene                   
#[3] "TW_Muscle_Skeletal:TCP11L1"  = "Unknown"               
#[4] "TW_Prostate:AP3S2"= T2D.Gene                        
#[5] "TW_Skin_Not_Sun_Exposed_Suprapubic:FAM229B" = "Unknown" # Be sure to factor this
#[6] "TW_Whole_Blood_DGN:KLHDC5"=KLHL42=T2D.Gene

# Evaluation data frame 
# Using sig df or T2Dcore.df 

fdr.df <- dplyr::select(readRDS(rds.dir %&% "FDR05.sig.df.RDS"),
                        one_of("gene","model","type"))
# add missing values 
fdr.df <- rbind(fdr.df,c("CCNE2","Cells-EBV-transformed-lymphocytes","T2D.Gene"))
fdr.df <- rbind(fdr.df,c("HCG27","Lung","T2D.Gene"))
fdr.df <- rbind(fdr.df,c("TCP11L1","Muscle-Skeletal","Unknown"))
fdr.df <- rbind(fdr.df,c("AP3S2","Prostate","T2D.Gene"))
fdr.df <- rbind(fdr.df,c("FAM229B","Skin-Not-Sun-Exposed-Suprapubic","Unknown"))
# correct error 
fdr.df[116,] <- c("KLHL42","Artery-Tibial","T2D.Gene")
saveRDS(fdr.df,file=rds.dir%&%"fdr05.processed.df.RDS")
# Excluding KLHDC5, the number of unique genes is 90 

# The number of fdr05 genes within T2D loci (including T2D.Gene)
length(unique(filter(fdr.df,type!="Unknown")$gene)) # 52 
# The number of fdr05 genes outside T2D loci 
length(unique(filter(fdr.df,type=="Unknown")$gene)) # 38 
# The number of fdr05 genes that are T2D genes 
length(unique(filter(fdr.df,type=="T2D.Gene")$gene)) # 21 
# total number of NOVEL associations 
90 - 21 # 69 

# Recall of the 20 Bonferonni Genes 
# 17 fall within T2D loci with 
# 10 reported genes and 
# 7 novel loci genes 

```

```{r quick_look_replicatedFDR05,echo=FALSE}

fdr.df <- readRDS(rds.dir %&% "FDR05.sig.df.RDS")
df <- filter(fdr.df,type=="Unknown",replicated==TRUE)

```



# QQ plot evaluation of known T2D loci genes 


```{r qqeval, echo=FALSE}

myqq <- function (pvector, ymax, xmax,...) 
{
    if (!is.numeric(pvector)) 
        stop("Input must be numeric.")
    pvector <- pvector[!is.na(pvector) & !is.nan(pvector) & !is.null(pvector) & 
        is.finite(pvector) & pvector < 1 & pvector > 0]
    o = -log10(sort(pvector, decreasing = FALSE))
    e = -log10(ppoints(length(pvector)))
    plot(x=e,y=o, pch = 20, xlim = c(0, xmax), ylim = c(0, 
        ymax), xlab = expression(Expected ~ ~-log[10](italic(p))), 
        ylab = expression(Observed ~ ~-log[10](italic(p))))
    abline(0, 1, col = "red")
}


rds.dir <-root.dir %&% "RDS.gtexV6p/"
loci.df <- readRDS(rds.dir%&%"loci.df.RDS")
t2d.loci.genes <- unique(loci.df$nearby.gene)
p.vec.total <- na.omit(as.numeric(as.matrix(select(p.EN.df,contains("TW_")))))
p.vec.t2dloci <- na.omit(as.numeric(as.matrix(filter(p.EN.df,Gene%in%t2d.loci.genes) %>% select(contains("TW_")))))
p.vec.unk <- na.omit(as.numeric(as.matrix(filter(p.EN.df,!(Gene%in%t2d.loci.genes)) %>% select(contains("TW_")))))
par(mfrow=c(1,3))
qq(p.vec.total,ylim=c(0,25),main="Total Genes")
qq(p.vec.t2dloci,ylim=c(0,25),main=("T2D Loci Genes"))
qq(p.vec.unk,ylim=c(0,25),main="Non-T2D Loci Genes")

```

