---
title: "Results-Summary.Rmd"
author: "Jason Torres"
date: "August 8, 2016"
output: pdf_document
---

Here, I summarize key results to discuss in the manuscript: 

# Results at T2D loci 

Referring to the top 1000 SNPs from the DIAGRAM trans-ethnic meta-analysis of GWAS (accessed from AMP-T2D, 2016), considered the set of 89 genes that were annotated to these genes. For each annotated gene (reported T2D gene implicated by proximity to GWAS signal), considered the set of GWAS SNPs annotated to this gene, noted the most distal SNPs on the terminal ends w/ respect to each other. Then expanded the window by 1Mb (i.e. 89 windows). Then delineated non-overlapping windows as the boundaries of the gene loci windows overlapped in several instances. This resulted in 57 unique, non-overlapping windows. We consider all genome-wide and locus-wide (w/ respect to tests performed within the non-overlapping window) gene-level associations. 

```{r t2d_loci, echo=FALSE}

"%&%" <- function(a,b) paste0(a,b)

library("data.table")
library("dplyr")

file.dir <- "/Users/jtorres/Google Drive/Projects/MetaXcan/MetaXcanT2D/"
rds.dir <- file.dir %&% "RDS/"
df1 <- readRDS(rds.dir %&% "msig.loci.df.RDS")
rep.genes <- unique(readRDS(rds.dir %&% "loci.df.RDS")$reported.gene)

```


### Genome-wide significant vs locus-wide significant 

There are **`r dim(df1)[1]`** non-overlapping T2D loci, of which:

**`r length(na.omit(df1$mtxn.sig.gw.genes))`** have genome-wide significant (across all models) MetaXcan associations 

**`r length(na.omit(df1$mtxn.sig.lw.genes))`** have locus-wide significant (across all models) MetaXcan associations 

### Locus profile 

There are **`r dim(filter(df1,group==1))[1]`** loci where _at least one_ reported gene yielded a genome-wide signficant association

There are **`r dim(filter(df1,group==2))[1]`** loci where _at least one_ reported gene yielded a locus-wide **but not genome-wide signficant** association 

There are **`r dim(filter(df1,group==3))[1]`** loci where an unreported T2D gene yielded a genome-wide association but no reported gene yielded an association an any signifcance level

There are **`r dim(filter(df1,group==4))[1]`** loci where an unreported T2D gene yielded a locus-wide association but no reported gene yielded an association an any signifcance level

There are **`r dim(df1[is.na(df1$group),])[1]`** loci where there are **no significant associations** 

## Genome-wide Signifcant genes and models 

```{r ebi_nhgri, echo=FALSE}

ebi.dir <- "/Volumes/im-lab/nas40t2/jason/projects/ReferenceData/NHGRI/"
ebi.file <- ebi.dir %&% "nhgri_ebi-trait-genes.txt.gz"

ebi.df <- fread("cat " %&% ebi.file %&% " | zmore")
names(ebi.df)<-c("trait","genes")
ebi.df <- filter(ebi.df,trait=="Type_2_diabetes")
ref.vec <- c()
for (r in ebi.df$genes){
  vec <- strsplit(r,",_")[[1]]
  ref.vec <- append(ref.vec,vec)
}
ref.vec <- unique(ref.vec)
rep.genes <- unique(append(rep.genes,ref.vec))
```




```{r sig_gene, echo=FALSE}

df1.sig <- df1[!is.na(df1$group),]
gw.genes <- df1.sig$mtxn.sig.gw.genes
gw.mods <- df1.sig$mtxn.sig.gw.models
gw.genes <- na.omit(as.character(gw.genes))
gw.mods <- na.omit(as.character(gw.mods))
gvec <- c()
for (g in gw.genes){
  vec <- strsplit(g,",")[[1]]
  gvec <- append(gvec,vec)
}
mvec <- c()
for (m in gw.mods){
  vec <- strsplit(m,",")[[1]]
  mvec <- append(mvec,vec)
}

summary(mvec)
```

There are **`r length(gvec)`** _genome-wide_ signficant gene-level associations

There are **`r length(unique(gvec))`** _unique_ genome-wide significant genes:

```{r echo=FALSE}
gvec1 <- gvec
gvec <- unique(gvec)
gvec

```

Of these genes, **`r length(gvec[gvec %in% rep.genes])`** genes are **reported** T2D genes: 

```{r}

gvec[gvec %in% rep.genes]

```


Of these genes, **`r length(gvec[!(gvec %in% rep.genes)])`** genes are **novel** T2D genes: 

```{r}

gvec[!(gvec %in% rep.genes)]

```


Here are the frequency counts for the 10 most frequent genes: 

```{r echo=FALSE}

tab <- sort(table(gvec1),decreasing=TRUE)
head(tab,10)

```


These correspond to **`r length(unique(mvec))`** unique models

Here are the frequency counts for the 10 most frequent models : 

```{r echo=FALSE}

tab <- sort(table(mvec),decreasing=TRUE)
head(tab,10)

```


## Locus-wide Signifcant genes and models 

```{r sig_gene_lw, echo=FALSE}

df1.sig <- df1[!is.na(df1$group),]
lw.genes <- df1.sig$mtxn.sig.lw.genes
lw.mods <- df1.sig$mtxn.sig.lw.models
lw.genes <- na.omit(as.character(lw.genes))
lw.mods <- na.omit(as.character(lw.mods))
gvec <- c()
for (g in lw.genes){
  vec <- strsplit(g,",")[[1]]
  gvec <- append(gvec,vec)
}
mvec <- c()
for (m in lw.mods){
  vec <- strsplit(m,",")[[1]]
  mvec <- append(mvec,vec)
}

```

There are **`r length(gvec)`** _locus-wide_ signficant gene-level associations

There are **`r length(unique(gvec))`** _unique_ locus-wide significant genes:

```{r echo=FALSE}

gvec1 <- gvec 
gvec <- unique(gvec)
gvec

```

Of these genes, **`r length(gvec[gvec %in% rep.genes])`** genes are **reported** T2D genes: 

```{r}

gvec[gvec %in% rep.genes]

```


Of these genes, **`r length(gvec[!(gvec %in% rep.genes)])`** genes are **novel** T2D genes: 

```{r}

gvec[!(gvec %in% rep.genes)]

```



Here are the frequency counts for the 10 most frequent genes: 

```{r echo=FALSE}

tab <- sort(table(gvec1),decreasing=TRUE)
head(tab,10)

```

These correspond to **`r length(unique(mvec))`** unique models

Here are the frequency counts for the 10 most frequent models : 

```{r echo=FALSE}

tab <- sort(table(mvec),decreasing=TRUE)
head(tab,10)

```


# Genome-wide Signifant associations 

```{r gw, echo=FALSE}

df2 <- readRDS(rds.dir %&% "T2Dcore.df.RDS")
df2 <- filter(df2, diag.mult.correct==TRUE)

```


There are **`r dim(df2)[1]`** GENOME-WIDE associations

This corresponds to **`r length(unique(df2$gene))`** unique genes 

Of these genes, **`r length(unique(df2$gene[df2$gene %in% rep.genes]))`** unique genes are reprted T2D genes and **`r length(unique(df2$gene[!(df2$gene %in% rep.genes)]))`** are **novel** genes 

Reported genes: 

```{r}

unique(df2$gene[df2$gene %in% rep.genes])

```

Novel genes: 

```{r}

unique(df2$gene[!(df2$gene %in% rep.genes)])

```

Genes mapping to "Unknown" T2D Loci: 

```{r}

unique(filter(df2,type=="Unknown")$gene)

```

## Replicated in GERA at _p-value_ < 0.05

```{r replicated, echo=FALSE}

replic.df <- filter(df2,replicated==TRUE)

```

There are **`r dim(replic.df)[1]`** associations that replicate in the GERA-T2D study

Corresponding to **`r unique(replic.df$gene)`** unique genes 

```{r echo=FALSE}

replic.genes <- unique(replic.df$gene)

```

Of these replicated genes, **`r length(unique(replic.genes[replic.genes %in% rep.genes]))`** genes are reported T2D genes:  

```{r echo=FALSE}

unique(replic.genes[replic.genes %in% rep.genes])

```


Of these replicated genes, **`r length(unique(replic.genes[!(replic.genes %in% rep.genes)]))`** genes are **novel** genes:  

```{r echo=FALSE}

unique(replic.genes[!(replic.genes %in% rep.genes)])

```

