---
title: "Results-Summary.Rmd"
author: "Jason Torres"
date: "September 19, 2016"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Here, I summarize key results to discuss in the manuscript: 


# Setup

GTEx V6p; HapMap SNPs; Prediction q-value <= 0.05 

```{r libraries}

"%&%" <- function(a,b) paste0(a,b)

library("data.table")
library("dplyr")

file.dir <- "/Users/jtorres/Google Drive/Projects/MetaXcan/MetaXcanT2D/"
rds.dir <- file.dir %&% "RDS.gtexV6p/"
ggd.df <- readRDS(rds.dir%&%"ggd.ref.df.RDS")
rep.genes <- unique(readRDS(rds.dir %&% "loci.df.RDS")$reported.gene)
loci.df <- readRDS(rds.dir%&%"loci.df.RDS")
```


```{r ebi_nhgri, echo=FALSE}

ebi.dir <- "/Volumes/im-lab/nas40t2/jason/projects/ReferenceData/NHGRI/"
ebi.file <- ebi.dir %&% "nhgri_ebi-trait-genes.txt.gz"

ebi.df <- fread("cat " %&% ebi.file %&% " | zmore")
names(ebi.df)<-c("trait","genes")
ebi.df <- filter(ebi.df,trait=="Type_2_diabetes")
ref.vec <- c()
for (r in ebi.df$genes){
  vec <- strsplit(r,",_")[[1]]
  ref.vec <- append(ref.vec,vec)
}
ref.vec <- unique(ref.vec)
rep.genes <- unique(append(rep.genes,ref.vec))

```



# Genome-wide Signifant associations 

## Bonferonni-adjusted for the total number of tests performed across all tissue models 

```{r gw, echo=FALSE}

df <- readRDS(rds.dir %&% "T2Dcore.df.RDS")
df <- filter(df, diag.mult.correct==TRUE)
```
 
There are **`r dim(df)[1]`** GENOME-WIDE associations

This corresponds to **`r length(unique(df$gene))`** unique genes 

Of these genes, **`r length(unique(df$gene[df$gene %in% rep.genes]))`** genes are nearest to GWAS associations ($p \leq 5 \times 10^{-6}$) and **`r length(unique(df$gene[!(df$gene %in% rep.genes)]))`** are not _nearest to GWAS associations_ genes 

_Nearest to GWAS hits_ genes: 

```{r}

unique(df$gene[df$gene %in% rep.genes])

```

_Not Nearest_ genes: 

```{r}

unique(df$gene[!(df$gene %in% rep.genes)])

```

Genes mapping to "Unknown" T2D Loci: 

```{r}

unique(filter(df,type=="Unknown")$gene)[!(filter(df,type=="Unknown")$gene %in% unique(loci.df$nearby.gene))]

```


### GERA Replicated at $p \leq 0.05$ 

```{r replicated, echo=FALSE}

replic.df <- filter(df,replicated==TRUE)
replic.df$diag.z.score <- as.numeric(replic.df$diag.z.score)
replic.df$gera.z.score <- as.numeric(replic.df$gera.z.score)
sign.consistent <- sign(replic.df$diag.z.score) == sign(replic.df$gera.z.score)
replic.df <- cbind(replic.df,sign.consistent)
```

There are **`r dim(replic.df)[1]`** associations that replicate in the GERA-T2D study

```{r}

replic.df <- filter(replic.df,sign.consistent==TRUE)

```

There are **`r dim(replic.df)[1]`** associations that replicate in the GERA-T2D study in the same direction as DIAGRAM


Corresponding to **`r length(unique(replic.df$gene))`** unique genes 

```{r echo=FALSE}

(replic.genes <- unique(replic.df$gene))

```

Of these replicated genes, **`r length(unique(replic.genes[replic.genes %in% rep.genes]))`** genes are reported T2D genes:  

```{r echo=FALSE}

unique(replic.genes[replic.genes %in% rep.genes])

```


Of these replicated genes, **`r length(unique(replic.genes[!(replic.genes %in% rep.genes)]))`** genes are **novel** genes:  

```{r echo=FALSE}

unique(replic.genes[!(replic.genes %in% rep.genes)])

```


### GSEA of MetaXcan-significant gene sets 

```{r gsea, echo=FALSE}

gsea.df <- readRDS(rds.dir%&%"gsea.p05.df.RDS")
gsea.t2d.df <- readRDS(rds.dir%&%"gsea.p05.t2d.df.RDS")
gsea.novel.df <- readRDS(rds.dir%&%"gsea.p05.novel.df.RDS")

```

There are **`r sum(gsea.df$bonfer)`** GO BP pathways that are Bonferonni-significant

GSEA for all genes in set: 

```{r}
head(dplyr::select(gsea.df,one_of("over_represented_pvalue","term")))
```

GSEA for reported T2D genes in set: 

```{r}
head(dplyr::select(gsea.t2d.df,one_of("over_represented_pvalue","term")))
```

GSEA for novel genes in set: 

```{r}
head(dplyr::select(gsea.novel.df,one_of("over_represented_pvalue","term")))
```


# Analyses at Loci harboring T2D GWAS associations 

```{r}

df1 <- readRDS(rds.dir %&% "msig.loci.df.RDS")

```

### Genome-wide significant vs locus-wide significant 

There are **`r dim(df1)[1]`** non-overlapping T2D loci, of which:

**`r length(na.omit(df1$mtxn.sig.gw.genes))`** have genome-wide significant (across all models) MetaXcan associations 

**`r length(na.omit(df1$mtxn.sig.lw.genes))`** have locus-wide significant (across all models) MetaXcan associations 


### Locus profile 

There are **`r dim(filter(df1,group==1))[1]`** loci where _at least one_ reported gene yielded a genome-wide signficant association

There are **`r dim(filter(df1,group==2))[1]`** loci where _at least one_ reported gene yielded a locus-wide **but not genome-wide signficant** association 

There are **`r dim(filter(df1,group==3))[1]`** loci where an unreported T2D gene yielded a genome-wide association but no reported gene yielded an association an any signifcance level

There are **`r dim(filter(df1,group==4))[1]`** loci where an unreported T2D gene yielded a locus-wide association but no reported gene yielded an association an any signifcance level

There are **`r dim(df1[is.na(df1$group),])[1]`** loci where there are **no significant associations** 

## Genome-wide Signifcant genes and models 


```{r sig_gene, echo=FALSE}

df1.sig <- df1[!is.na(df1$group),]
gw.genes <- df1.sig$mtxn.sig.gw.genes
gw.mods <- df1.sig$mtxn.sig.gw.models
gw.genes <- na.omit(as.character(gw.genes))
gw.mods <- na.omit(as.character(gw.mods))
gvec <- c()
for (g in gw.genes){
  vec <- strsplit(g,",")[[1]]
  gvec <- append(gvec,vec)
}
mvec <- c()
for (m in gw.mods){
  vec <- strsplit(m,",")[[1]]
  mvec <- append(mvec,vec)
}

summary(mvec)
```

There are **`r length(gvec)`** _genome-wide_ signficant gene-level associations

There are **`r length(unique(gvec))`** _unique_ genome-wide significant genes:

```{r echo=FALSE}
gvec1 <- gvec
gvec <- unique(gvec)
gvec

```

Of these genes, **`r length(gvec[gvec %in% rep.genes])`** genes are **reported** T2D genes: 

```{r}

gvec[gvec %in% rep.genes]

```


Of these genes, **`r length(gvec[!(gvec %in% rep.genes)])`** genes are **novel** T2D genes: 

```{r}

gvec[!(gvec %in% rep.genes)]

```


Here are the frequency counts for the 10 most frequent genes: 

```{r echo=FALSE}

tab <- sort(table(gvec1),decreasing=TRUE)
head(tab,10)

```


These correspond to **`r length(unique(mvec))`** unique models

Here are the frequency counts for the 10 most frequent models : 

```{r echo=FALSE}

tab <- sort(table(mvec),decreasing=TRUE)
head(tab,10)

```


## Locus-wide Signifcant genes and models 

```{r sig_gene_lw, echo=FALSE}

df1.sig <- df1[!is.na(df1$group),]
lw.genes <- df1.sig$mtxn.sig.lw.genes
lw.mods <- df1.sig$mtxn.sig.lw.models
lw.genes <- na.omit(as.character(lw.genes))
lw.mods <- na.omit(as.character(lw.mods))
gvec <- c()
for (g in lw.genes){
  vec <- strsplit(g,",")[[1]]
  gvec <- append(gvec,vec)
}
mvec <- c()
for (m in lw.mods){
  vec <- strsplit(m,",")[[1]]
  mvec <- append(mvec,vec)
}

```

There are **`r length(gvec)`** _locus-wide_ signficant gene-level associations

There are **`r length(unique(gvec))`** _unique_ locus-wide significant genes:

```{r echo=FALSE}

gvec1 <- gvec 
gvec <- unique(gvec)
gvec

```

Of these genes, **`r length(gvec[gvec %in% rep.genes])`** genes are **reported** T2D genes: 

```{r}

gvec[gvec %in% rep.genes]

```


Of these genes, **`r length(gvec[!(gvec %in% rep.genes)])`** genes are **novel** T2D genes: 

```{r}

gvec[!(gvec %in% rep.genes)]

```



Here are the frequency counts for the 10 most frequent genes: 

```{r echo=FALSE}

tab <- sort(table(gvec1),decreasing=TRUE)
head(tab,10)

```

These correspond to **`r length(unique(mvec))`** unique models

Here are the frequency counts for the 10 most frequent models : 

```{r echo=FALSE}

tab <- sort(table(mvec),decreasing=TRUE)
head(tab,10)

```


### GERA replication of Locus-wide significant associations at $p \leq 0.05$ 

```{r}
#**`r `**
lw.df <- readRDS(file=rds.dir%&%"lw.rep.df.RDS") 

```


There are **`r dim(lw.df)[1]`** locus-wide significant associations in DIAGRAM 

Of these, **`r dim(filter(lw.df,gw.p05==TRUE))[1]`** replicate in GERA-T2D Study 

Of these, **`r dim(filter(lw.df,gw.p05==TRUE,same.direct==TRUE))[1]`** replicate in GERA-T2D Study in the **same direction**

```{r}

lw.df <- arrange(lw.df,desc(abs(diag.z)))

print.df <- filter(lw.df,gw.p05==TRUE,same.direct==TRUE) %>% dplyr::select(one_of("gene","model","diag.z","gera.z"))

print.genes <- unique(print.df$gene)

```


This corresponds to **`r length(print.genes)`** unique genes 

```{r}

print.genes

```

