---
title: "PredictorSNPs-manhattan.Rmd"
author: "Jason Torres"
date: "October 10, 2016"
output:
  pdf_document:
    toc: yes
  html_document:
    fig_caption: yes
    highlight: textmate
    number_sections: yes
    theme: journal
    toc: yes
---

```{r setup}

library("LDheatmap")
library("data.table")
library("dplyr")
library("ggplot2")
library("gridExtra")
library("RColorBrewer")
library("ggrepel")

"%&%" <- function(a,b) paste0(a,b)

root.dir <- "/Users/jtorres/Google Drive/Projects/MetaXcan/MetaXcanT2D/"
rds.dir <- root.dir %&% "RDS.gtexV6p/"
tarbell.dir <- "/Volumes/im-lab/nas40t2/jason/projects/MetaXcan/results/predictors/gtexV6p/gtex_files/"

ggd.df <- readRDS(rds.dir%&%"ggd.ref.df.RDS")

save.dir <- "/Users/jtorres/Google Drive/Projects/MetaXcan/MetaXcanT2D/txt.gtexV6p/snp_lists/"

beta.dir <- "/Volumes/im-lab/nas40t2/jason/projects/MetaXcan/db_files/gtex_v6p/alpha_0.5/allBetas/"
res.dir <- "/Volumes/im-lab/nas40t2/jason/projects/MetaXcan/results/DIAGRAM_ImpG_0.8_gtexV6p/alpha_0.5/"

```


```{r make_snpref_df,echo=FALSE}

library("SNPlocs.Hsapiens.dbSNP144.GRCh37")
snps144 = SNPlocs.Hsapiens.dbSNP144.GRCh37
snps.Gpos = snpsById(snps144,ids=diag.df$SNP,ifnotfound="drop") 
rs.vec <- mcols(snps.Gpos)$RefSNP_id # rs ids 
chrom.vec <- as.character(seqnames(snps.Gpos)) # chromosomes 
pos.vec <- start(ranges(snps.Gpos)) # positions 
snp.df <- as.data.frame(cbind(chrom.vec,rs.vec,pos.vec),stringsAsFactors=FALSE)

```


```{r functions,echo=FALSE}

get_snp_pos <- function(snp.vec){
  pos.vec <- c()
  for (snp in snp.vec){
    pos <- filter(snp.df,SNP==snp)$POS
    if (length(pos)==0){
      pos.vec <- append(pos.vec,NA)  
    } else{
      pos.vec <- append(pos.vec,pos)  
    }
  }
  return(pos.vec)
}

get_model_snps <- function(mygene,model){
  print("Looking up model SNPs: " %&% model %&% " ; " %&% mygene)
  file.path <- beta.dir %&% model %&% ".allBetas.txt.gz"
  temp.df <- fread("cat " %&% file.path %&% " | zmore")
  ensid <- filter(ggd.df,symbol==mygene)$ensgene
  suppressWarnings(out.df <- filter(temp.df,grepl(ensid, temp.df$gene)))
  if (dim(out.df)[1]==0 | length(ensid) > 1){
    t.df <- fread("cat " %&% res.dir  %&% "TW_" %&% model %&% ".zscores.csv")
    ensid <- filter(t.df,gene_name==mygene)$gene
    out.df <- filter(temp.df,grepl(ensid, temp.df$gene))
  }
  return(out.df)
}


```


# TCF7L2 analysis 

```{r ld_plots, echo=FALSE}

source("modified.hm.R")

assign_color <- function(snp.vec,snpset1,snpset2,col1,col2,colboth){
  col.vec <- c()
  for (snp in snp.vec){
    if (snp %in% snpset1 & snp %in% snpset2){
      col.vec <- append(col.vec,colboth)
    } else if (snp %in% snpset1){
      col.vec <- append(col.vec,col1)
    } else if (snp %in% snpset2){
      col.vec <- append(col.vec,col2)
    } else{
      print("WTF")
    }
  }
  return(col.vec)
}

tcf7l2.AA.snps <- get_model_snps("TCF7L2","Artery_Aorta")
tcf7l2.AA.snps <- unique(tcf7l2.AA.snps$rsid)
write.table(tcf7l2.AA.snps,file=save.dir%&%"tcf7l2.AA.snps.txt",quote=FALSE,sep="\n",row.names=FALSE,col.names=FALSE)

dclre1a.CTF.snps <- get_model_snps("DCLRE1A","Cells_Transformed_fibroblasts")
dclre1a.CTF.snps <- unique(dclre1a.CTF.snps$rsid)
write.table(dclre1a.CTF.snps,file=save.dir%&%"dclre1a.CTF.snps.txt",quote=FALSE,sep="\n",row.names=FALSE,col.names=FALSE)

dclre1a.Thy.snps <- get_model_snps("DCLRE1A","Thyroid")
dclre1a.Thy.snps <- unique(dclre1a.Thy.snps$rsid)
write.table(dclre1a.Thy.snps,file=save.dir%&%"dclre1a.Thy.snps.txt",quote=FALSE,sep="\n",row.names=FALSE,col.names=FALSE)
key.snp <- "rs12243326"
write.table(unique(c(tcf7l2.AA.snps,key.snp))
            ,file=save.dir%&%"set5.snps.txt",quote=FALSE,sep="\n",row.names=FALSE,col.names=FALSE)

ld.df <- dplyr::select(fread(save.dir%&%"r2.set5.txt"),-contains("RS_number"))
snps <- names(ld.df)
pos.vec <- get_snp_pos(snps)
na.snps <- snps[is.na(pos.vec)]
mat <- as.matrix(ld.df)
row.names(mat) <- snps
if (length(na.snps)>=1){
  remove.index <- match(na.snps,row.names(mat))
  mat <- mat[-remove.index,-remove.index]
}
pos.vec <- pos.vec[!is.na(pos.vec)]
snps <- row.names(mat)
  
Lab.palette <- colorRampPalette(c("dodgerblue4", "ghostwhite"),space = "Lab")

mycolors <- assign_color(snps,tcf7l2.AA.snps,key.snp,
                         col1="blue",col2="black",colboth="red")
keep.index <- !grepl("blue",mycolors)

hm <- LDheatmap_jmt(mat,color=Lab.palette(100),
            genetic.distances=pos.vec,distances="physical",add.map=TRUE,flip=TRUE,
            SNP.name=snps[keep.index],geneMapLocation=0.20,name = "myLDgrob",
            mycol=mycolors[keep.index],mycex=0.4)

hm <- LDheatmap_jmt(mat,color=Lab.palette(100),
            genetic.distances=pos.vec,distances="physical",add.map=TRUE,flip=TRUE,
            geneMapLocation=0.15,name = "myLDgrob",
            mycol=mycolors[keep.index],mycex=0.4)

#hm <- LDheatmap(mat,color=Lab.palette(100),
#            genetic.distances=pos.vec,distances="physical",add.map=TRUE,flip=TRUE,
#            SNP.name=snps,geneMapLocation=0.25,name = "myLDgrob")

#LDheatmap.addGenes(hm,chromosome="chr10",splice_variants=FALSE,non_coding=FALSE,genome="hg19")
# key start: chr10:114,238,129-115,203,305
# start = 114238129
# end = 116369485
# chr10:114,238,129-116,369,485

```

```{r ld_plots_prc1, echo=FALSE}

source("modified.hm.R")


rccd1.BC.snps <- get_model_snps("RCCD1","Brain_Cortex")
rccd1.BC.snps <- unique(rccd1.BC.snps$rsid)
write.table(rccd1.BC.snps,file=save.dir%&%"trccd1.BC.snps.txt",quote=FALSE,sep="\n",row.names=FALSE,col.names=FALSE)

ap3s2.SI.snps <- get_model_snps("AP3S2","Small_Intestine_Terminal_Ileum")
ap3s2.SI.snps <- unique(ap3s2.SI.snps$rsid)
write.table(ap3s2.SI.snps,file=save.dir%&%"ap3s2.SI.snps.txt",quote=FALSE,sep="\n",row.names=FALSE,col.names=FALSE)

prc1.Panc.snps <- get_model_snps("PRC1","Pancreas")
prc1.Panc.snps <- unique(prc1.Panc.snps$rsid)
write.table(prc1.Panc.snps,file=save.dir%&%"prc1.Panc.snps.txt",quote=FALSE,sep="\n",row.names=FALSE,col.names=FALSE)

write.table(unique(c(ap3s2.SI.snps,rccd1.BC.snps))
            ,file=save.dir%&%"set2.snps.txt",quote=FALSE,sep="\n",row.names=FALSE,col.names=FALSE)

write.table(unique(c(rccd1.BC.snps,prc1.Panc.snps))
            ,file=save.dir%&%"set3.snps.txt",quote=FALSE,sep="\n",row.names=FALSE,col.names=FALSE)


ld.df <- dplyr::select(fread(save.dir%&%"r2_set3.txt"),-contains("RS_number"))
snps <- names(ld.df)
pos.vec <- get_snp_pos(snps)
na.snps <- snps[is.na(pos.vec)]
mat <- as.matrix(ld.df)
row.names(mat) <- snps
if (length(na.snps)>=1){
  remove.index <- match(na.snps,row.names(mat))
  mat <- mat[-remove.index,-remove.index]
}
pos.vec <- pos.vec[!is.na(pos.vec)]
snps <- row.names(mat)
  
Lab.palette <- colorRampPalette(c("dodgerblue4", "ghostwhite"),space = "Lab")

mycolors <- assign_color(snps,rccd1.BC.snps,prc1.Panc.snps,
                         col1="blue",col2="black",colboth="red")
keep.index <- !grepl("blue",mycolors)


hm <- LDheatmap_jmt(mat,color=Lab.palette(100),
            genetic.distances=pos.vec,distances="physical",add.map=TRUE,flip=TRUE,
            SNP.name=snps[keep.index],geneMapLocation=0.20,name = "myLDgrob",
            mycol=mycolors[keep.index],mycex=0.48)

#hm <- LDheatmap(mat,color=Lab.palette(100),
#            genetic.distances=pos.vec,distances="physical",add.map=TRUE,flip=TRUE,
#            SNP.name=snps,geneMapLocation=0.25,name = "myLDgrob")

#LDheatmap.addGenes(hm,chromosome="chr10",splice_variants=FALSE,non_coding=FALSE,genome="hg19")

# start = 114238129
# end = 116369485
# chr10:114,238,129-116,369,485

```


