---
title: "PredictorSNPs-GWAS-GTEx-Miami.Rmd"
author: "Jason Torres"
date: "June 7, 2016"
output:
  pdf_document:
    toc: yes
  html_document:
    fig_caption: yes
    highlight: textmate
    number_sections: yes
    theme: journal
    toc: yes
---

```{r setup}

library("LDheatmap")
library("data.table")
library("dplyr")
library("ggplot2")
library("gridExtra")
library("RColorBrewer")
library("ggrepel")

"%&%" <- function(a,b) paste0(a,b)

rds.dir <- "/Users/jtorres/Google Drive/Projects/MetaXcan/DIAGRAM-T2D-version3/RDS/"
ld.dir <- "/Users/jtorres/Google Drive/Projects/MetaXcan/DIAGRAM-T2D-version3/LD/"
eqtl.dir <- "/Volumes/im-lab/nas40t2/jason/projects/MetaXcan/results/predictors/by_locus/"
fig.dir <- "/Users/jtorres/Google Drive/Projects/MetaXcan/DIAGRAM-T2D-version3/figures/"

diagram.df <- readRDS("/Volumes/im-lab/nas40t2/jason/projects/ReferenceData/" %&% "diagramT2D/build37/diagram.df.RDS")
gera.df <- readRDS("/Volumes/im-lab/nas40t2/jason/projects/MetaXcan/meta_files/gera/gera.expanded.df.RDS")


modsnp.df <- readRDS(file=rds.dir %&% "mtxnsig-gene-pred.df.RDS")
modsnp.gtex.df <- readRDS(file=rds.dir %&% "modsnp.gtex.df.RDS")
uniq.loci.df <- readRDS(file=rds.dir %&% "uniq.loci.df.RDS")
mtxn.sig.loci.df <- readRDS(file=rds.dir%&%"mtxn.sig.loci.df.RDS")

loc_win_df <- function(){
  temp.df <- select(uniq.loci.df,one_of("chrom","locus","locus.start","locus.end"))
  out.df <- temp.df[!duplicated(temp.df),]
  return(out.df)
}

loc.win.df <- loc_win_df()

```


Generate a manhattan plot of GWAS SNP results within a specified window, highlighting select model SNPs  

```{r window_manhat, echo=FALSE}

window_manhat <- function(loc.refnum,gwas,mod,genename){
  # Note: Locus References number ranges from 1 to 57 
  chromo <- filter(loc.win.df,locus==loc.refnum)[1,1]
  loc.start <- filter(loc.win.df,locus==loc.refnum)[1,3]
  loc.end <- filter(loc.win.df,locus==loc.refnum)[1,4]
  # GWAS Manhattan data frame
  if (gwas=="GERA"){
    df <- gera.df 
  } else if (gwas=="DIAGRAM"){
    df <- diagram.df
  } else{
    print("gwas must be either DIAGRAM or GERA")
  }
  gwas.df <- filter(df,CHR==chromo,POS>=loc.start,POS<=loc.end) %>% 
    select(one_of("SNP","CHR","POS","P"))
  names(gwas.df) <- c("rsid","chr","pos","p")
  # Get Model SNP info for the interval 
  pred.df <- filter(modsnp.df,chr==chromo,
                    pos>=loc.start,pos<=loc.end,
                    gene==genename,model==mod) %>% 
    select(one_of("chr","rsid","gene","model","pos"))
  #if (!is.null(mod)) {pred.df <- filter(pred.df,model==mod)}
  #if (!is.null(genename)) {pred.df <- filter(pred.df,gene==genename)}
  p <- c()
  for (i in 1:dim(pred.df)[1]){
    snp <- pred.df$rsid[i]
    temp.df <- filter(gwas.df,rsid==snp)
    temp.df <- as.data.frame(temp.df)
    if (dim(temp.df)[1]==0){
      p <- append(p,NA)
    } else{
      pval <- temp.df[1,4]
      p <- append(p,pval)
    }
  }
  genemod <- paste(pred.df$gene," : ",gsub("TW_","",pred.df$model))
  pred.df <- cbind(pred.df,p,genemod)
  #gw.genes <- unique(strsplit(as.character(filter(mtxn.sig.loci.df,
  #                                                locus==loc.refnum)$mtxn.sig.gw.genes),
  #                            split=",")[[1]])
  #pred.df <- filter(pred.df,gene %in% gw.genes)
  maxval <- max(-log(gwas.df$p,base=10))
  if (maxval >= 50 & maxval < 100){
    ymax <- 100
  } else if (maxval > 20 & maxval < 50){
    ymax <- 50
  } else if (maxval > 10 & maxval<= 20){
    ymax <- 20
  } else if (maxval <= 10){
    ymax <- 10 
  } else{
    ymax <- 200 
  }
  #colfun <- colorRampPalette(brewer.pal(n=8,name="Accent"))
  plt <- ggplot(data=gwas.df,aes(x=pos/10^6,y=-log(p,base=10))) + 
    geom_point(size=2.5,alpha=0.7,shape=21,color="darkgray",fill="grey") + 
    theme_bw() + coord_cartesian(xlim=c(loc.start/10^6,loc.end/10^6),
                                 ylim=c(0,ymax+1),expand=FALSE) + 
    scale_y_continuous(breaks=seq(0,ymax,by=ymax/10)) + 
    xlab("Position on Chromosome " %&% chromo %&% " (Mb)") + 
    ylab(expression(-log[10](italic("p-value")))) + 
    geom_hline(yintercept=-log(5*10^-8,base=10)) + 
    ggtitle(gwas) + 
    geom_text_repel(data=arrange(pred.df,p)[1:20,],aes(x=pos/10^6,y=-log(p,base=10),
                                     label=rsid),size=2.25) + 
    geom_point(data=pred.df,aes(x=pos/10^6,y=-log(p,base=10),fill=genemod),
               size=2.5,shape=21) + 
    theme(plot.title = element_text(face="bold"),
          legend.position = "bottom",
          legend.key = element_rect(colour = "black"),
          legend.key.size = unit(0.4, "cm"),
          legend.text = element_text(size = 6),
          panel.grid.minor=element_blank()) +
    scale_fill_discrete(name="Model");plt
  return(plt)
}

```

Add a Chicago (Miami) plot of GTEx associations, highlighting select model SNPs

```{r add_miami, echo=FALSE}

window_miami_gtex <- function(loc.refnum,gwas,mod,genename){
  pltA <- window_manhat(loc.refnum,gwas,mod,genename)
  
  gtex.file <- eqtl.dir %&% mod %&% "_" %&% genename %&% ".gtex.txt.gz"
  gtex.df <- fread("cat " %&% gtex.file %&% " | zmore")
  chromo <- filter(loc.win.df,locus==loc.refnum)[1,1]
  loc.start <- filter(loc.win.df,locus==loc.refnum)[1,3]
  loc.end <- filter(loc.win.df,locus==loc.refnum)[1,4]
  maxval <- max(-log(gtex.df$gtex.p,base=10))
  if (maxval >= 50 & maxval < 100){
    ymax <- 100
  } else if (maxval > 20 & maxval < 50){
    ymax <- 50
  } else if (maxval > 10 & maxval<= 20){
    ymax <- 20
  } else if (maxval <= 10){
    ymax <- 10 
  } else{
    ymax <- 200 
  }  
  
  pred.df <- filter(modsnp.gtex.df,chr==chromo,
                    pos>=loc.start,pos<=loc.end,gene==genename,model==mod) %>% 
    select(one_of("chr","rsid","gene","model","pos","gtex.p"))
  genemod <- paste(pred.df$gene," : ",gsub("TW_","",pred.df$model))
  pred.df <- cbind(pred.df,genemod)
  
  pltB <- ggplot(data=gtex.df,aes(x=pos/10^6,y=-log(gtex.p,base=10))) + 
    geom_point(size=2.5,alpha=0.7,shape=21,color="darkgray",fill="grey") + 
    theme_bw() + coord_cartesian(xlim=c(loc.start/10^6,loc.end/10^6),
                                 ylim=c(0,ymax+1),expand=FALSE) + 
    ylab(expression(-log[10](italic("p-value")))) + 
    geom_text_repel(data=arrange(pred.df,gtex.p)[1:20,],aes(x=pos/10^6,y=-log(gtex.p,base=10),
                                     label=rsid),size=2.25) + 
    geom_point(data=pred.df,aes(x=pos/10^6,y=-log(gtex.p,base=10),fill=genemod),
               size=2.5,shape=21) + 
    theme(plot.title = element_text(face="bold"),
          legend.position = "none",
          panel.grid.minor=element_blank(),
          axis.text.x=element_blank(),
          axis.title.x=element_text(face="bold"),
          axis.ticks.x=element_blank())+
    scale_fill_discrete(name="Model") + 
    scale_y_reverse(breaks=seq(0,ymax,by=ymax/10)) + 
    ggtitle("GTEx") + xlab("\n\n\n"); pltB
  return(grid.arrange(pltA,pltB,ncol=1))
}

```


Differently formatted Manhattan/Miami Plot

```{r alt_plots, echo=FALSE}

window_manhatV2 <- function(loc.refnum,gwas,mod,genename){
  # Note: Locus References number ranges from 1 to 57 
  chromo <- filter(loc.win.df,locus==loc.refnum)[1,1]
  loc.start <- filter(loc.win.df,locus==loc.refnum)[1,3]
  loc.end <- filter(loc.win.df,locus==loc.refnum)[1,4]
  # GWAS Manhattan data frame
  if (gwas=="GERA"){
    df <- gera.df 
  } else if (gwas=="DIAGRAM"){
    df <- diagram.df
  } else{
    print("gwas must be either DIAGRAM or GERA")
  }
  gwas.df <- filter(df,CHR==chromo,POS>=loc.start,POS<=loc.end) %>% 
    select(one_of("SNP","CHR","POS","P"))
  names(gwas.df) <- c("rsid","chr","pos","p")
  # Get Model SNP info for the interval 
  pred.df <- filter(modsnp.df,chr==chromo,
                    pos>=loc.start,pos<=loc.end,
                    gene==genename,model==mod) %>% 
    select(one_of("chr","rsid","gene","model","pos"))
  #if (!is.null(mod)) {pred.df <- filter(pred.df,model==mod)}
  #if (!is.null(genename)) {pred.df <- filter(pred.df,gene==genename)}
  p <- c()
  for (i in 1:dim(pred.df)[1]){
    snp <- pred.df$rsid[i]
    temp.df <- filter(gwas.df,rsid==snp)
    temp.df <- as.data.frame(temp.df)
    if (dim(temp.df)[1]==0){
      p <- append(p,NA)
    } else{
      pval <- temp.df[1,4]
      p <- append(p,pval)
    }
  }
  genemod <- paste(pred.df$gene," : ",gsub("TW_","",pred.df$model))
  pred.df <- cbind(pred.df,p,genemod)
  #gw.genes <- unique(strsplit(as.character(filter(mtxn.sig.loci.df,
  #                                                locus==loc.refnum)$mtxn.sig.gw.genes),
  #                            split=",")[[1]])
  #pred.df <- filter(pred.df,gene %in% gw.genes)
  maxval <- max(-log(gwas.df$p,base=10))
  if (maxval >= 50 & maxval < 100){
    ymax <- 100
  } else if (maxval > 20 & maxval < 50){
    ymax <- 50
  } else if (maxval > 10 & maxval<= 20){
    ymax <- 20
  } else if (maxval <= 10){
    ymax <- 10 
  } else{
    ymax <- 200 
  }
  #colfun <- colorRampPalette(brewer.pal(n=8,name="Accent"))
  plt <- ggplot(data=gwas.df,aes(x=pos/10^6,y=-log(p,base=10))) + 
    geom_point(size=2.5,alpha=0.7,shape=21,color="darkgray",fill="grey") + 
    theme_bw() + coord_cartesian(xlim=c(loc.start/10^6,loc.end/10^6),
                                 ylim=c(0,ymax+1),expand=FALSE) + 
    scale_y_continuous(breaks=seq(0,ymax,by=ymax/10)) + 
   # xlab("") + 
    #xlab("Position on Chromosome " %&% chromo %&% " (Mb)") + 
    ylab(expression(-log[10](italic("p-value")))) + 
    geom_hline(yintercept=-log(5*10^-8,base=10)) + 
    ggtitle("\n\n"%&%gwas) + 
    geom_text_repel(data=arrange(pred.df,p)[1:20,],aes(x=pos/10^6,y=-log(p,base=10),
                                     label=rsid),size=2.25) + 
    geom_point(data=pred.df,aes(x=pos/10^6,y=-log(p,base=10),fill=genemod),
               size=2.5,shape=21) + 
    theme(plot.title = element_text(face="bold"),
          #axis.text.x=element_blank(),
          #axis.ticks.x=element_blank(),
          axis.title.x=element_blank(),
          #legend.position = "bottom",
          legend.position = "none",
          legend.key = element_rect(colour = "black"),
          legend.key.size = unit(0.4, "cm"),
          legend.text = element_text(size = 6),
          panel.grid.minor=element_blank()) +
    scale_fill_discrete(name="Model");plt
  return(plt)
}

window_miami_gtexV2 <- function(loc.refnum,gwas,mod,genename){
  pltA <- window_manhatV2(loc.refnum,gwas,mod,genename)
  
  gtex.file <- eqtl.dir %&% mod %&% "_" %&% genename %&% ".gtex.txt.gz"
  gtex.df <- fread("cat " %&% gtex.file %&% " | zmore")
  chromo <- filter(loc.win.df,locus==loc.refnum)[1,1]
  loc.start <- filter(loc.win.df,locus==loc.refnum)[1,3]
  loc.end <- filter(loc.win.df,locus==loc.refnum)[1,4]
  maxval <- max(-log(gtex.df$gtex.p,base=10))
  if (maxval >= 50 & maxval < 100){
    ymax <- 100
  } else if (maxval > 20 & maxval < 50){
    ymax <- 50
  } else if (maxval > 10 & maxval<= 20){
    ymax <- 20
  } else if (maxval <= 10){
    ymax <- 10 
  } else{
    ymax <- 200 
  }  
  
  pred.df <- filter(modsnp.gtex.df,chr==chromo,
                    pos>=loc.start,pos<=loc.end,gene==genename,model==mod) %>% 
    select(one_of("chr","rsid","gene","model","pos","gtex.p"))
  genemod <- paste(pred.df$gene," : ",gsub("TW_","",pred.df$model))
  pred.df <- cbind(pred.df,genemod)
  
  pltB <- ggplot(data=gtex.df,aes(x=pos/10^6,y=-log(gtex.p,base=10))) + 
    geom_point(size=2.5,alpha=0.7,shape=21,color="darkgray",fill="grey") + 
    theme_bw() + coord_cartesian(xlim=c(loc.start/10^6,loc.end/10^6),
                                 ylim=c(0,ymax+1),expand=FALSE) + 
    ylab(expression(-log[10](italic("p-value")))) + 
    xlab("Position on Chromosome " %&% chromo %&% " (Mb)") + 
    geom_text_repel(data=arrange(pred.df,gtex.p)[1:20,],aes(x=pos/10^6,y=-log(gtex.p,base=10),
                                     label=rsid),size=2.25) + 
    geom_point(data=pred.df,aes(x=pos/10^6,y=-log(gtex.p,base=10),fill=genemod),
               size=2.5,shape=21) + 
    theme(plot.title = element_text(face="bold"),
          #axis.title.y=element_text(margin=margin(0,12.25,0,0)),
          #axis.title.y=element_text(margin=margin(0,10,0,2.25)),
          legend.position = "bottom",
          legend.key = element_rect(colour = "black"),
          legend.key.size = unit(0.4, "cm"),
          legend.text = element_text(size = 6),
          axis.title.x=element_text(face="bold"),
          panel.grid.minor=element_blank()) + #, 
          #axis.text.x=element_blank(),
          #axis.title.x=element_text(face="bold")) + 
          #axis.ticks.x=element_blank())+
    scale_fill_discrete(name="Model") + 
    scale_y_reverse(breaks=seq(0,ymax,by=ymax/10)); pltB 
    #ggtitle("GTEx") + 
    #xlab("\n\n\n"); pltB
  return(grid.arrange(pltA,pltB,ncol=1))
}

```


Write loci predictor plots 

```{r write_miami, echo=FALSE}

save_all_miami_files <- function(){
  # Iterate over loci in mtxn.sig.loci.df
  for (i in 1:dim(mtxn.sig.loci.df)[1]){
    locnum <- mtxn.sig.loci.df[i,2]
    print(locnum)
    temp.df <- filter(mtxn.sig.loci.df,locus==locnum)
    lw.genes <- strsplit(as.character(temp.df$mtxn.sig.lw.genes),",")[[1]]
    lw.mods <- strsplit(as.character(temp.df$mtxn.sig.lw.models),",")[[1]]
    if (!is.na(lw.genes) & !is.na(lw.mods) & length(lw.genes)==length(lw.mods)){
      save.dir <- fig.dir %&% "predictors/locus_" %&% locnum %&% "/"
      if (dir.exists(save.dir)==FALSE){
        dir.create(save.dir)
      }
      for (j in 1:length(lw.genes)){
        plt1<-NULL;plt2<-NULL
        g <- lw.genes[j]
        m <- lw.mods[j]
        print(c(g," : ",m))
        # <- window_miami_gtex(3,gwas="DIAGRAM",mod=m,genename=g)
        save.name <- save.dir %&% "loc" %&% locnum %&% 
          "." %&% g %&% "." %&% m %&% "." %&% "DIAGRAM.miami.pdf"
        tryCatch(plt1 <- window_miami_gtexV2(locnum,gwas="DIAGRAM",mod=m,genename=g),
                 error=function(e) e,finally="try statement")   
        tryCatch(ggsave(plot=plt1,filename=save.name,width=6,height=8),
                 error=function(e) e)
        save.name2 <- save.dir %&% "loc" %&% locnum %&% 
          "." %&% g %&% "." %&% m %&% "." %&% "GERA.miami.pdf"
        tryCatch(plt2 <- window_miami_gtexV2(locnum,gwas="GERA",mod=m,genename=g),
                 error=function(e) e,finally="try statement")   
        tryCatch(ggsave(plot=plt2,filename=save.name2,width=6,height=8),
                 error=function(e) e)        
      }
    }
  }
}

save_all_miami_files()

```





```{r testing}

test <- window_miami_gtexV2(37,"DIAGRAM","TW_Artery-Tibial","GPAM")
test2 <- window_miami_gtexV2(37,gwas="GERA","TW_Artery-Tibial","GPAM")

test <- window_miami_gtex(37,"DIAGRAM","TW_Artery-Tibial","GPAM")
test2 <- window_miami_gtex(37,gwas="GERA","TW_Artery-Tibial","GPAM")

ggsave(test,filename=fig.dir %&% "TW_Artery-Tibial_GPAM_DIAGRAM.miami.pdf",
       width=6,height=8)
ggsave(test2,filename=fig.dir %&% "TW_Artery-Tibial_GPAM_GERA.miami.pdf",
       width=6,height=8)

```








```{r produce_sq_matrix, echo=FALSE}

ld_heatmap <- function(locnum,gwas="DIAGRAM"){
  temp.df <- filter(mtxn.sig.loci.df,locus==locnum) %>% 
    select(one_of("chrom","locus","locus.start","locus.end"))
  temp.df <- temp.df[!duplicated(temp.df),]
  chromo <- temp.df$chrom
  df <- as.data.frame(fread(ld.dir %&% "locus."%&%locnum%&%"."%&%gwas%&%".ld.txt"))
  snps <- df$RS_number
  if (gwas=="DIAGRAM"){
    gwas.df <- filter(diagram.df,CHR==chromo) %>% 
      arrange(P)
  } else if(gwas=="GERA"){
    gwas.df <- filter(gera.df,CHR==chromo) %>% 
      arrange(P)
  } else{
    print("Must enter valid gwas name: DIAGRAM or GERA")
  }
  pos.vec <- c()
  for (snp in snps){
    pos <- filter(gwas.df,SNP==snp)$POS
    #print(pos)
    if (length(pos)==1){
      pos.vec <- append(pos.vec,pos)
    } else{
      pos.vec <- append(pos.vec,NA)
    }
  }
  # Ensure there are no NA values in the square matrix 
  na.snps <- snps[is.na(pos.vec)]
  mat <- as.matrix(df[,2:dim(df)[2]])
  row.names(mat) <- snps
  if (length(na.snps)>=1){
    remove.index <- match(na.snps,row.names(mat))
    mat <- mat[-remove.index,-remove.index]
  }
  pos.vec <- pos.vec[!is.na(pos.vec)]
  snps <- row.names(mat)
  # Get most significant SNP 
  maxsnp <- (filter(gwas.df,SNP %in% snps) %>% arrange(P))$SNP[1]
  # Get model SNPs
  mod.snps <- setdiff(intersect(write_locus_snps(locnum,gwas),snps),maxsnp)
  # Label no more than the top 20 SNPs in the plot annotation
  lab.df <- filter(gwas.df,SNP %in% mod.snps) %>% arrange(P)
  if (dim(lab.df)[1]>=20){
    mod.snps <- lab.df$SNP[1:20]
  }
  # Determine plotting SNPs 
  plot.snps <- append(maxsnp,mod.snps)
  col.vec <- c("red2",rep("dodgerblue4",length(mod.snps)))
  #cex.vec <- c(0.6, rep(0.475,length(mod.snps)))
  # Set palette for LDheatmap 
  Lab.palette <- colorRampPalette(c("dodgerblue4", "ghostwhite"),
                                space = "Lab")
  # Generate the plot
  hm <- LDheatmap_jmt(mat, LDmeasure = "r", genetic.distances = pos.vec,
                  title = "", add.map = TRUE,
                  color = Lab.palette(100),flip=TRUE,
                  SNP.name=plot.snps, #SNP.name= maxsnp
                  name = "myLDgrob", add.key = TRUE, 
                  mycol=col.vec,#,"black"
                  mycex=0.4)#cex.vec)
  return(hm)
  #hm.genes <- LDheatmap.addGenes(hm, chr="chr3", genome="hg19",splice_variants=FALSE,non_coding=FALSE)  
  
}


save_locus <- function(locnum){
  pltA <- window_manhat(locnum,gwas="DIAGRAM")
  pltB <- ld_heatmap(locnum, gwas="DIAGRAM")
  pltC <- window_manhat(locnum,gwas="GERA")
  pltD <- ld_heatmap(locnum, gwas="GERA")
  plt <- grid.arrange(ggplotGrob(pltA),ggplotGrob(pltC),
               pltB$LDheatmapGrob,pltD$LDheatmapGrob,
               ncol=2,nrow=2)
  ggsave(plt,file=ld.dir%&%"locus."%&%locnum%&%".manhatLD.pdf",
         width=12,height=8)
}

save_locus(6)
save_locus(37)
save_locus(39)
save_locus(42)
#ld_heatmap(6, gwas="DIAGRAM")

```


```{r make_plots, echo=FALSE}

gw.loci <- na.omit(select(mtxn.sig.loci.df,one_of("locus","mtxn.sig.gw.genes")))$locus
#lw.loci <- na.omit(select(mtxn.sig.loci.df,one_of("locus","mtxn.sig.lw.genes")))$locus
test <- mtxn.sig.loci.df[grepl("KLHDC5",mtxn.sig.loci.df$mtxn.sig.gw.genes),]$locus
test <- mtxn.sig.loci.df[grepl("KCNJ11",mtxn.sig.loci.df$mtxn.sig.gw.genes),]$locus
test <- mtxn.sig.loci.df[grepl("TCF7L2",mtxn.sig.loci.df$mtxn.sig.gw.genes),]$locus


window_manhat(loc.refnum=37,gwas="DIAGRAM")
window_manhat(loc.refnum=6,gwas="GERA")


```




```{r trying-things-out}


### HOW TO DO A CHICAGO PLOT 
#tpltA <- ggplot(data=gwas.df,aes(x=pos/10^6,y=-log(p,base=10))) + geom_point() + theme_bw()
#tpltB <- ggplot(data=gwas.df,aes(x=pos/10^6,y=-log(p,base=10))) + geom_point() + scale_y_reverse() + theme_bw() + 
#  theme(axis.text.x=element_blank(),
#        axis.title.x=element_blank(),
#        axis.ticks.x=element_blank())
#grid.arrange(tpltA,tpltB,ncol=1)

#tt <- LDheatmap.addGrob(hm,rectGrob(gp=gpar(col="white")),height=1)
#pushViewport(viewport(x=.48,y=.8,width=0.85,height=1))
#grid.draw(ggplotGrob(plt))
#popViewport(1)
#dev.off()

```
