---
title: "09.3_PredictorSNPs-LDHeatmap.Rmd"
author: "Jason Torres"
date: "July 1, 2016"
output: pdf_document
---

```{r setup}

library("LDheatmap")
library("data.table")
library("dplyr")
library("ggplot2")

"%&%" <- function(a,b) paste0(a,b)


root.dir <- "/Users/jtorres/Google Drive/Projects/MetaXcan/DIAGRAM-T2D-version3/"
rds.dir <- root.dir %&% "RDS/"
ld.dir <- root.dir %&% "LD/"

source(root.dir %&% "modified.hm.R")

diagram.df <- readRDS("/Volumes/im-lab/nas40t2/jason/projects/ReferenceData/" %&% "diagramT2D/build37/diagram.df.RDS")
gera.df <- readRDS("/Volumes/im-lab/nas40t2/jason/projects/MetaXcan/meta_files/gera/gera.expanded.df.RDS")  

modsnp.df <- readRDS(file=rds.dir %&% "mtxnsig-gene-pred.df.RDS")
uniq.loci.df <- readRDS(file=rds.dir %&% "uniq.loci.df.RDS")
mtxn.sig.loci.df <- readRDS(file=rds.dir%&%"mtxn.sig.loci.df.RDS")


```


```{r write_mod_snp_list, echo=FALSE}

write_locus_snps <- function(locnum,gwas="DIAGRAM"){
  temp.df <- filter(mtxn.sig.loci.df,locus==locnum) %>% 
    select(one_of("chrom","locus","locus.start","locus.end"))
  temp.df <- temp.df[!duplicated(temp.df),]
  chromo <- temp.df$chrom
  lstart <- temp.df$locus.start; lend <- temp.df$locus.end
  if (gwas=="DIAGRAM"){
    gwas.df <- filter(diagram.df,CHR==chromo,POS >= lstart, POS <= lend) %>% 
      arrange(POS)
  } else if(gwas=="GERA"){
    gwas.df <- filter(gera.df,CHR==chromo,POS >= lstart, POS <= lend) %>% 
      arrange(POS)
  } else{
    print("Must enter valid gwas name: DIAGRAM or GERA")
  }
  ##head.df <- head(gwas.df,n=100)
  ##snp.vec <- head.df$SNP
  ##min.snps <- gwas.df$SNP[1:3]
  ##max.snps <- gwas.df$SNP[(length(gwas.df$SNP)-3):length(gwas.df$SNP)]
  snp.vec <- c(head(arrange(gwas.df,P)$SNP,1))
  #snp.vec <- append(c(min.snps,max.snps),snp.vec)
  # Ensure model SNPs included 
  siggenes <- strsplit(as.character(filter(mtxn.sig.loci.df,locus==locnum)$mtxn.sig.gw.genes),split=",")[[1]]
  sigmods <- strsplit(as.character(filter(mtxn.sig.loci.df,locus==locnum)$mtxn.sig.gw.models),split=",")[[1]]
  mod.snps <- c()
  for (i in 1:length(siggenes)){
    g <- siggenes[i]; m <- sigmods[i]
    track.snps <- unique(filter(modsnp.df,gene==g,model==m)$rsid)
    track.snps <- track.snps[track.snps %in% gwas.df$SNP]
    mod.snps <- append(mod.snps,track.snps)
  }
  snp.vec <- unique(append(snp.vec,mod.snps))
  index <- match(snp.vec,gwas.df$SNP)
  pos.vec <- gwas.df$POS[index]
  snp.vec <- snp.vec[!is.na(pos.vec)]
  write.table(snp.vec,file=ld.dir%&%"locus."%&%locnum%&%"."%&%gwas%&%".snps.txt",
              sep="\t",quote=FALSE,row.names=FALSE)
  return(unique(mod.snps))
}

```

```{r write_files, echo=FALSE}

gw.loci <- na.omit(select(mtxn.sig.loci.df,one_of("locus","mtxn.sig.gw.genes")))$locus

for (index in gw.loci){
  print(index)
  file1 <- write_locus_snps(index,gwas="DIAGRAM")
  file2 <- write_locus_snps(index,gwas="GERA")  
}

```



Need to manually generate LD Link - LD Matrix files, saves as locus.locname.gwasname.ld.txt




